---
title: "neoadjuvant pancancer WES analysis"
author: "Pedro Batista Tan"
date: "`r strftime(Sys.Date(), '%B %d %Y')`"
output: 
  html_document:
    toc: true
    number_sections: true
params:
  export: TRUE
  output_dir: "WES_analysis"
  metadata: ""
  filter_and_parse_vcfs: FALSE
  exclude_niche_msi: FALSE
---

Author comment: File description comment: The purpose of this script is to parse results from WES analysis vcf files, quality logs, comparing with RNAseq data filter_and_parse_vcfs


# Loading Packages

```{r, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, warning = FALSE, message = FALSE, dpi = 300, error = TRUE
  # fig.width = 10, fig.height = 6, output: html_notebook
)

# Install if necessary install.packages("tidyverse")
# Install if necessary BiocManager::install("DESeq2")

# library("devtools")
# devtools::install("~/Tools/TMBleR")
# BiocManager::install("VariantAnnotation")
# BiocManager::install("maftools")

library(tidyverse)
library(readxl)
# library(TMBleR)
library(maftools)
library(ICAMS)
source("helper_functions.R")
source("oncomatrix.R")
library(VariantAnnotation)
library(ggpubr)
library(REdaS)
library(paletteer)
library(corrplot)
library(ggpointdensity)
library(export)
library(meta)
# library(metafor)
library(lsa)
library(forestplot)
library(export)
library(MutationalPatterns)
# library(data.table)
# devtools::install_github("mskcc/facets-suite")
library("facetsSuite")
# BiocManager::install("MutationalPatterns")

# install.packages("forestplot")
# library("devtools")
# devtools::install_github("ashishjain1988/MAFDash")
# library(MAFDash)
# library(data.table)

MakeDataFrameFromVCF <- getFromNamespace("MakeDataFrameFromVCF", "ICAMS")

set.seed(42)
select <- dplyr::select
rename <- dplyr::rename
here <- here::here

packageVersion("GSVA")
```

# Setup output folder

```{r Setup output, echo = FALSE, include = FALSE}
print(params)

export <- params$export
filter_and_parse_vcfs <- params$filter_and_parse_vcfs

# generate output directory
output_dir <- here("data", "processed", params$output_dir)
wes_dir <- here("data", "processed", "WES_analysis")

metadata_dir <- here("data", "processed", "metadata")
print(output_dir)
# if output_dir does not exist, create it
if (!dir.exists(output_dir)) {
    dir.create(output_dir)
    dir.create(here(output_dir, "pathway_proportions"))
    
    } else {
  print("Output folder already exists")
}

```

# ----------

# Load metadata

```{r Load metadata}
dna_metadata <- read_csv2(here(metadata_dir, "combined_dna_metadata.csv"))

# WES target sizes
WES_target_sizes <- data.frame(
  Trial = c("IMCISION", "PRADO", "OpACIN_neo", "MATISSE",
            "NICHE-MSI", "NICHE-MSS", "NABUCCO", "BELLINI"),
  Kit = c(rep("Twist exome Refseq", 4), rep("Exome-IDT V1", 4)),
  WES_target_size = c(rep(36.72, 4), rep(39.02, 4)))

WES_target_sizes

histology_df <- data.frame(
  Trial = c("BELLINI", "IMCISION", "NABUCCO", "NICHE-MSS", "NICHE-MSI",
            "OpACIN_neo", "PRADO", "MATISSE"),
  Histology = c("TBNC", "HNSCC", "Urothelial", "Colon", "Colon",
                "LN_melanoma", "LN_melanoma", "CSCC"))

compare_r_nr_trials <- c("Combined", "BELLINI", "IMCISION", "MATISSE",
                         "NABUCCO", "NICHE-MSS", "OpACIN_neo", "PRADO")

trial_list <- c("Combined", "BELLINI", "IMCISION", "MATISSE", "NABUCCO", "NICHE-MSS",  "NICHE-MSI", "OpACIN_neo", "PRADO")

trial_group_map <- data.frame(Trial = c(trial_list, trial_list, trial_list),
           Group = c(trial_list, sprintf("%s_R", trial_list), sprintf("%s_NR", trial_list)),
           Response = c(
             rep("Combined", length(trial_list)),
             rep("R", length(trial_list)),
             rep("NR", length(trial_list)))
)

all_cohort_patient_metadata <- read_csv2(
  here(metadata_dir, "all_cohort_patient_metadata.csv"))

```

# Annotation color

```{r}

# Define colours for each group
Color_NR <- "#F8766D"
Color_R <- "#00BA38" # "#619CFF"
Color_NA <- "#CD9600"
Color_Pre <- "#619CFF"
Color_Post <- "#F564E3"
Color_Depleted <- "#7CAE00"
Color_Fibrotic <- "#00BFC4"
Color_Inflamed_fibrotic <- "#C77CFF"
Color_Inflamed_non_fibrotic <- "#F8766D"
Color_HNSCC <- "#009292"
Color_CSCC <- "#CD9600"
Color_Urothelial <- "#B66DFF"
Color_LN_melanoma <- "#00BFC4"
Color_Colon <- "#F8766D"
Color_Colon_MSI <- "#B73A3A"
Color_Colon_MSS <- "#23C26F"
Color_TNBC <- "#FEB5DA"
Color_BELLINI <- "#FEB5DA"
Color_IMCISION <- "#009292"
Color_MATISSE <- "#CD9600"
Color_NABUCCO <- "#B66DFF"
Color_BELLINI <- "#FEB5DA"
Color_NICHE_MSS <- "#23C26F"
Color_NICHE_MSI <- "#B73A3A"
Color_OpACIN_neo <- "#00BFC4"
Color_PRADO <- "#619CFF"

annotation_color_list = list(
  Response = c(NR = Color_NR, R = Color_R, R_NR = "#CD9600"),
  Treatment = c(Pre = Color_Pre, Post = Color_Post),
  Histology = c(LN_melanoma = Color_LN_melanoma, Urothelial = Color_Urothelial,
                Colon = Color_Colon, HNSCC = Color_HNSCC, TBNC = Color_TNBC,
                CSCC = Color_CSCC),
  Trial = c(BELLINI = Color_BELLINI, IMCISION = Color_IMCISION,
            MATISSE = Color_MATISSE, NABUCCO = Color_NABUCCO,
            `NICHE-MSS` = Color_NICHE_MSS, `NICHE-MSI` = Color_NICHE_MSI,
            OpACIN_neo = Color_OpACIN_neo, PRADO = Color_PRADO))

trial_colors <- c(BELLINI = Color_BELLINI, IMCISION = Color_IMCISION,
            MATISSE = Color_MATISSE, NABUCCO = Color_NABUCCO,
            `NICHE-MSS` = Color_NICHE_MSS, `NICHE-MSI` = Color_NICHE_MSI,
            OpACIN_neo = Color_OpACIN_neo, PRADO = Color_PRADO)

trial_colors_combined <- c(Combined = "#999999", BELLINI = Color_BELLINI, IMCISION = Color_IMCISION,
            MATISSE = Color_MATISSE, NABUCCO = Color_NABUCCO,
            `NICHE-MSS` = Color_Colon_MSS, `NICHE-MSI` = Color_NICHE_MSI,
            OpACIN_neo = Color_OpACIN_neo, PRADO = Color_PRADO)

Bagaev_tme_colors <- c(Color_Depleted, Color_Fibrotic,
                       Color_Inflamed_fibrotic, Color_Inflamed_non_fibrotic)

histology_msi_mss_colors <- c(`Colon-MSI`= Color_Colon_MSI, `Colon-MSS`= Color_Colon_MSS,
                              CSCC = Color_CSCC, HNSCC = Color_HNSCC, 
                              LN_melanoma = Color_LN_melanoma, TBNC = Color_TNBC,
                              Urothelial = Color_Urothelial)

```

# ----------

# Load MAF files

```{r}
combined_maf_raw <- readRDS(here(output_dir, "combined_maf_raw.rds"))
combined_maf_df <- read_csv2(here(output_dir, "combined_maf_df.csv"))

combined_maf_df %>% filter(is.na(Trial)) %>% pull(Source_MAF) %>% table(useNA = "ifany")
# combined_maf_df_dmmr <- combined_maf_df %>% filter((Trial %in% "NICHE-MSI" & Tumor_nr == "Tumor_2"))

combined_maf_df <- combined_maf_df %>%
  filter(!CF_sample_tumor %in% c("S2697Nr44", "S2697Nr58")) %>%
  filter(!Patient %in% c("NAB_S33", "NAB_S52", "NAB_S41")) %>%
  filter(!is.na(CF_sample_tumor)) %>%
  filter(!is.na(Trial)) %>%
  filter(!(Trial %in% "NICHE-MSI" & Tumor_nr == "Tumor_2"))

combined_maf_df %>% select(Trial, Patient) %>% unique() %>% dim()
combined_maf_df %>% select(Trial, Patient) %>% unique() %>% pull(Trial) %>% table()

sample_name_df <- read_csv2(here(output_dir, "sample_name_df.csv"))
dna_metadata <- dna_metadata %>% left_join(sample_name_df) 

# Only a few patients do not have WES data
dna_metadata$Patient[!dna_metadata$Patient %in% combined_maf_df$Patient]

# Remove files with missing data on MAF
dna_metadata %>%
  filter(!CF_sample_tumor %in% combined_maf_df$CF_sample_tumor) %>%
  factor_summary()

dna_metadata_f <- dna_metadata %>%
  filter(CF_sample_tumor %in% combined_maf_df$CF_sample_tumor) %>%
  filter(!Patient %in% c("NAB_S33", "NAB_S52", "NAB_S41")) %>%
  filter(!(Trial %in% "NICHE-MSI" & Tumor_nr == "Tumor_2"))

dna_metadata_f %>% write_csv2(here(output_dir, "dna_metadata_maf.csv"))
dna_metadata_f <- read_csv2(here(wes_dir, "dna_metadata_maf.csv"))

# Sensitivity analysis excluding NICHE2 dMMR MSI samples
if (params$exclude_niche_msi) {
  dna_metadata_f <- dna_metadata_f %>% filter(!Trial %in% c("NICHE-MSI"))
}

dna_metadata_f$Response %>% table()
dna_metadata_f$Response %>% table() %>% prop.table() %>% round(2)

sample_name_df$CF_sample_tumor[
  !sample_name_df$CF_sample_tumor %in% dna_metadata$CF_sample_tumor]

combined_maf <- subsetMaf(combined_maf_raw,
                          tsb = dna_metadata_f$Tumor_Sample_Barcode)

combined_maf@data$Source_MAF %>% n_distinct()
combined_maf@data$Tumor_Sample_Barcode %>% n_distinct()
dna_metadata %>% filter(!Patient %in% combined_maf_df$Patient)

combined_maf_df_samples <- combined_maf_df %>%
  select(Patient, Trial, Tumor_Sample_Barcode) %>%
  unique() %>%
  filter(!is.na(Patient))

combined_maf_df_patients <- combined_maf_df %>%
  select(Patient, Trial) %>%
  unique() %>%
  filter(!is.na(Patient))

n_distinct(combined_maf_df_patients$Patient)
table(combined_maf_df_patients$Trial)

combined_maf_df_samples$Patient %>%
  table() %>%
  as.data.frame() %>%
  filter(Freq > 1)

combined_maf_df %>%
  select(Patient, Trial) %>%
  unique() %>%
  filter(!is.na(Patient)) %>%
  pull(Trial) %>%
  table()
```

# Common variants, remove?
```{r}
table(combined_maf_df$FILTER)
combined_maf_df %>% filter(FILTER == "common_variant") %>% view()

```

# ----------

# Add maf clinical annotations

```{r}
combined_maf@clinical.data <- combined_maf@clinical.data %>%
  left_join(dna_metadata %>%
              select(-c(Source_MAF, CF_sample_tumor, CF_sample_gl))) 

combined_maf@data$Symbol_Variant <- sprintf(
  "%s_%s", combined_maf@data$SYMBOL,
  combined_maf@data$HGVSp_Short)

```

# Explore MAF

```{r, fig.width = 16, fig.height = 16}

plotmafSummary(maf = combined_maf, rmOutlier = TRUE,
               addStat = 'median', dashboard = TRUE, titvRaw = FALSE)

oncoplot(combined_maf,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list)

combined_maf@variants.per.sample

combined_maf_titv = titv(maf = combined_maf, plot = FALSE,
                         useSyn = TRUE)

plotTiTv(res = combined_maf_titv)

titv_counts <- combined_maf_titv$raw.counts %>%
  left_join(dna_metadata_f)  %>%
  filter(!is.na(Trial)) %>%
  pivot_longer(cols = c("C>A", "C>G", "C>T", "T>C", "T>A", "T>G"),
               names_to = "TiTv", values_to = "Freq") 

titv_counts %>%
  ggplot(aes(x = TiTv, y = Freq)) + geom_boxplot() +
           facet_wrap(~Trial)+
  labs(x = "") +
  theme_Publication()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

titv_fractions <- combined_maf_titv$fraction.contribution %>%
  left_join(dna_metadata_f) %>%
  filter(!is.na(Trial)) %>%
  pivot_longer(cols = c("C>A", "C>G", "C>T", "T>C", "T>A", "T>G"),
               names_to = "TiTv", values_to = "Proportion")

titv_fractions %>%
  ggplot(aes(x = TiTv, y = Proportion)) + geom_boxplot() +
           facet_wrap(~Trial) +
  labs(x = "") +
  theme_Publication()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

combined_maf_titv$TiTv.fractions %>%
  mutate(Ti_tv_ratio = Ti / Tv) %>%
  left_join(dna_metadata_f) %>%
  filter(!is.na(Trial)) %>%
  ggplot(aes(x = Ti_tv_ratio, fill = Trial)) +
  geom_density(alpha = 0.5)+
  theme_Publication() 

lollipopPlot(
  maf = combined_maf,
  gene = 'TP53',
  AACol = 'HGVSp',
  showMutationRate = TRUE
)

plotProtein(gene = "TP53", refSeqID = "NM_000546")

rainfallPlot(maf = combined_maf, detectChangePoints = TRUE, pointSize = 0.4)

laml.mutload = tcgaCompare(
  maf = combined_maf, cohortName = 'Neoadj. Pancan',
  logscale = TRUE, capture_size = 39)

plotVaf(combined_maf)

somaticInteractions(maf = combined_maf, top = 25, pvalue = c(0.05, 0.1))
```

# Plot variant statistics

```{r Plot variant statistics, fig.width = 5, fig.height = 6}
combined_maf_df %>%
  select(Hugo_Symbol, Chromosome, Variant_Classification,
         Variant_Type,IMPACT, Feature_type, BIOTYPE, Consequence,
         IMPACT, CLIN_SIG, VARIANT_CLASS, AF, EUR_AF,
         gnomAD_AF, gnomAD_NFE_AF, VAF, n_depth) %>%
  mutate_if(is.character, as.factor) %>%
  summary(maxsum = 25)

table(combined_maf_df$Chromosome) %>%
  as.data.frame() %>%
  setNames(c("Chromosome", "Freq")) %>%
  ggplot(aes(x = reorder(Chromosome, Freq), y = Freq)) +
  geom_col() + theme_Publication() + coord_flip() +
  labs(x = "Chromosome")

table(combined_maf_df$Variant_Classification) %>%
  as.data.frame() %>%
  setNames(c("Variant_Classification", "Freq")) %>%
  ggplot(aes(x = Variant_Classification , y = Freq)) +
  geom_col() + theme_Publication() + coord_flip()

table(combined_maf_df$Variant_Type) %>%
  as.data.frame() %>%
  setNames(c("Variant_Type", "Freq")) %>%
  ggplot(aes(x = Variant_Type , y = Freq)) +
  geom_col() + theme_Publication()

table(combined_maf_df$IMPACT) %>%
  as.data.frame() %>%
  setNames(c("IMPACT", "Freq")) %>%
  ggplot(aes(x = IMPACT , y = Freq)) +
  geom_col() + theme_Publication()

combined_maf_df %>%
  ggplot(aes(x = n_depth, fill = Trial)) + geom_density(alpha = 0.5) +
  theme_Publication()

combined_maf_df %>%
  ggplot(aes(x = VAF, fill = Trial)) + geom_density(alpha = 0.5) +
  theme_Publication()

# Indels
##INFO=<ID=IC,Number=1,Type=Integer,Description="Number of times RU repeats in the indel allele">
##INFO=<ID=IHP,Number=1,Type=Integer,Description="Largest reference interrupted homopolymer length intersecting with the indel">
##INFO=<ID=MQ,Number=1,Type=Float,Description="RMS Mapping Quality">
##INFO=<ID=MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">
##INFO=<ID=NT,Number=1,Type=String,Description="Genotype of the normal in all data tiers, as used to classify somatic variants. One of {ref,het,hom,conflict}.">
##INFO=<ID=OVERLAP,Number=0,Type=Flag,Description="Somatic indel possibly overlaps a second indel.">
##INFO=<ID=QSI,Number=1,Type=Integer,Description="Quality score for any somatic variant, ie. for the ALT haplotype to be present at a significantly different frequency in the tumor and normal">
##INFO=<ID=QSI_NT,Number=1,Type=Integer,Description="Quality score reflecting the joint probability of a somatic variant and NT">
##INFO=<ID=RC,Number=1,Type=Integer,Description="Number of times RU repeats in the reference allele">
##INFO=<ID=RU,Number=1,Type=String,Description="Smallest repeating sequence unit in inserted or deleted sequence">
##INFO=<ID=SGT,Number=1,Type=String,Description="Most likely somatic genotype excluding normal noise states">
##INFO=<ID=SOMATIC,Number=0,Type=Flag,Description="Somatic mutation">
##INFO=<ID=SomaticEVS,Number=1,Type=Float,Description="Somatic Empirical Variant Score (EVS) expressing the phred-scaled probability of the call being a false positive observation.">
##INFO=<ID=TQSI,Number=1,Type=Integer,Description="Data tier used to compute QSI">
##INFO=<ID=TQSI_NT,Number=1,Type=Integer,Description="Data tier used to compute QSI_NT">
##INFO=<ID=ANN,Number=.,Type=String,Description="Functional annotations: 'Allele | Annotation | Annotation_Impact | Gene_Name | Gene_ID | Feature_Type | Feature_ID | Transcript_BioType | Rank | HGVS.c | HGVS.p | cDNA.pos / cDNA.length | CDS.pos / CDS.length | AA.pos / AA.length | Distance | ERRORS / WARNINGS / INFO' ">
##INFO=<ID=LOF,Number=.,Type=String,Description="Predicted loss of function effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected'">
##INFO=<ID=NMD,Number=.,Type=String,Description="Predicted nonsense mediated decay effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected'">
##VEP="v106" time="2023-03-28 23:15:47" cache="/.vep/homo_sapiens/106_GRCh38" ensembl=106.f4b50c6 ensembl-funcgen=106.027e023 ensembl-variation=106.2aa7a5d ensembl-io=106.6eafdaa 1000genomes="phase3" COSMIC="92" ClinVar="202109" ESP="V2-SSA137" HGMD-PUBLIC="20204" assembly="GRCh38.p13" dbSNP="154" gencode="GENCODE 40" genebuild="2014-07" gnomAD="r2.1.1" polyphen="2.2.2" regbuild="1.0" sift="sift5.2.2"
##INFO=<ID=CSQ,Number=.,Type=String,Description="Consequence annotations from Ensembl VEP. Format: Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|VARIANT_CLASS|SYMBOL_SOURCE|HGNC_ID|CANONICAL|MANE_SELECT|MANE_PLUS_CLINICAL|TSL|APPRIS|CCDS|ENSP|SWISSPROT|TREMBL|UNIPARC|UNIPROT_ISOFORM|GENE_PHENO|SIFT|PolyPhen|DOMAINS|miRNA|AF|AFR_AF|AMR_AF|EAS_AF|EUR_AF|SAS_AF|AA_AF|EA_AF|gnomAD_AF|gnomAD_AFR_AF|gnomAD_AMR_AF|gnomAD_ASJ_AF|gnomAD_EAS_AF|gnomAD_FIN_AF|gnomAD_NFE_AF|gnomAD_OTH_AF|gnomAD_SAS_AF|MAX_AF|MAX_AF_POPS|FREQS|CLIN_SIG|SOMATIC|PHENO|PUBMED|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|TRANSCRIPTION_FACTORS">

# SNVS
##FORMAT=<ID=AU,Number=2,Type=Integer,Description="Number of 'A' alleles used in tiers 1,2">
##FORMAT=<ID=CU,Number=2,Type=Integer,Description="Number of 'C' alleles used in tiers 1,2">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read depth for tier1 (used+filtered)">
##FORMAT=<ID=FDP,Number=1,Type=Integer,Description="Number of basecalls filtered from original read depth for tier1">
##FORMAT=<ID=GU,Number=2,Type=Integer,Description="Number of 'G' alleles used in tiers 1,2">
##FORMAT=<ID=SDP,Number=1,Type=Integer,Description="Number of reads with deletions spanning this site at tier1">
##FORMAT=<ID=SUBDP,Number=1,Type=Integer,Description="Number of reads below tier1 mapping quality threshold aligned across this site">
##FORMAT=<ID=TU,Number=2,Type=Integer,Description="Number of 'T' alleles used in tiers 1,2">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Combined depth across samples">
##INFO=<ID=MQ,Number=1,Type=Float,Description="RMS Mapping Quality">
##INFO=<ID=MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">
##INFO=<ID=NT,Number=1,Type=String,Description="Genotype of the normal in all data tiers, as used to classify somatic variants. One of {ref,het,hom,conflict}.">
##INFO=<ID=PNOISE,Number=1,Type=Float,Description="Fraction of panel containing non-reference noise at this site">
##INFO=<ID=PNOISE2,Number=1,Type=Float,Description="Fraction of panel containing more than one non-reference noise obs at this site">
##INFO=<ID=QSS,Number=1,Type=Integer,Description="Quality score for any somatic snv, ie. for the ALT allele to be present at a significantly different frequency in the tumor and normal">
##INFO=<ID=QSS_NT,Number=1,Type=Integer,Description="Quality score reflecting the joint probability of a somatic variant and NT">
##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref read-position in the tumor">
##INFO=<ID=SGT,Number=1,Type=String,Description="Most likely somatic genotype excluding normal noise states">
##INFO=<ID=SNVSB,Number=1,Type=Float,Description="Somatic SNV site strand bias">
##INFO=<ID=SOMATIC,Number=0,Type=Flag,Description="Somatic mutation">
##INFO=<ID=SomaticEVS,Number=1,Type=Float,Description="Somatic Empirical Variant Score (EVS) expressing the phred-scaled probability of the call being a false positive observation.">
##INFO=<ID=TQSS,Number=1,Type=Integer,Description="Data tier used to compute QSS">
##INFO=<ID=TQSS_NT,Number=1,Type=Integer,Description="Data tier used to compute QSS_NT">
```

# ----------

# Plot n_variants and TMB

```{r}
response_rates <- read_csv2(here(metadata_dir, "Trial_response.csv"))
response_rates %>% arrange(Response_rate) %>% pull(Response_rate) %>% round(2)

nrow(combined_maf_df)
n_distinct(combined_maf_df %>%
             select(CF_sample_tumor, Hugo_Symbol, vcf_pos, Chromosome))
n_distinct(combined_maf_df %>%
             select(Patient, Hugo_Symbol, vcf_pos, Chromosome))

combined_maf@variants.per.sample %>%
  left_join(combined_maf@data %>%
              select(Tumor_Sample_Barcode, Source_MAF) %>%
              unique()) %>%
  left_join(dna_metadata_f) %>% view()

n_variants_tmb <- combined_maf@variants.per.sample %>%
  left_join(combined_maf@data %>%
              select(Tumor_Sample_Barcode, Source_MAF) %>%
              unique()) %>%
  left_join(dna_metadata_f) %>%
  left_join(WES_target_sizes) %>%
  mutate(TMB = Variants / WES_target_size,
         Histology_MSI_MSS = ifelse(
    Trial %in% c("NICHE-MSS"), "Colon-MSS",
    ifelse(Trial %in% c("NICHE-MSI"), "Colon-MSI",
           Histology))) %>%
  unique()

table(n_variants_tmb$Trial, n_variants_tmb$Response)

dna_metadata_f$Patient[!dna_metadata_f$Patient %in% n_variants_tmb$Patient]

tmb_response_rate <- n_variants_tmb %>%
  group_by(Trial, Histology, Histology_MSI_MSS) %>%
  summarize(mean_TMB = mean(TMB), median_TMB = median(TMB)) %>%
  ungroup()  %>%
  left_join(response_rates %>% select(-Histology), by = c("Trial")) %>%
  mutate(Yarchoan_lm_rr = ((10.8 * log(median_TMB)) - 0.7) / 100) %>%
  arrange(median_TMB)

tmb_response_rate %>%
  ggplot(aes(x = median_TMB, y = Response_rate, col = Histology_MSI_MSS)) +
  geom_point() +
  expand_limits(x = 0, y = 0) +
  theme_Publication() +
  labs(x = "median_TMB (mutations/Mb)", col = "") +
  scale_color_manual(values = histology_msi_mss_colors)

# Compare with Yarchoan et al. 2017 plot and response linear regression
tmb_response_rate %>%
  ggplot(aes(x = median_TMB, y = Response_rate, col = Histology_MSI_MSS)) +
  geom_point() +
  expand_limits(x = 0, y = 0) +
  theme_Publication() +
  labs(x = "median_TMB (mutations/Mb)", col = "") +
  scale_x_continuous(trans = "log", breaks = c(1, 10, 20, 30, 40, 50)) +
  scale_color_manual(values = histology_msi_mss_colors)

tmb_response_rate %>%
  ggplot(aes(x = Yarchoan_lm_rr, y = Response_rate, col = Histology_MSI_MSS)) +
  geom_point() +
  expand_limits(x = 0, y = 0) +
  theme_Publication() +
  geom_abline(slope = 1, intercept = 0) +
  labs(x = "Yarchoan et al. 2017\nLinear regression RR ~ TMB", col = "")+
  scale_color_manual(values = histology_msi_mss_colors)

n_variants_tmb %>%
  ggplot(aes(x = factor(Histology_MSI_MSS,
                        levels = unique(tmb_response_rate$Histology_MSI_MSS)),
             y = TMB, fill = Histology)) +
  geom_boxplot() + theme_Publication() +
  guides(fill = FALSE) +
  labs(x = "", y = "TMB (mutations/Mb)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))+
  scale_fill_manual(values = c("#B73A3A", "#CD9600", "#23C26F", "#00BFC4",
                               "#009292", "#B66DFF", "#FEB5DA"))

proportion_above_10 <- n_variants_tmb %>%
  group_by(Histology_MSI_MSS) %>%
  summarise(proportion = mean(TMB > 10))

 n_variants_tmb %>%
  mutate(Histology_MSI_MSS = ifelse(
    Histology_MSI_MSS == "cSCC", "CSCC", Histology_MSI_MSS)) %>%
  ggplot(aes(x = factor(Histology_MSI_MSS,
                        levels = unique(tmb_response_rate$Histology_MSI_MSS)),
             y = TMB, fill = Histology_MSI_MSS)) + 
  stat_boxplot(geom = "errorbar", width = 0.1) +
  geom_boxplot() + theme_Publication() +
  guides(fill = FALSE) +
  labs(x = "", y = "TMB (mutations/Mb)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))+
  scale_fill_manual(values = histology_msi_mss_colors) + 
  geom_hline(yintercept = 10, linetype = "dashed", alpha = 0.5) +
  scale_y_continuous(expand = expand_scale(mult = c(0, 0.1)))+
  geom_text(data = proportion_above_10,
            aes(x = Histology_MSI_MSS, y = 168,
                label = paste0(round(proportion * 100), "%")), vjust = -1.5)

n_variants_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB, fill = Histology_MSI_MSS)) +
  geom_boxplot() + theme_Publication() +
  labs(y = "TMB (mutations/Mb)") +
  scale_fill_manual(values = histology_msi_mss_colors)

n_variants_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB)) +
  geom_boxplot(aes(fill = Response)) + theme_Publication() +
  facet_grid(~Histology) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)") +
  stat_compare_means(method = "wilcox.test", label = "p.format")

n_variants_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB)) +
  geom_boxplot(aes(fill = Response)) + theme_Publication() +
  facet_grid(~Histology_MSI_MSS) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)") +
  stat_compare_means(method = "wilcox.test", label = "p.format")

n_variants_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB, fill = Response)) +
  geom_boxplot() + theme_Publication() +
  facet_wrap(~Histology, scales = "free") +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)")

stat.test <- compare_means(formula = TMB ~ Response,
                           method = "wilcox.test",
                           data = n_variants_tmb, paired = FALSE)

n_variants_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication()  +
  stat_pvalue_manual(stat.test, y.position = 150,
                     label = "wilcoxon, p = {p.format}") +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)")

n_variants_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = factor(Trial, levels = c(unique(tmb_response_rate$Trial))),
             y = TMB, fill = Response)) +
  geom_boxplot() + theme_Publication() +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(x = "Trial", y = "TMB (mutations/Mb)") +
  geom_hline(yintercept = 10, linetype = "dashed") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

n_variants_tmb_group <- n_variants_tmb %>%
  mutate(TMB_group = ifelse(TMB > 10, "TMB-H", "TMB-L")) %>%
  filter(!is.na(Response))

table(n_variants_tmb_group$TMB_group)
table(n_variants_tmb_group$Trial, n_variants_tmb_group$TMB_group)

table(n_variants_tmb_group$TMB_group, n_variants_tmb_group$Response)

table(n_variants_tmb_group$TMB_group, n_variants_tmb_group$Response) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total)

tmb_group_response <- table(n_variants_tmb_group$TMB_group,
                            n_variants_tmb_group$Response) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total)

tmb_group_response

tmb_group_response %>%
  ggplot(aes(x = TMB_group, y = Proportion)) + geom_col() +
  theme_Publication() + labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25))

tmb_group_trial_response <- table(n_variants_tmb_group$TMB_group,
                            n_variants_tmb_group$Response,
                            n_variants_tmb_group$Trial) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Trial", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total,
         Trial_TMB_group = sprintf("%s (%s)", Trial, Total))

tmb_group_trial_response$TMB_group <- factor(
  tmb_group_trial_response$TMB_group, levels = c("TMB-L", "TMB-H"))
tmb_group_trial_response$Trial <- factor(
  tmb_group_trial_response$Trial,
  levels = c(unique(tmb_response_rate$Trial)))

tmb_group_trial_response %>%
  ggplot(aes(x = TMB_group, y = Proportion)) + geom_col() +
  theme_Publication() + labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
scale_y_continuous(breaks = seq(0, 1, by = 0.25))  + facet_wrap(~Trial)

tmb_group_trial_response %>%
  ggplot(aes(x = factor(Trial, levels = c(unique(tmb_response_rate$Trial))),
             y = Proportion,fill = TMB_group)) +
  geom_col(position = "dodge2") +
  theme_Publication() +
  labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c("#00BA38", "#619CFF"))

tmb_group_trial_response %>%
  select(Trial, TMB_group, Proportion) %>%
  mutate(Proportion = round(Proportion, 2)) %>%
  pivot_wider(names_from = TMB_group, values_from = Proportion) %>%
  left_join(response_rates) %>%
  view()

tmb_group_trial_response %>%
  ggplot(aes(x = reorder(Trial_TMB_group, Proportion),
             y = Proportion,fill = TMB_group)) +
  geom_col(position = "dodge2") +
  theme_Publication() +
  labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c("#00BA38", "#619CFF")) +
  facet_wrap(~TMB_group, scales = "free_x")

n_variants_tmb <- read_csv2(here(output_dir, "n_variants_tmb.csv"))
n_variants_tmb_group_excl_dmmr <- n_variants_tmb_group %>%
  filter(!Histology_MSI_MSS %in% c("Colon-MSI"))

 table(n_variants_tmb_group_excl_dmmr$TMB_group,
                            n_variants_tmb_group_excl_dmmr$Response) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total)
```

## Filter with DP and VAF

```{r}
# Minor impact
combined_maf_dp20_vaf <- subsetMaf(combined_maf, query = "n_depth >= 20")
combined_maf_dp20_vaf <- subsetMaf(
  combined_maf_dp20_vaf, query = "t_alt_count / t_depth >= 0.05")

combined_maf_dp20_vaf_df <- combined_maf_dp20_vaf@data

combined_maf_dp20_vaf_df %>%
  write_csv2(here(output_dir, "combined_maf_dp20_vaf_df.csv"))

plotmafSummary(maf = combined_maf_dp20_vaf, rmOutlier = TRUE,
               addStat = 'median', dashboard = TRUE, titvRaw = FALSE)

combined_maf_dp20_vaf_df <- combined_maf_dp20_vaf_df %>%
  left_join(dna_metadata_f, by = c("Source_MAF", "Tumor_Sample_Barcode")) %>%
  left_join(WES_target_sizes) %>%
  mutate(VAF = t_alt_count / t_depth)

combined_maf_dp20_vaf_df %>%
  write_csv2(here(output_dir, "combined_maf_dp20_vaf_df.csv"))

combined_maf_dp20_vaf_df <- read_csv2(here(wes_dir, "combined_maf_dp20_vaf_df.csv"))

combined_maf_dp20_vaf_df %>% select(VAF) %>% summary()

n_variants_dp20_vaf_tmb <- combined_maf_dp20_vaf@variants.per.sample %>%
  left_join(combined_maf_dp20_vaf@data %>%
              select(Tumor_Sample_Barcode, Source_MAF) %>%
              unique()) %>%
  left_join(dna_metadata_f) %>%
  left_join(WES_target_sizes) %>%
  mutate(TMB = Variants / WES_target_size,
         Histology_MSI_MSS = ifelse(
    Trial %in% c("NICHE-MSS"), "Colon-MSS",
    ifelse(Trial %in% c("NICHE-MSI"), "Colon-MSI",
           Histology)))

compare_tmb_f <- n_variants_tmb %>%
  select(Patient, Variants, TMB) %>%
  left_join(n_variants_dp20_vaf_tmb %>%
  select(Patient, Variants, TMB) %>%
  setNames(c("Patient", "Variants_dp20_vaf" ,"TMB_dp20_vaf"))
  ) %>%
  left_join(dna_metadata_f %>% select(Patient, Trial)) %>%
  mutate(Delta_TMB = TMB - TMB_dp20_vaf,
         Delta_variants = Variants - Variants_dp20_vaf) %>%
  unique()

# Remove MAT_71 that did not match
compare_tmb_f <- compare_tmb_f %>%
  filter(!(Patient == "MAT_71" & Delta_variants %in% c(-281, 284)))

compare_tmb_f %>%
  ggplot(aes(x = Variants, y = Variants_dp20_vaf, col = Trial)) +
  geom_point() + theme_Publication() +
  geom_abline(slope = 1, intercept = 0) +
  scale_x_continuous(trans = "log2") +
  scale_y_continuous(trans = "log2")

compare_tmb_f %>%
  ggplot(aes(x = TMB, y = TMB_dp20_vaf, col = Trial)) +
  geom_point() + theme_Publication() +
  geom_abline(slope = 1, intercept = 0)+
  scale_x_continuous(trans = "log2") +
  scale_y_continuous(trans = "log2")

compare_tmb_f %>%
  filter(Trial == "IMCISION") %>%
  ggplot(aes(x = TMB, y = TMB_dp20_vaf, col = Trial)) +
  geom_point() + theme_Publication() +
  geom_abline(slope = 1, intercept = 0)+
  scale_x_continuous(trans = "log2") +
  scale_y_continuous(trans = "log2")

compare_tmb_f %>%
  filter(Trial == "NICHE-MSI") %>%
  ggplot(aes(x = TMB, y = TMB_dp20_vaf, col = Trial)) +
  geom_point() + theme_Publication() +
  geom_abline(slope = 1, intercept = 0)+
  scale_x_continuous(trans = "log2") +
  scale_y_continuous(trans = "log2")

compare_tmb_f %>%
  mutate(Delta_TMB = TMB - TMB_dp20_vaf,
         Delta_variants = Variants - Variants_dp20_vaf) %>%
  summary()
```

## Frameshift burden

```{r}
combined_maf_dp20_vaf_df$Variant_Classification %>% unique()

combined_maf_dp20_vaf_df$Patient %>% n_distinct()
frameshift_df20_vaf <- combined_maf_dp20_vaf_df %>%
  filter(Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins"))
 
sample_barcodes <- unique(combined_maf_dp20_vaf_df$Tumor_Sample_Barcode)

frameshift_df20_vaf_tmb <- table(frameshift_df20_vaf$Tumor_Sample_Barcode) %>%
  as.data.frame() %>%
  setNames(c("Tumor_Sample_Barcode", "Frameshift_load")) %>%
  rbind(data.frame(
    Tumor_Sample_Barcode = sample_barcodes[
      !sample_barcodes %in% frameshift_df20_vaf$Tumor_Sample_Barcode],
    Frameshift_load = 0)) %>%
  left_join(combined_maf_dp20_vaf@data %>%
              select(Tumor_Sample_Barcode, Source_MAF) %>%
              unique()) %>%
  left_join(dna_metadata_f) %>%
  left_join(WES_target_sizes) %>%
  mutate(TMB_frameshift = Frameshift_load / WES_target_size,
         Histology_MSI_MSS = ifelse(
    Trial %in% c("NICHE-MSS"), "Colon-MSS",
    ifelse(Trial %in% c("NICHE-MSI"), "Colon-MSI",
           Histology)))

n_variants_dp20_vaf_tmb <- n_variants_dp20_vaf_tmb %>%
  left_join(frameshift_df20_vaf_tmb %>%
              select(Tumor_Sample_Barcode, TMB_frameshift, Frameshift_load))

```

## Plot n_variants and filtered TMB

```{r}
nrow(combined_maf_dp20_vaf_df)
table(n_variants_dp20_vaf_tmb$Trial, n_variants_dp20_vaf_tmb$Response)

dna_metadata_f$Patient[
  !dna_metadata_f$Patient %in% n_variants_dp20_vaf_tmb$Patient]

tmb_response_rate <- n_variants_dp20_vaf_tmb %>%
  group_by(Trial, Histology, Histology_MSI_MSS) %>%
  summarize(mean_TMB = mean(TMB), median_TMB = median(TMB)) %>%
  ungroup()  %>%
  left_join(response_rates %>% select(-Histology), by = c("Trial")) %>%
  mutate(Yarchoan_lm_rr = ((10.8 * log(median_TMB)) - 0.7) / 100) %>%
  arrange(median_TMB)

tmb_response_rate %>%
  ggplot(aes(x = median_TMB, y = Response_rate, col = Histology_MSI_MSS)) +
  geom_point() +
  expand_limits(x = 0, y = 0) +
  theme_Publication() +
  labs(x = "median_TMB (mutations/Mb)", col = "") +
  scale_color_manual(values = histology_msi_mss_colors)

# Compare with Yarchoan et al. 2017 plot and response linear regression
tmb_response_rate %>%
  ggplot(aes(x = median_TMB, y = Response_rate, col = Histology_MSI_MSS)) +
  geom_point() +
  expand_limits(x = 0, y = 0) +
  theme_Publication() +
  labs(x = "median_TMB (mutations/Mb)", col = "") +
  scale_x_continuous(trans = "log", breaks = c(1, 10, 20, 30, 40, 50)) +
  scale_color_manual(values = histology_msi_mss_colors)
ggsave(here(output_dir, "median_TMB_response.png"), height = 5, width = 6)

tmb_response_rate %>%
  ggplot(aes(x = Yarchoan_lm_rr, y = Response_rate, col = Histology_MSI_MSS)) +
  geom_point() +
  expand_limits(x = 0, y = 0) +
  theme_Publication() +
  geom_abline(slope = 1, intercept = 0) +
  labs(x = "Yarchoan et al. 2017\nLinear regression RR ~ TMB", col = "")+
  scale_color_manual(values = histology_msi_mss_colors)
ggsave(here(output_dir, "Yarchoan_TMB_RR.png"), height = 5, width = 6)

n_variants_dp20_vaf_tmb %>%
  ggplot(aes(x = factor(Histology_MSI_MSS,
                        levels = unique(tmb_response_rate$Histology_MSI_MSS)),
             y = TMB, fill = Histology)) +
  geom_boxplot() + theme_Publication() +
  guides(fill = FALSE) +
  labs(x = "", y = "TMB (mutations/Mb)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))+
  scale_fill_manual(values = c("#B73A3A", "#CD9600", "#23C26F", "#00BFC4",
                               "#009292", "#B66DFF", "#FEB5DA"))

n_variants_dp20_vaf_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB, fill = Histology_MSI_MSS)) +
  geom_boxplot() + theme_Publication() +
  labs(y = "TMB (mutations/Mb)") +
  scale_fill_manual(values = histology_msi_mss_colors)
ggsave(here(output_dir, "TMB_histology.png"), height = 5, width = 5)

proportion_above_10 <- n_variants_dp20_vaf_tmb %>%
  group_by(Histology_MSI_MSS) %>%
  summarise(proportion = mean(TMB > 10))

n_variants_dp20_vaf_tmb %>%
  mutate(Histology_MSI_MSS = ifelse(
    Histology_MSI_MSS == "cSCC", "CSCC", Histology_MSI_MSS)) %>%
  ggplot(aes(x = factor(Histology_MSI_MSS,
                        levels = unique(tmb_response_rate$Histology_MSI_MSS)),
             y = TMB, fill = Histology_MSI_MSS)) + 
  stat_boxplot(geom = "errorbar", width = 0.1) +
  geom_boxplot() + theme_Publication() +
  guides(fill = FALSE) +
  labs(x = "", y = "TMB (mutations/Mb)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))+
  scale_fill_manual(values = histology_msi_mss_colors) + 
  geom_hline(yintercept = 10, linetype = "dashed", alpha = 0.5) +
  scale_y_continuous(expand = expand_scale(mult = c(0, 0.1)))+
  geom_text(
    data = proportion_above_10,
    aes(x = Histology_MSI_MSS, y = 165,
        label = paste0(round(proportion * 100), "%")), vjust = -1.5)
 
n_variants_dp20_vaf_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB)) +
  geom_boxplot(aes(fill = Response)) + theme_Publication() +
  facet_grid(~Histology) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)") +
  stat_compare_means(method = "wilcox.test", label = "p.format")
ggsave(here(output_dir, "TMB_response_histology.png"), height = 5, width = 8)

n_variants_dp20_vaf_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB)) +
  geom_boxplot(aes(fill = Response)) + theme_Publication() +
  facet_grid(~Histology_MSI_MSS) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)") +
  stat_compare_means(method = "wilcox.test", label = "p.format")
ggsave(here(output_dir, "TMB_response_histology_MSI_MSS.png"),
       height = 5, width = 9)

n_variants_dp20_vaf_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB, fill = Response)) +
  geom_boxplot() + theme_Publication() +
  facet_wrap(~Histology, scales = "free") +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)")

stat.test <- compare_means(formula = TMB ~ Response,
                           method = "wilcox.test",
                           data = n_variants_dp20_vaf_tmb, paired = FALSE)
n_variants_dp20_vaf_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = Response, y = TMB)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication()  +
  stat_pvalue_manual(stat.test, y.position = 150,
                     label = "wilcoxon, p = {p.format}") +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(y = "TMB (mutations/Mb)")
ggsave(here(output_dir, "TMB_response.png"), height = 5, width = 4)

n_variants_dp20_vaf_tmb %>%
  filter(!is.na(Response)) %>%
  ggplot(aes(x = factor(Trial, levels = c(unique(tmb_response_rate$Trial))),
             y = TMB, fill = Response)) +
  geom_boxplot() + theme_Publication() +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  expand_limits(y = 0) +
  labs(x = "Trial", y = "TMB (mutations/Mb)") +
  geom_hline(yintercept = 10, linetype = "dashed") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

n_variants_dp20_vaf_tmb_group <- n_variants_dp20_vaf_tmb %>%
  mutate(TMB_group = ifelse(TMB > 10, "TMB-H", "TMB-L")) %>%
  filter(!is.na(Response))

table(n_variants_dp20_vaf_tmb_group$TMB_group)
table(n_variants_dp20_vaf_tmb_group$Trial,
      n_variants_dp20_vaf_tmb_group$TMB_group)

table(n_variants_dp20_vaf_tmb_group$TMB_group,
      n_variants_dp20_vaf_tmb_group$Response)

table(n_variants_dp20_vaf_tmb_group$TMB_group,
      n_variants_dp20_vaf_tmb_group$Response) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total)

tmb_group_response <- table(n_variants_dp20_vaf_tmb_group$TMB_group,
                            n_variants_dp20_vaf_tmb_group$Response) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total)

tmb_group_response

tmb_group_response %>%
  ggplot(aes(x = TMB_group, y = Proportion)) + geom_col() +
  theme_Publication() + labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25))
ggsave(here(output_dir, "TMB_group_response.png"), height = 5, width = 6)


tmb_group_trial_response <- table(n_variants_dp20_vaf_tmb_group$TMB_group,
                            n_variants_dp20_vaf_tmb_group$Response,
                            n_variants_dp20_vaf_tmb_group$Trial) %>%
  as.data.frame() %>% setNames(c("TMB_group", "Response", "Trial", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Proportion = R / Total,
         Trial_TMB_group = sprintf("%s (%s)", Trial, Total))

tmb_group_trial_response$TMB_group <- factor(
  tmb_group_trial_response$TMB_group, levels = c("TMB-L", "TMB-H"))
tmb_group_trial_response$Trial <- factor(
  tmb_group_trial_response$Trial,
  levels = c(unique(tmb_response_rate$Trial)))

tmb_group_trial_response %>%
  ggplot(aes(x = TMB_group, y = Proportion)) + geom_col() +
  theme_Publication() + labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
scale_y_continuous(breaks = seq(0, 1, by = 0.25))  + facet_wrap(~Trial)

tmb_group_trial_response %>%
  ggplot(aes(x = factor(Trial, levels = c(unique(tmb_response_rate$Trial))),
             y = Proportion,fill = TMB_group)) +
  geom_col(position = "dodge2") +
  theme_Publication() +
  labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c("#00BA38", "#619CFF"))

tmb_group_trial_response %>%
  select(Trial, TMB_group, Proportion) %>%
  mutate(Proportion = round(Proportion, 2)) %>%
  pivot_wider(names_from = TMB_group, values_from = Proportion) %>%
  left_join(response_rates) %>%
  view()

tmb_group_trial_response %>%
  ggplot(aes(x = reorder(Trial_TMB_group, Proportion),
             y = Proportion,fill = TMB_group)) +
  geom_col(position = "dodge2") +
  theme_Publication() +
  labs(x = "TMB group (cutoff at 10 Mut/Mb)",
                             y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c("#00BA38", "#619CFF")) +
  facet_wrap(~TMB_group, scales = "free_x")
```


```{r}
# Frameshift
n_variants_dp20_vaf_tmb %>%
  ggplot(aes(x = TMB, y= TMB_frameshift, col = Histology_MSI_MSS)) +
  theme_Publication() + geom_point() +
  scale_color_manual(values = histology_msi_mss_colors)

n_variants_dp20_vaf_tmb %>%
  ggplot(aes(x = TMB, y= TMB_frameshift, col = Trial)) +
  theme_Publication() + geom_point() +
  scale_color_manual(values = trial_colors)

n_variants_dp20_vaf_tmb %>%
  ggplot(aes(x = Histology_MSI_MSS, y= Frameshift_load)) + geom_boxplot() +
  theme_Publication() + geom_point()

n_variants_dp20_vaf_tmb %>%
  ggplot(aes(x = Histology_MSI_MSS, y= TMB_frameshift)) + geom_boxplot() +
  theme_Publication() + geom_point()

n_variants_dp20_vaf_tmb %>%
  filter(!Trial %in% c("NICHE-MSI")) %>%
  ggplot(aes(x = Response, y = Frameshift_load)) + geom_boxplot() +
  theme_Publication() + geom_point()
```


```{r}
tmb_response_rate %>%
  write_csv2(here(output_dir, "tmb_response_rate.csv"))

n_variants_dp20_vaf_tmb %>%
  write_csv2(here(output_dir, "n_variants_dp20_vaf_tmb.csv"))

tmb_group_trial_response %>%
  write_csv2(here(output_dir, "tmb_group_trial_response.csv"))

```

# Add TMB MAF

```{r}
# Add TMB as a clinical feature
combined_maf@clinical.data <- combined_maf@clinical.data %>%
  left_join(n_variants_dp20_vaf_tmb %>% select(Tumor_Sample_Barcode, TMB))

# Replace combined_maf with dp20 vaf filtering
combined_maf_dp20_vaf@clinical.data <- combined_maf_dp20_vaf@clinical.data %>%
  left_join(n_variants_dp20_vaf_tmb %>% select(Tumor_Sample_Barcode, TMB))
```

# ----------

# Gene sets and mutational pathways

```{r}
dmmr_ptns <- c("MLH1", "MSH2", "MSH6", "PMS2")

# Genetic Immune Evasion GIE Immune escape mutation sets
# signatures from  Martínez-Jiménez 2023 https://doi.org/10.1038/s41588-023-01367-1
immune_escape_genes_df <- read.csv2(
  here("data", "external", "paper_supplementaries",
       "Martínez-Jiménez et al. - 2023 Supp Table 1.csv"))
immune_escape_genes_df <- immune_escape_genes_df %>%
  mutate(Pathway.general = ifelse(
    Pathway.general == "IFN-? pathway", "IFN-y pathway", Pathway.general))%>%
  mutate(Pathway.general = ifelse(
    Pathway.general == "Antigen Presentation ", "Antigen Presentation",
    Pathway.general))

immune_escape_genes_df %>% select(Gene, Pathway.specific, Pathway.general)

# Gene lists from PMID: 34873480
ddr_ptns <- c("MRE11", "RAD50", "MLH1", "MSH2", "MSH6", "PMS2", "POLD1",
              "ATM", "ATR", "BARD1", "BLM", "BRCA1", "BRCA2", "BRIP1", "CHEK1",
              "CHEK2", "FANCA", "FANCC", "FANCD2", "FANCG", "MUTYH", "PALB2",
              "POLE", "TP53", "PRKDC", "CUL3", "ERCC1", "FANCE", "FANCF",
              "FANCL", "RAD51", "WEE1")

crc_risk_ptns <- c("APC", "MLH1", "MSH2", "MSH6", "PMS2", "MUTYH", "PTEN",
                   "TP53", "STK11", "SMAD1", "SMAD2")

immune_escape_genes <- immune_escape_genes_df$Gene

# Expand with other drivers
crc_risk_ptns <- c(crc_risk_ptns,
                   c("KRAS", "NRAS", "HRAS", "BRAF", "EGFR", "PIK3CA",
                     "EPCAM"))

# Grasso 2018 TCGA MSS genes
grasso_2018_mss <- c("RNF43", "SMAD4", "PIK3CA", "PTEN", "CTNNB1", "TGFBR2",
                     "ZFP36L2", "B2M", "ACVR2A", "FBXW7", "PCBP1", "SOX9",
                     "SMAD2", "ARID1A", "TCF7L2", "AMER1", "TGIF1", "ELF3",
                     "TP53", "APC", "KRAS", "NRAS", "BRAF")

grasso_mss_wnt_signaling <- c("RNF43", "FBXW7", "SOX9", "PTEN", "SMAD2",
                          "SMAD4", "ELF3", "AMER1") # Leaving out "APC", "TP53"

grasso_mss_tgfb_signaling <- c("ACVR2A","SMAD2", "SMAD4", "PCBP1")

# Check oncokb cancergene list
oncoKB_cancerGeneList <- read.delim(
  here("data", "external", "oncoKB_cancerGeneList_20240406.tsv"))

list.files(system.file('extdata', package = 'maftools'))

# Bailey et al., https://doi.org/10.1016/j.cell.2018.02.060]
maftools_BP_SMG_pathways <- data.table::fread(
  file = system.file("extdata", "BP_SMGs.txt.gz", 
                     package = "maftools"), skip = "Gene")

# All genes on SMGBP are assigned to a unique pathway
maftools_BP_SMG_pathways %>% select(Gene, Pathway) %>% unique() %>% pull(Gene) %>% table() %>% as.data.frame() %>% filter(Freq > 1)


maftools_BP_SMG_pathways %>% factor_summary()
maftools_BP_SMG_pathways %>% filter(Cancer_type == "PANCAN") %>% factor_summary()
maftools_BP_SMG_pathways$Gene[!maftools_BP_SMG_pathways$Gene %in% oncoKB_cancerGeneList$Hugo.Symbol]

maftools_BP_SMG_pathways %>%
  write_csv2(here("data", "external", "maftools_BP_SMG_pathways.csv"))

# Sanchez-Vega et al., https://doi.org/10.1016/j.cell.2018.03.035]")
maftools_sigpw_pathways <- data.table::fread(file = system.file("extdata", "oncogenic_sig_patwhays.tsv", 
                package = "maftools"), skip = "Gene")
maftools_sigpw_pathways %>% select(Gene, Pathway) %>% unique() %>% pull(Gene) %>% table() %>% as.data.frame() %>% filter(Freq > 1)
maftools_sigpw_pathways %>% filter(Gene == "SCRIB")

maftools_sigpw_pathways %>% write_csv2(here(output_dir, "maftools_sigpw_pathways"))
# pathdb <- system.file("extdata", "oncogenic_sig_patwhays.tsv", package = "maftools")

maftools_sigpw_pathways %>% factor_summary()

maftools_sigpw_oncogenes <- maftools_sigpw_pathways %>% filter(OG_TSG == "OG")
maftools_sigpw_oncogenes %>% factor_summary()
maftools_sigpw_tumor_suppressors <- maftools_sigpw_pathways %>% filter(OG_TSG == "TSG")
maftools_sigpw_tumor_suppressors %>% factor_summary()

table(oncoKB_cancerGeneList$Is.Oncogene, oncoKB_cancerGeneList$Is.Tumor.Suppressor.Gene, useNA = "ifany")
oncoKB_cancerGeneList %>% filter(Is.Oncogene == "Yes") %>% filter(Is.Tumor.Suppressor.Gene == "Yes") %>% pull(Hugo.Symbol)

oncoKB_oncogenes <- oncoKB_cancerGeneList %>% filter(Is.Oncogene == "Yes") %>% filter(Is.Tumor.Suppressor.Gene != "Yes")
oncoKB_tumor_suppressors <- oncoKB_cancerGeneList %>% filter(Is.Tumor.Suppressor.Gene == "Yes") %>% filter(Is.Oncogene != "Yes")

table(maftools_sigpw_oncogenes$Gene %in% oncoKB_oncogenes$Hugo.Symbol)
table(oncoKB_oncogenes$Hugo.Symbol %in% maftools_sigpw_oncogenes$Gene) 

table(maftools_sigpw_tumor_suppressors$Gene %in% oncoKB_tumor_suppressors$Hugo.Symbol) 
table(oncoKB_tumor_suppressors$Hugo.Symbol %in% maftools_sigpw_tumor_suppressors$Gene) 

```

```{r}
genes_to_assess_shortlist <- unique(
  c(ddr_ptns, crc_risk_ptns, immune_escape_genes,
                     "CD58", "CD274", "SETDB1", "CDKN2A", "TRAF2", "CCND1")) %>%
  unique()

genes_to_assess_shortlist <- genes_to_assess_shortlist[
  genes_to_assess_shortlist %in% combined_maf_dp20_vaf_df$Hugo_Symbol]
length(genes_to_assess_shortlist)

genes_to_assess_shortlist[
  !genes_to_assess_shortlist %in% oncoKB_cancerGeneList$Hugo.Symbol]

# Expand genes to assess
genes_to_assess <- unique(
  c(genes_to_assess_shortlist, maftools_sigpw_pathways$Gene,
    maftools_BP_SMG_pathways$Gene, oncoKB_cancerGeneList$Hugo.Symbol))
length(genes_to_assess)

genes_to_assess[!genes_to_assess %in% oncoKB_cancerGeneList$Hugo.Symbol]

genes_to_asses_cnv <- unique(c(
  "CD274", "CD58", "CDKN2A", "SETDB1", "TRAF2", "CCND1",
  immune_escape_genes_df %>%
    filter(Pathway.general %in% c("Antigen Presentation", "IFN-y pathway")) %>%
    pull(Gene)))

table(combined_maf_dp20_vaf_df$Variant_Classification,
      combined_maf_dp20_vaf_df$Variant_Type)

variant_eval_columns <- c("Symbol_Variant", "SYMBOL",
                          "CLIN_SIG", "SIFT", "PolyPhen",
                          "Variant_Classification",
                          "Variant_Type", "VARIANT_CLASS",
                          "IMPACT", "AF",  "gnomAD_AF", 
                          "n_depth", "VAF") # "EUR_AF", "gnomAD_NFE_AF",

table(oncoKB_cancerGeneList$Hugo.Symbol %in% genes_to_assess)
table(maftools_BP_SMG_pathways$Gene %in% genes_to_assess)
table(maftools_sigpw_pathways$Gene %in% genes_to_assess)

maftools_BP_SMG_pathways$Gene[!maftools_BP_SMG_pathways$Gene %in% genes_to_assess]
maftools_sigpw_pathways$Gene[!maftools_sigpw_pathways$Gene %in% genes_to_assess]

```

## Genetic Immune Evasion Martínez-Jiménez 2023

```{r, fig.width = 14, fig.height = 14}
oncoplot(combined_maf_dp20_vaf,
         pathways = immune_escape_genes_df %>%
           select(Gene, Pathway.general),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)
```

## Oncogenic Pathways TCGA pathways sigpw Sanchez Vega

```{r, fig.width = 12, fig.height = 10}
# PlotOncogenicPathways uses maftools_sigpw_pathways

# Available pathways 
# "Cell_Cycle" "Hippo"      "MYC"        "NOTCH"      "NRF2"       "PI3K"       "RTK-RAS"    "TGF-Beta"   "TP53"       "WNT"
# Color for tumor suppressor genes. Default red
# Color for oncogenes. Default royalblue
tumor_supressor_color <- "#007958"
oncogene_color <- "#FF4500"

PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "Cell_Cycle", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "Hippo", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "MYC", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "NOTCH", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "NRF2", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "PI3K", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "RTK-RAS", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "TGF-Beta", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "TP53", tsgCol = tumor_supressor_color, ogCol = oncogene_color)
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "WNT", tsgCol = tumor_supressor_color, ogCol = oncogene_color)

?OncogenicPathways
?PlotOncogenicPathways

```

## Oncoplots SMGBP Bailey 2018

```{r, fig.width = 14, fig.height = 14}
# <https://doi.org/10.1016/j.cell.2018.02.060>

# Genome_integrity, MAPK_signaling, Cell_cycle, Chromatin_other,
# Chromatin_SWI/SNF_complex, Transcription_factor, RNA_abundance
# Chromatin_histone_modifiers, RTK_signaling
# Wnt/B-catening_signaling PI3K_signaling
# NOTCH_signaling Protein_homeostatis/ubiquitination
# Histone_modification  TGFB_signaling
# TOR_signaling Metabolism Immune_signaling
# Splicing Epigenetics_DNA_modifiers
#  NFKB_signaling Apoptosis

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         selectedPathways = c("Wnt/B-catenin_signaling", "TGFB_signaling"),
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "sigpw", 
         selectedPathways = c("Cell_cycle", "MAPK_signaling", "RTK_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list) # removeNonMutated = FALSE

oncoplot(combined_maf_dp20_vaf, pathways = "sigpw", 
         selectedPathways = c("Genome_integrity",
                              "Protein homeostasis/ubiquitination",
                              "TGFB_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list)

oncoplot(combined_maf_dp20_vaf, pathways = "sigpw", 
         selectedPathways = c("Epigenetics_DNA_modifiers", "Chromatin_other",
                              "Chromatin_histone_modifiers"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB")

oncoplot(combined_maf_dp20_vaf, pathways = "sigpw", 
         selectedPathways = c("Epigenetics_DNA_modifiers", "NOTCH_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB")

```

## Oncoplots, compare similar pathways sigpw vs SMGBP

```{r}
# Choose pathway versions
maftools_BP_SMG_pathways$Pathway %>% unique()
maftools_sigpw_pathways$Pathway %>% unique()
```

```{r, fig.width = 14, fig.height = 14}
# Careful! Oncoplot pathways is using and returning smgbp pathways, even if you indicate pathways = "sigpw"
# The name of pathways is matching smgbp instead of sigpw... this cannot be used reliably and caused a lot of confusion before
# loaded maftools_sigpw_pathways separately

# This does not work, although the pathway list should explicitly be "sigpw", it's using the name from smgbp, "RTK_signaling" (!)
# oncoplot(combined_maf_dp20_vaf,
#          pathways = "sigpw",
#          selectedPathways = c("RTK_signaling"),
#          clinicalFeatures = c("Trial", "Response"),
#          topBarData = "TMB",
#          annotationColor = annotation_color_list,
#          gene_mar = 8, fontSize = 0.6,
#          sortByAnnotation = TRUE)

# Cell cycle
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "Cell_Cycle",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)

oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "Cell_Cycle") %>%
           select(Gene, Pathway),
         selectedPathways = c("Cell_Cycle"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         selectedPathways = c("Cell_cycle"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoKB_cancerGeneList %>% filter(Hugo.Symbol %in% c("BTG2", "CDK12"))

# NOTCH
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "NOTCH",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)
oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "NOTCH") %>%
           select(Gene, Pathway),
         selectedPathways = c("NOTCH"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# NOTCH alone is not working, only one gene
oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         selectedPathways = c("Cell_cycle", "NOTCH_signaling"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# PI3K
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "PI3K",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)
oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "PI3K") %>%
           select(Gene, Pathway),
         selectedPathways = c("PI3K"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         selectedPathways = c("PI3K_signaling"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoKB_cancerGeneList %>% filter(Hugo.Symbol %in% c("INPPL1", "MAPK1", "DEPDC5"))

# TGFb
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "TGF-Beta",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)
oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "TGF-Beta") %>%
           select(Gene, Pathway),
         selectedPathways = c("TGF-Beta"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         selectedPathways = c("TGFB_signaling"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoKB_cancerGeneList %>% filter(Hugo.Symbol %in% c("DACH1", "ACVR1"))

# RTK RAS
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "RTK-RAS",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)

oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "RTK-RAS") %>%
           select(Gene, Pathway),
         selectedPathways = c("RTK-RAS"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         selectedPathways = c("RTK_signaling"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoKB_cancerGeneList %>% filter(Hugo.Symbol %in% c("EPHA2", "PLCG1", "JAK1"))


# TP53
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "TP53",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)

oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "TP53") %>%
           select(Gene, Pathway),
         selectedPathways = c("TP53"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         selectedPathways = c("Genome_integrity"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# Wnt
PlotOncogenicPathways(combined_maf_dp20_vaf, pathways = "WNT",
                      tsgCol = tumor_supressor_color, ogCol = oncogene_color)

oncoplot(combined_maf_dp20_vaf, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "WNT") %>%
           select(Gene, Pathway),
         selectedPathways = c("WNT"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_dp20_vaf, pathways = "smgbp",
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         selectedPathways = c("Wnt/B-catenin_signaling"),
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

maftools_sigpw_pathways %>% filter(Pathway == "WNT") %>% pull(Gene) %>% unique()
maftools_BP_SMG_pathways %>% filter(Pathway == "Wnt/B-catenin signaling") %>% pull(Gene) %>% unique()

oncoKB_cancerGeneList %>% filter(Hugo.Symbol %in% c("PTCH1", "CTNND1", "LZTR1", "NDP", "CHD4", "CHD8"))

```

# Oncoplot TOP100 in pathways

```{r}
top100_genes <- read_csv2(
  here("data", "processed", "PanCancer_pathway_mutations", "all_pathways",
       "Mutated_genes_ALL_patways_TOP_100_ensembl_gene_list.csv"))

top100_genes <- read_delim(
  here("data", "processed", "PanCancer_pathway_mutations",
       "Oncoplot_top_mutated_genes_oncoKB",
       "OncoPlot_pathway_mutated_genes_oncoKB.txt"), col_names =  FALSE) %>%
  setNames(c("Gene_description", "Description"))

top100_genes <- read_csv2(
  here("data", "processed", "PanCancer_pathway_mutations",
       "Oncoplot_top_mutated_genes_oncoKB",
       "OncoPlot_pathway_mutated_genes_oncoKB.csv")) 

top100_genes <- top100_genes %>% mutate()

top100_genes$Gene %in% combined_maf_dp20_vaf_df$Hugo_Symbol

oncoplot(combined_maf_dp20_vaf,
         genes = top100_genes$Gene,
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         sortByAnnotation = TRUE,
         removeNonMutated = FALSE
         )

# Dev off only works from the console, and not from the rmd chunk. Copy and run
# do not run from chunks and knit
if (FALSE) {
pdf(file = here(output_dir, "_oncoplot_top100_genes_pathway_response.pdf"),
    width = 14, height = 18)
oncoplot(combined_maf_dp20_vaf, genes = top100_genes$Gene,
         gene_mar = 8, fontSize = 0.75,
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE,
         removeNonMutated = FALSE
         )
dev.off()

}
```

# ----------

# Check Sift x Polyphen predictions and clinsig annotations
## Enrich for pathogenic functional mutations
```{r}
colnames(combined_maf_dp20_vaf_df)
combined_maf_dp20_vaf_df$CLIN_SIG %>% unique()

polyphen_sift_comparison <- combined_maf_dp20_vaf_df %>%
  select(CLIN_SIG, IMPACT, SIFT, PolyPhen) %>%
  mutate(SIFT_score = parse_number(SIFT),
         PolyPhen_score = parse_number(PolyPhen))

polyphen_sift_comparison %>%
  ggplot(aes(x = SIFT_score, y = PolyPhen_score)) +
  geom_pointdensity() +
  theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

polyphen_sift_comparison %>%
  ggplot(aes(x = SIFT_score, y = PolyPhen_score)) +
  geom_pointdensity() + facet_wrap(~IMPACT) +
  theme_Publication() + 
  guides(colour = guide_colourbar(barwidth = 10))

pathogenic_categories_vep <- unique(polyphen_sift_comparison$CLIN_SIG)
pathogenic_categories_vep <- pathogenic_categories_vep[str_detect(pathogenic_categories_vep, "pathogenic")]
pathogenic_categories_vep <- pathogenic_categories_vep[!pathogenic_categories_vep %in% c(NA, "benign&likely_benign&benign/likely_benign", "likely_benign", "likely_benign&uncertain_significance&conflicting_interpretations_of_pathogenicity", "benign&likely_benign", "benign&likely_benign&uncertain_significance&benign/likely_benign", "benign&likely_benign&benign/likely_benign&conflicting_interpretations_of_pathogenicity", "benign&other", "benign", "benign/likely_benign&conflicting_interpretations_of_pathogenicity&benign&likely_benign", "benign&not_provided", "uncertain_significance&conflicting_interpretations_of_pathogenicity&likely_benign", "likely_benign&uncertain_significance", "likely_benign&conflicting_interpretations_of_pathogenicity",
                              "benign&benign/likely_benign&likely_benign", "benign&conflicting_interpretations_of_pathogenicity", "uncertain_significance&benign&conflicting_interpretations_of_pathogenicity&likely_benign",
                              "benign,conflicting_interpretations_of_pathogenicity",
                              "benign,conflicting_interpretations_of_pathogenicity,likely_benign",
                              "benign/likely_benign,conflicting_interpretations_of_pathogenicity",
                              "conflicting_interpretations_of_pathogenicity,likely_benign,benign/likely_benign",
                              "benign,uncertain_significance,likely_benign,benign/likely_benign,conflicting_interpretations_of_pathogenicity", "likely_benign,conflicting_interpretations_of_pathogenicity","benign,uncertain_significance,likely_benign,conflicting_interpretations_of_pathogenicity", "uncertain_significance,conflicting_interpretations_of_pathogenicity,benign/likely_benign",
                              "uncertain_significance,conflicting_interpretations_of_pathogenicity,likely_benign",
                              "uncertain_significance,likely_benign,conflicting_interpretations_of_pathogenicity",
                              "likely_benign,uncertain_significance,conflicting_interpretations_of_pathogenicity",
                              "uncertain_significance,benign,conflicting_interpretations_of_pathogenicity,likely_benign",
                              "benign,likely_benign,conflicting_interpretations_of_pathogenicity",
                              "uncertain_significance,benign,conflicting_interpretations_of_pathogenicity,likely_benign,benign/likely_benign"
                              
                              )]

pathogenic_categories_vep_f <- pathogenic_categories_vep[!pathogenic_categories_vep %in% c("uncertain_significance&benign&conflicting_interpretations_of_pathogenicity", "uncertain_significance&benign&conflicting_interpretations_of_pathogenicity", "conflicting_interpretations_of_pathogenicity", "benign&uncertain_significance&likely_benign&conflicting_interpretations_of_pathogenicity", "conflicting_interpretations_of_pathogenicity&likely_benign", "benign&conflicting_interpretations_of_pathogenicity&likely_benign", "uncertain_significance&likely_benign&conflicting_interpretations_of_pathogenicity", "conflicting_interpretations_of_pathogenicity&likely_benign&benign/likely_benign", "uncertain_significance&benign&conflicting_interpretations_of_pathogenicity&likely_benign&benign/likely_benign","benign&conflicting_interpretations_of_pathogenicity&likely_benign&benign/likely_benign", "likely_benign&benign/likely_benign&conflicting_interpretations_of_pathogenicity", "benign&likely_benign&conflicting_interpretations_of_pathogenicity", "benign/likely_benign&conflicting_interpretations_of_pathogenicity&uncertain_significance&benign&likely_benign",
"conflicting_interpretations_of_pathogenicity&uncertain_significance", "conflicting_interpretations_of_pathogenicity&benign&likely_benign", "benign&uncertain_significance&benign/likely_benign&conflicting_interpretations_of_pathogenicity", "benign/likely_benign&likely_benign&conflicting_interpretations_of_pathogenicity", "likely_benign&uncertain_significance&benign/likely_benign&conflicting_interpretations_of_pathogenicity", "uncertain_significance&conflicting_interpretations_of_pathogenicity&likely_benign&not_provided", "uncertain_significance&benign/likely_benign&conflicting_interpretations_of_pathogenicity", "conflicting_interpretations_of_pathogenicity&benign/likely_benign", "uncertain_significance&conflicting_interpretations_of_pathogenicity&likely_benign&benign/likely_benign", "conflicting_interpretations_of_pathogenicity&uncertain_significance&likely_benign", "benign/likely_benign&conflicting_interpretations_of_pathogenicity&likely_benign", "benign&uncertain_significance&conflicting_interpretations_of_pathogenicity", "not_provided&benign/likely_benign&conflicting_interpretations_of_pathogenicity&benign",
"benign,pathogenic,likely_benign","benign/likely_benign,likely_pathogenic,benign",
"uncertain_significance,benign,conflicting_interpretations_of_pathogenicity",
"uncertain_significance,benign,association,conflicting_interpretations_of_pathogenicity,likely_benign",
"uncertain_significance,benign,likely_benign,pathogenic", "not_provided,benign,likely_benign,pathogenic", "conflicting_interpretations_of_pathogenicity,likely_benign", "benign,uncertain_significance,conflicting_interpretations_of_pathogenicity", "likely_benign,benign/likely_benign,conflicting_interpretations_of_pathogenicity", "benign,conflicting_interpretations_of_pathogenicity,likely_benign,benign/likely_benign", "likely_benign,uncertain_significance,benign/likely_benign,conflicting_interpretations_of_pathogenicity",  "benign,likely_benign,benign/likely_benign,conflicting_interpretations_of_pathogenicity",
"benign/likely_benign,likely_benign,conflicting_interpretations_of_pathogenicity",
"pathogenic,uncertain_significance,benign,likely_benign", "benign,benign/likely_benign,pathogenic,conflicting_interpretations_of_pathogenicity", "uncertain_significance,conflicting_interpretations_of_pathogenicity,likely_benign,not_provided", "uncertain_significance,benign,likely_pathogenic,likely_benign", "uncertain_significance,benign,likely_pathogenic,likely_benign", "benign/likely_benign,conflicting_interpretations_of_pathogenicity,uncertain_significance,benign,likely_benign", "benign,benign/likely_benign,conflicting_interpretations_of_pathogenicity", "conflicting_interpretations_of_pathogenicity,uncertain_significance,benign")]

pathogenic_categories_vep_f[str_detect(pathogenic_categories_vep_f, "benign")]
pathogenic_categories_vep_f[str_detect(pathogenic_categories_vep_f, "uncertain")]
pathogenic_categories_vep_f[!str_detect(pathogenic_categories_vep_f, "uncertain|benign")]
pathogenic_categories_vep_f[!str_detect(pathogenic_categories_vep_f, "uncertain|benign|pathogenic")]

polyphen_sift_comparison %>%
  filter(CLIN_SIG %in% pathogenic_categories_vep) %>%
  ggplot(aes(x = SIFT_score, y = PolyPhen_score)) +
  theme_Publication() +
  geom_pointdensity() +
  guides(colour = guide_colourbar(barwidth = 10)) +
  labs(title = "pathogenic / conflicting / uncertain variants")
 
polyphen_sift_comparison_pathogenic_f <- polyphen_sift_comparison %>%
  filter(CLIN_SIG %in% pathogenic_categories_vep_f)

polyphen_sift_comparison_npathogenic_f <- polyphen_sift_comparison %>%
  filter(!CLIN_SIG %in% pathogenic_categories_vep_f)

polyphen_sift_comparison_pathogenic_f %>%
  ggplot(aes(x = SIFT_score, y = PolyPhen_score)) +
  theme_Publication() +
  geom_pointdensity() +
  guides(colour = guide_colourbar(barwidth = 10))+
  labs(title = "Pathogenic variants (stricter categories)")
 
polyphen_sift_comparison %>% filter(SIFT_score < 0.05) %>%
  mutate(SIFT = str_replace(SIFT, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate(PolyPhen = str_replace(PolyPhen, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate_if(is.character, as.factor) %>% summary()

polyphen_sift_comparison %>% filter(SIFT_score >= 0.05) %>%
  mutate(SIFT = str_replace(SIFT, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate(PolyPhen = str_replace(PolyPhen, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate_if(is.character, as.factor) %>% summary()

polyphen_sift_comparison %>% filter(PolyPhen_score > 0.446) %>%
  mutate(SIFT = str_replace(SIFT, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate(PolyPhen = str_replace(PolyPhen, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate_if(is.character, as.factor) %>% summary()

polyphen_sift_comparison %>% filter(PolyPhen_score <= 0.446) %>%
  mutate(SIFT = str_replace(SIFT, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate(PolyPhen = str_replace(PolyPhen, "\\s*\\([^\\)]+\\)", "" )) %>%
  mutate_if(is.character, as.factor) %>% summary()

nrow(polyphen_sift_comparison)
polyphen_sift_comparison %>% filter(SIFT_score < 0.05) %>% filter(PolyPhen_score <= 0.446) %>% nrow()
polyphen_sift_comparison %>% filter(SIFT_score < 0.05) %>% filter(PolyPhen_score > 0.446) %>% nrow()
polyphen_sift_comparison %>% filter(SIFT_score >= 0.05) %>% filter(PolyPhen_score > 0.446) %>% nrow()

# Agreement
polyphen_sift_comparison %>% filter(SIFT_score >= 0.05) %>% filter(PolyPhen_score <= 0.446) %>% nrow()

table(polyphen_sift_comparison$SIFT_score < 0.05,
      polyphen_sift_comparison$PolyPhen_score > 0.446) %>%
    `rownames<-`(c("SIFT_tolerated", "SIFT_deleterious")) %>%
    `colnames<-`(c("PolyPhen_benign", "PolyPhen_damaging"))

table(polyphen_sift_comparison$SIFT_score < 0.05,
      polyphen_sift_comparison$PolyPhen_score > 0.446) %>%
    `rownames<-`(c("SIFT_tolerated", "SIFT_deleterious")) %>%
    `colnames<-`(c("PolyPhen_benign", "PolyPhen_damaging")) %>%
  prop.table() %>%
  round(2)

table(polyphen_sift_comparison_pathogenic_f$SIFT_score < 0.05,
      polyphen_sift_comparison_pathogenic_f$PolyPhen_score > 0.446) %>%
    `rownames<-`(c("SIFT_tolerated", "SIFT_deleterious")) %>%
    `colnames<-`(c("PolyPhen_benign", "PolyPhen_damaging"))

table(polyphen_sift_comparison_pathogenic_f$SIFT_score < 0.05,
      polyphen_sift_comparison_pathogenic_f$PolyPhen_score > 0.446) %>%
    `rownames<-`(c("SIFT_tolerated", "SIFT_deleterious")) %>%
    `colnames<-`(c("PolyPhen_benign", "PolyPhen_damaging")) %>%
  prop.table() %>%
  round(2)

table(polyphen_sift_comparison_npathogenic_f$SIFT_score < 0.05,
      polyphen_sift_comparison_npathogenic_f$PolyPhen_score > 0.446) %>%
    `rownames<-`(c("SIFT_tolerated", "SIFT_deleterious")) %>%
    `colnames<-`(c("PolyPhen_benign", "PolyPhen_damaging")) %>%
  prop.table() %>%
  round(2)
```

# ----------

# Check dMMR proteins

```{r}
oncoplot(combined_maf_dp20_vaf, genes = dmmr_ptns,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE)

combined_maf_dp20_vaf_df_dmmr <- combined_maf_dp20_vaf_df %>%
  filter(SYMBOL %in% dmmr_ptns) %>%
  mutate(Symbol_Variant = sprintf("%s_%s", SYMBOL, HGVSp_Short))

combined_maf_dp20_vaf_df_dmmr %>% select(all_of(c(variant_eval_columns, "Patient", "Trial", "Tumor_Sample_Barcode"))) %>% view()

combined_maf_dp20_vaf_df_dmmr_f <- combined_maf_dp20_vaf_df_dmmr %>%
  filter(!Symbol_Variant %in% c("PMS2_p.S459L", "MLH1_p.E234K", "MSH6_p.L310F",
"PMS2_p.P509S", "PMS2_p.P470S", "MSH2_p.G426E", "PMS2_p.S547L", "PMS2_p.P470F",
"MSH6_p.P831S", "PMS2_p.E491K", "MLH1_p.S421G", "MSH6_p.R361H", "MSH2_p.V932A",
"PMS2_p.S498G", "MLH1_p.P169S", "PMS2_p.S164F", "MLH1_p.S421G", 
"MSH6_p.A23V", "MSH6_p.E368K", "MSH6_p.R361H", "MLH1_p.I655M", "PMS2_p.E5K") |
  Trial == "NICHE-MSI")

combined_maf_dp20_vaf_df_dmmr_removed <- combined_maf_dp20_vaf_df_dmmr %>%
  filter(!Symbol_Variant %in% combined_maf_dp20_vaf_df_dmmr_f$Symbol_Variant)

combined_maf_dp20_vaf_df_dmmr_removed %>%
  select(all_of(c(variant_eval_columns, "Patient", "Tumor_Sample_Barcode")))

combined_maf_dp20_vaf_df_dmmr_f %>%
  select(all_of(c(variant_eval_columns, "Patient", "Tumor_Sample_Barcode")))

# "MSH6_p.X1180_splice", "MSH2_p.P696L",
#                                "MLH1_p.S44F", "PMS2_p.R563*", "MSH6_p.P903S",
#                                "MSH6_p.R300W", "MSH2_p.S325F", "MLH1_p.I8V",
#                                "MLH1_p.R265C", "PMS2_p.P760S"

# add missing patients as factor levels, still used in table with 0 counts
combined_maf_dp20_vaf_df_dmmr_f$Patient <- factor(
  combined_maf_dp20_vaf_df_dmmr_f$Patient,
  levels = unique(combined_maf_dp20_vaf_df$Patient))

combined_maf_dp20_vaf_df_dmmr_f %>%
   select(all_of(variant_eval_columns))

combined_maf_dp20_vaf_df_dmmr_f %>%
   select(all_of(variant_eval_columns)) %>%
  view()

combined_maf_dp20_vaf_df_dmmr_f %>%
  select(all_of(variant_eval_columns)) %>%
  mutate_if(is.character, as.factor) %>%
  summary(maxsum = 10)


combined_maf_dp20_vaf_df_dmmr_f_unique <- combined_maf_dp20_vaf_df_dmmr_f %>%
  select(Patient, Trial, Response) %>% unique()
table(combined_maf_dp20_vaf_df_dmmr_f_unique$Trial)
table(combined_maf_dp20_vaf_df_dmmr_f_unique$Response)

table(combined_maf_dp20_vaf_df_dmmr_f_unique$Trial,
      combined_maf_dp20_vaf_df_dmmr_f_unique$Response)

# Not accounting for all dMMR MSI tumors from NICHE
combined_maf_dp20_vaf_df_dmmr_f %>% filter(Trial ==  "NICHE-MSI") %>% nrow()
combined_maf_dp20_vaf_df_dmmr_f_unique %>% filter(Trial ==  "NICHE-MSI") %>% nrow()
table(combined_maf_dp20_vaf_df_dmmr_f$Trial)
table(combined_maf_dp20_vaf_df_dmmr_f_unique$Trial,
      combined_maf_dp20_vaf_df_dmmr_f_unique$Response)

# Added NICHE MSI manually
# dMMR status assigned from independent test, not from WES data
# e.g. MLH1 hypermethylation + IHC loss
dmmr_trial_response <- table(combined_maf_dp20_vaf_df_dmmr_f_unique$Trial,
                             combined_maf_dp20_vaf_df_dmmr_f_unique$Response) %>%
  as.data.frame() %>% setNames(c("Trial", "Response", "Freq")) 
  # mutate(Freq = ifelse(Trial == "NICHE-MSI" & Response == "R", 19, Freq))

dmmr_trial_response %>% pivot_wider(names_from = Response, values_from = Freq)
dmmr_trial_response %>% filter(!Trial %in% c("NICHE-MSI")) %>%
  pivot_wider(names_from = Response, values_from = Freq)


dmmr_trial_response %>%
  ggplot(aes(x = Trial, y = Freq, fill = Response)) + geom_col() +
  theme_Publication() + labs(x = "", y = "dMMR Freq") +
  scale_fill_manual(values = c(Color_NR, Color_R))
ggsave(here(output_dir, "dmmr_trial_response.png"), height = 5, width = 6.5)

dmmr_trial_response %>%
  group_by(Response) %>% summarize(Total = sum(Freq)) %>%
  ggplot(aes(x = Response, y = Total, fill = Response)) + geom_col() +
  theme_Publication() + labs(x = "", y = "dMMR Freq") +
  scale_fill_manual(values = c(Color_NR, Color_R))
ggsave(here(output_dir, "dmmr_response.png"), height = 5, width = 4)

dmmr_trial_response %>%
  group_by(Response) %>% summarize(Freq = sum(Freq)) %>%
  mutate(Total = sum(Freq), percentage = Freq / Total * 100) %>%
  ggplot(aes(x = "dMMR", y = percentage, fill = Response)) + geom_col() +
  theme_Publication() + labs(x = "", y = "dMMR % Response") +
  scale_fill_manual(values = c(Color_NR, Color_R))
ggsave(here(output_dir, "dmmr_response_percentage.png"), height = 5, width = 4)

combined_maf_dp20_vaf_df_dmmr_f %>%
  write_csv2(here(output_dir, "combined_maf_dp20_vaf_df_dmmr_f.csv"))

combined_maf_dp20_vaf_df_dmmr_f <- read_csv2(
  here(output_dir, "combined_maf_dp20_vaf_df_dmmr_f.csv"))
  
table(combined_maf_dp20_vaf_df_dmmr_f$Trial)
# PMS2 lower TMB, less immunogenic, less pCR?
# Look at WES, RNA and IHC for msi proteins

combined_maf_dp20_vaf_df_dmmr_f_unique %>%
  left_join(n_variants_dp20_vaf_tmb) %>%
  ggplot(aes(x = Trial, y = TMB)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(col = Trial, shape = Response)) +
  theme_Publication() +
  geom_hline(yintercept = 10, linetype = "dashed")

combined_maf_dp20_vaf_df_dmmr_f %>%
  # select(-Arm) %>%
  left_join(n_variants_dp20_vaf_tmb) %>%
  ggplot(aes(x = SYMBOL, y = TMB)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(col = Trial, shape = Response)) +
  theme_Publication() +
  geom_hline(yintercept = 10, linetype = "dashed")

dmmr_mutation_df <- table(combined_maf_dp20_vaf_df_dmmr_f$Patient,
                          combined_maf_dp20_vaf_df_dmmr_f$Hugo_Symbol) %>%
  as.data.frame() %>%
  setNames(c("Patient", "Gene_somatic_germline", "Freq")) %>%
  pivot_wider(names_from = Gene_somatic_germline, values_from = Freq) %>%
  left_join(dna_metadata_f)

dmmr_mutation_df %>% write_csv2(here(output_dir,
                                         "dmmr_mutation_df.csv"))
dmmr_mutation_df <- read_csv2(here(output_dir,
                                         "dmmr_mutation_df.csv"))

dmmr_mutation_df %>% filter(Trial == "NICHE-MSI") %>% view()

combined_maf_dp20_vaf_df_dmmr_f %>% filter(Trial == "BELLINI") %>%
  filter(Hugo_Symbol %in% dmmr_ptns) %>%
  select(Patient, variant_eval_columns) %>% view()

combined_maf_dp20_vaf_df_dmmr_f %>% filter(Trial == "NABUCCO") %>%
  filter(Hugo_Symbol %in% dmmr_ptns) %>%
  select(Patient, variant_eval_columns) %>% view()

combined_maf_dp20_vaf_df_dmmr_f %>% filter(Trial == "MATISSE") %>%
  filter(Hugo_Symbol %in% dmmr_ptns) %>%
  select(Patient, variant_eval_columns) %>% view()

combined_maf_dp20_vaf_df_dmmr_f %>% filter(Trial == "NICHE-MSI") %>%
  filter(Hugo_Symbol %in% dmmr_ptns) %>%
  select(Patient, variant_eval_columns) %>% view()

n_variants_dp20_vaf_tmb %>%
  mutate(dMMR_hit = ifelse(
    Tumor_Sample_Barcode %in% combined_maf_dp20_vaf_df_dmmr_f$Tumor_Sample_Barcode,
    "dMMR", "no dMMR hit")) %>%
  ggplot(aes(x = dMMR_hit, y = TMB, color = Histology)) + geom_point() +
  theme_Publication() +
  scale_color_manual(values = histology_msi_mss_colors)

combined_maf_dmmr_f <- subsetMaf(
  combined_maf, tsb = combined_maf_dp20_vaf_df_dmmr_f$Tumor_Sample_Barcode)

combined_maf_dp20_vaf_df_dmmr_f %>% select(Trial, Patient) %>% unique() %>% pull(Trial) %>% table()

oncoplot(combined_maf_dmmr_f, genes = dmmr_ptns)

plotVaf(combined_maf, genes = dmmr_ptns)

```

# ----------

# Check gene mutation frequencies

```{r}
# Check genes from oncokb not present in the cohort
table(oncoKB_cancerGeneList$Hugo.Symbol %in% combined_maf_dp20_vaf_df$Hugo_Symbol)
oncoKB_cancerGeneList$Hugo.Symbol[!oncoKB_cancerGeneList$Hugo.Symbol %in% combined_maf_dp20_vaf_df$Hugo_Symbol]

combined_maf_dp20_vaf_df %>% select(Tumor_Sample_Barcode) %>% n_distinct()

unique_sample_gene_variants <- combined_maf_dp20_vaf_df %>% select(Hugo_Symbol, Tumor_Sample_Barcode) %>% unique()

top_genes_n_samples <- table(unique_sample_gene_variants$Hugo_Symbol) %>%
  as.data.frame() %>%
  setNames(c("Gene", "Freq_all")) %>%
  arrange(-Freq_all) %>%
  mutate(Total = dna_metadata_f %>%
           select(Tumor_Sample_Barcode) %>%
           n_distinct(),
         Prop_all = Freq_all / Total,
         OncoKB_list = Gene %in% oncoKB_cancerGeneList$Hugo.Symbol)

for (current_trial in unique(combined_maf_dp20_vaf_df$Trial)) {
  current_data <- combined_maf_dp20_vaf_df %>% filter(Trial == current_trial)
  trial_unique_sample_gene_variants <- current_data %>% select(Hugo_Symbol, Tumor_Sample_Barcode) %>% unique()
  
  Trial_total <- dna_metadata_f %>%
             filter(Trial == current_trial) %>% 
             select(Tumor_Sample_Barcode) %>%
             n_distinct()
  
  top_genes_n_samples_trial <- table(trial_unique_sample_gene_variants$Hugo_Symbol) %>%
    as.data.frame() %>%
    setNames(c("Gene", "Freq")) %>%
    arrange(-Freq) 
  
  top_genes_n_samples_trial <- top_genes_n_samples %>% select(Gene) %>% left_join(top_genes_n_samples_trial, by = "Gene")

  top_genes_n_samples_trial <- top_genes_n_samples_trial %>%
    mutate(Freq = ifelse(is.na(Freq), 0 , Freq)) %>%
    mutate(
      Trial_total = Trial_total,
      Prop = Freq / Trial_total) 
  
  top_genes_n_samples_trial <- top_genes_n_samples_trial %>%
    setNames(c("Gene", paste0("Freq_", current_trial),  paste0("Trial_total_", current_trial), paste0("Prop_", current_trial)))
  colnames(top_genes_n_samples_trial) <- gsub("-", "_", colnames(top_genes_n_samples_trial) )
  
  
  top_genes_n_samples <- top_genes_n_samples_trial %>% left_join(top_genes_n_samples, by = "Gene")
 
}

top_genes_n_samples %>% write_csv2(here(output_dir, "top_genes_n_samples.csv"))


top_genes_n_samples %>%
  ggplot(aes(x = Prop_all, fill = OncoKB_list, alpha = 0.5)) +
  geom_density() +
  theme_Publication()

```

# Number of genes over mutation threshold

```{r}
check_proportion_thresholds <- function(df, sequence = seq(0, 1, by = 0.01)) {
  result <- data.frame(threshold = numeric(), num_rows = numeric())
  
  for (threshold in sequence) {
    num_rows <- sum(apply(df, 1, function(row) all(row > threshold)))
    result <- rbind(result, data.frame(threshold = threshold, num_rows = num_rows))
  }
  
  return(result)
}

gene_proportion_thresholds_broad <- check_proportion_thresholds(
  df = top_genes_n_samples %>%
    select("Gene", starts_with("Prop")) %>%
    select(-Prop_all),
  sequence = seq(0, 1, by = 0.1))

gene_proportion_thresholds_broad

gene_proportion_thresholds <- check_proportion_thresholds(
  df = top_genes_n_samples %>%
    select("Gene", starts_with("Prop")) %>%
    select(-Prop_all),
  sequence = seq(0, 0.25, by = 0.005))

# gene_proportion_thresholds

gene_proportion_thresholds %>%
  filter(threshold > 0) %>%
  ggplot(aes(x = threshold, y = num_rows)) +
  geom_point() + theme_Publication() +
  labs(x = "Proportion threshold", y = "Number of genes")

mutation_proportions <- top_genes_n_samples %>%
    select("Gene", starts_with("Prop")) %>%
    select(-Prop_all) %>% column_to_rownames("Gene")

mutation_proportions[rowSums(mutation_proportions > 0.2) == ncol(mutation_proportions), ]
# mutation_proportions[rowSums(mutation_proportions > 0.1) == ncol(mutation_proportions), ]
# mutation_proportions[rowSums(mutation_proportions > 0.05) == ncol(mutation_proportions), ]

genes_above_threshold <- mutation_proportions[rowSums(mutation_proportions > 0.05) == ncol(mutation_proportions), ]

table(rownames(genes_above_threshold) %in% genes_to_assess)
rownames(genes_above_threshold)[rownames(genes_above_threshold) %in% genes_to_assess]
rownames(genes_above_threshold)[rownames(genes_above_threshold) %in% oncoKB_cancerGeneList$Hugo.Symbol]

top_genes_n_samples %>% filter(Gene %in% rownames(genes_above_threshold))

top_genes_n_samples %>%
  filter(Gene %in% genes_to_assess) %>%
  view()

top_genes_n_samples %>% filter(Gene %in% c("TP53" , "KRAS", "NRAS", "HRAS", "BRAF", "BRCA1", "BRCA2", "CDK2NA", "B2M"))
top_genes_n_samples %>% filter(Gene %in% c("TP53" , "KRAS", "NRAS", "HRAS", "BRAF", "BRCA1", "BRCA2", "CDK2NA", "B2M")) %>% view()

combined_maf_dp20_vaf_df_specific_f %>% select(variant_eval_columns) %>% view()
```

```{r}
mutation_proportions[
  (rowSums(mutation_proportions > 0.05) == ncol(mutation_proportions)), ]  %>%
  pheatmap()

mutation_proportions[
  (rowSums(mutation_proportions > 0.05) == ncol(mutation_proportions)) &
      (rownames(mutation_proportions) %in% genes_to_assess), ] %>%
  pheatmap()
```

```{r, fig.width = 5, fig.height = 13}

mutation_proportions[c((rowMeans(mutation_proportions) > 0.1) & (rownames(mutation_proportions) %in% genes_to_assess)), ]%>%
  pheatmap()

mutation_proportions[c((rowMeans(mutation_proportions) > 0.2) & (rownames(mutation_proportions) %in% genes_to_assess)), ]%>%
  pheatmap()

mutation_proportions[c((rowMeans(mutation_proportions) > 0.2) & (!rownames(mutation_proportions) %in% genes_to_assess)), ]

mutation_proportions[rowMeans(mutation_proportions) > 0.2, ] %>% pheatmap()
 
table(rowSums(mutation_proportions > 0.1) == ncol(mutation_proportions))
```

# PCA on mutation frequencies

```{r}
mutation_proportions_z_norm <- mutation_proportions %>% t() %>% scale()

mutation_proportions_z_norm[1:5, 1:5]
rownames(mutation_proportions_z_norm) <- gsub("Prop_", "", rownames(mutation_proportions_z_norm))

rowSums(mutation_proportions_z_norm)
head(colSums(mutation_proportions_z_norm))

pca <- prcomp(mutation_proportions_z_norm)

percent_var <- pca$sdev^2 / sum(pca$sdev^2)

percent_var <- percent_var %>%
  as.data.frame() %>%
  setNames("Var_exp") %>%
  rownames_to_column("PC") %>%
  mutate(Cumvar = cumsum(Var_exp), PC = as.numeric(PC))

percent_var %>%
  filter(PC < 50) %>%
  ggplot(aes(x = PC, y = Var_exp)) +
  geom_col() +
  geom_point(aes(x = PC, y = Cumvar)) +
  labs(x = "Principal Component", y = "(Cumulative) Variance Explained") +
  theme_Publication()

percent_var_pc1 <- percent_var[1, 2] * 100
percent_var_pc2 <- percent_var[2, 2] * 100
percent_var_pc3 <- percent_var[3, 2] * 100

n_pcs <- 8
pca_heatmap <- pheatmap(t(pca$x[, 1:n_pcs]),
         # annotation_col = annot_df,
         col = paletteer_d(palette = "RColorBrewer::RdYlBu", direction = -1),
         show_colnames = TRUE,
         cluster_rows = FALSE,
         )
print(pca_heatmap)


pca_data <- as.data.frame(pca$x) %>%
  rownames_to_column("Trial")
  
pca_data %>%
  ggplot(aes(x = PC1, y = PC2, col = Trial)) +
  geom_point(size = 2.5) +
  labs(x = sprintf("PC1: %s%% variance", round(percent_var_pc1, 2)),
       y = sprintf("PC2: %s%% variance", round(percent_var_pc2, 2)),
       col = "Response",
       title = "Genetic landscape - mutational frequencies") +
  theme_Publication()

pca_data %>% write_csv2(here(output_dir, "pca_data_mutation_frequencies.csv"))
saveRDS(pca, here(output_dir, "pca_mutation_frequencies.RDS"))

# pca$rotation[, "PC1"]
loadings <- as.data.frame(pca$rotation) %>%
  rownames_to_column("Gene") %>%
  arrange(PC1)


loadings[select_top_extremes(loadings$PC1, 10), ] %>%
  ggplot(aes(x = reorder(Gene, PC1), y = PC1)) +
  geom_col() +
  coord_flip() +
  labs(x = "PC1 Loadings") +
  theme_Publication()

loadings[select_top_extremes(loadings$PC2, 10), ] %>%
  ggplot(aes(x = reorder(Gene, PC2), y = PC2)) +
  geom_col() +
  coord_flip() +
  labs(x = "PC2 Loadings") +
  theme_Publication()
```

# ----------

# Assess mutations on gene sets

```{r, fig.width = 16, fig.height = 26}
oncoplot(combined_maf_dp20_vaf, genes = genes_to_assess_shortlist,
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")
plotVaf(combined_maf_dp20_vaf, genes = genes_to_assess_shortlist)

oncoplot(combined_maf_dp20_vaf, genes = maftools_sigpw_oncogenes$Gene,
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

# Oncogene ore stringent, OncoKB and MAFtools
oncoplot(combined_maf_dp20_vaf, genes = maftools_sigpw_oncogenes %>%
           filter(Gene %in% oncoKB_oncogenes$Hugo.Symbol) %>%
           pull(Gene),
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

plotVaf(combined_maf_dp20_vaf, genes = maftools_sigpw_oncogenes$Gene)

# TSG ore stringent, OncoKB and MAFtools
oncoplot(combined_maf_dp20_vaf, genes = maftools_sigpw_tumor_suppressors %>%
           filter(Gene %in% oncoKB_tumor_suppressors$Hugo.Symbol) %>%
           pull(Gene)
,
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")
plotVaf(combined_maf_dp20_vaf, genes = maftools_sigpw_tumor_suppressors$Gene)

```

# Functional filtering

```{r, fig.width = 11, fig.height = 13}
combined_maf_specific <- subsetMaf(
  combined_maf_dp20_vaf,
  tsb = combined_maf_dp20_vaf@clinical.data$Tumor_Sample_Barcode,
  genes = genes_to_assess)

# add column of filtered variants
# query by maf_specific_f based on this column
# make oncoplot

genes_to_assess[!genes_to_assess %in% combined_maf_dp20_vaf_df$SYMBOL]

combined_maf_dp20_vaf_df_specific <- combined_maf_dp20_vaf_df %>%
  filter(SYMBOL %in% genes_to_assess) %>%
  mutate(Symbol_Variant = sprintf("%s_%s", SYMBOL, HGVSp_Short))

nrow(combined_maf_dp20_vaf_df_specific)

n_distinct(combined_maf_dp20_vaf_df_specific$Symbol_Variant)

unique(combined_maf_dp20_vaf_df_specific$CLIN_SIG)

combined_maf_dp20_vaf_df_specific %>%
  select(Patient, Tumor_Sample_Barcode, all_of(variant_eval_columns)) %>% view()

remove_likely_benign_categories <- c("benign", "likely_benign", "uncertain_significance,likely_benign", "benign,likely_benign,benign/likely_benign", "likely_benign,uncertain_significance", "benign/likely_benign", "likely_benign,not_provided,uncertain_significance", "likely_benign,conflicting_interpretations_of_pathogenicity", "benign,not_provided", "benign/likely_benign,uncertain_significance,benign", "benign,conflicting_interpretations_of_pathogenicity", "benign,likely_benign,not_provided", "benign,likely_benign", "uncertain_significance,benign,likely_benign,benign/likely_benign", "benign,likely_benign,uncertain_significance", "uncertain_significance,benign,likely_benign,benign/likely_benign", "benign,uncertain_significance", "benign,likely_benign,conflicting_interpretations_of_pathogenicity", "uncertain_significance,benign,likely_benign,not_provided", "conflicting_interpretations_of_pathogenicity,likely_benign", "benign,uncertain_significance,likely_benign", "likely_benign,benign/likely_benign,conflicting_interpretations_of_pathogenicity", "benign,uncertain_significance,likely_benign,benign/likely_benign", "likely_benign,uncertain_significance,benign/likely_benign,conflicting_interpretations_of_pathogenicity", "benign,likely_benign,benign/likely_benign,conflicting_interpretations_of_pathogenicity", "benign,uncertain_significance,benign/likely_benign", "benign,conflicting_interpretations_of_pathogenicity,likely_benign", "uncertain_significance,benign,conflicting_interpretations_of_pathogenicity,likely_benign",  "uncertain_significance,benign","likely_benign,uncertain_significance,conflicting_interpretations_of_pathogenicity", "uncertain_significance,conflicting_interpretations_of_pathogenicity,likely_benign", "uncertain_significance,conflicting_interpretations_of_pathogenicity,likely_benign,not_provided", "benign,likely_benign,not_provided,benign/likely_benign","uncertain_significance,likely_benign,conflicting_interpretations_of_pathogenicity", "not_provided,benign,likely_benign", "conflicting_interpretations_of_pathogenicity,uncertain_significance,benign", "uncertain_significance,benign,conflicting_interpretations_of_pathogenicity")

# keep mutations is sift and polyphen are NA
# Filter tolerated and benign annotations, unless clin_sig is pathogenic, or NA
# Remove in cases of conflicting/uncertain pathogenicity
combined_maf_dp20_vaf_df_specific_f <- combined_maf_dp20_vaf_df_specific %>%
  filter(!(
    str_detect(combined_maf_dp20_vaf_df_specific$SIFT, "tolerated") &
      str_detect(combined_maf_dp20_vaf_df_specific$PolyPhen, "benign") &
      (!str_detect(combined_maf_dp20_vaf_df_specific$CLIN_SIG, "pathogenic") | is.na(CLIN_SIG) | CLIN_SIG %in% c("uncertain_significance,conflicting_interpretations_of_pathogenicity", "uncertain_significance,conflicting_interpretations_of_pathogenicity,benign/likely_benign", "conflicting_interpretations_of_pathogenicity", "conflicting_interpretations_of_pathogenicity,uncertain_significance", "uncertain_significance", "conflicting_interpretations_of_pathogenicity,uncertain_significance,benign"))
    ) | (is.na(PolyPhen) & is.na(SIFT))) %>%
  filter(!CLIN_SIG %in% remove_likely_benign_categories) 
  # %>% filter(!Symbol_Variant %in% remove_extra_variants)

unique(combined_maf_dp20_vaf_df_specific_f$CLIN_SIG)

combined_maf_dp20_vaf_df_specific_f %>% filter(CLIN_SIG %in% c("risk_factor,benign", "uncertain_significance,benign,conflicting_interpretations_of_pathogenicity", "likely_benign,uncertain_significance,conflicting_interpretations_of_pathogenicity", "uncertain_significance,conflicting_interpretations_of_pathogenicity,likely_benign", "uncertain_significance,conflicting_interpretations_of_pathogenicity,likely_benign,not_provided", "benign,likely_benign,not_provided,benign/likely_benign", "uncertain_significance,likely_benign,conflicting_interpretations_of_pathogenicity", "not_provided,benign,likely_benign", "uncertain_significance,conflicting_interpretations_of_pathogenicity")) %>% select(Patient, variant_eval_columns) %>% view()

table(is.na(combined_maf_dp20_vaf_df_specific_f$CLIN_SIG))
table(is.na(combined_maf_dp20_vaf_df_specific_f$CLIN_SIG)) %>% prop.table() %>% round(2)

combined_maf_dp20_vaf_df_specific_f %>% filter(CLIN_SIG %in% c("uncertain_significance,conflicting_interpretations_of_pathogenicity")) %>% select(Patient, variant_eval_columns)
combined_maf_dp20_vaf_df_specific_f %>% filter(CLIN_SIG %in% c("uncertain_significance")) %>% select(Patient, variant_eval_columns)

# Check oncogene hits
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_oncogenes$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_oncogenes$Gene) %>%
  select(Patient, all_of(variant_eval_columns)) %>%
  factor_summary()
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_oncogenes$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_oncogenes$Gene) %>%
  select(Patient, all_of(variant_eval_columns)) %>%
  view()
## Check NA for Clinsig
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_oncogenes$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_oncogenes$Gene) %>% pull(CLIN_SIG) %>% is.na() %>% table() 
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_oncogenes$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_oncogenes$Gene) %>% pull(CLIN_SIG) %>% is.na() %>% table() %>% prop.table() %>% round(2)

# Check tumor suppressor hits
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_tumor_suppressors$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_tumor_suppressors$Gene) %>%
  select(Patient, all_of(variant_eval_columns)) %>%
  factor_summary()
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_tumor_suppressors$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_tumor_suppressors$Gene) %>%
  select(Patient, all_of(variant_eval_columns))%>%
  view()

## Check NA for Clinsig
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_tumor_suppressors$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_tumor_suppressors$Gene)  %>% pull(CLIN_SIG) %>% is.na() %>% table() 
combined_maf_dp20_vaf_df_specific_f %>% filter(Hugo_Symbol %in% oncoKB_tumor_suppressors$Hugo.Symbol | Hugo_Symbol %in% maftools_sigpw_tumor_suppressors$Gene)  %>% pull(CLIN_SIG) %>% is.na() %>% table() %>% prop.table() %>% round(2)
oncoKB_cancerGeneList %>% filter(Is.Oncogene == "Yes") %>% view()
oncoKB_cancerGeneList %>% filter(Is.Tumor.Suppressor.Gene == "Yes") %>% view()

table(oncoKB_cancerGeneList$Is.Oncogene)

# Check variants remaining that have tolerated and benign predictions
# Only kept if there is evidence for variant pathogenicity
combined_maf_dp20_vaf_df_specific_f %>%
  filter(str_detect(combined_maf_dp20_vaf_df_specific_f$SIFT, "tolerated") &
      str_detect(combined_maf_dp20_vaf_df_specific_f$PolyPhen, "benign")) %>%
  select(Patient, all_of(variant_eval_columns)) %>% view()

combined_maf_dp20_vaf_df_specific_f %>%
  filter(str_detect(combined_maf_dp20_vaf_df_specific_f$SIFT, "tolerated") &
      str_detect(combined_maf_dp20_vaf_df_specific_f$PolyPhen, "benign")) %>%
  pull(CLIN_SIG) %>%
  table(useNA = "always")

combined_maf_dp20_vaf_df_specific_f %>%
  select(Patient, all_of(variant_eval_columns)) %>% view()

combined_maf_dp20_vaf_df_specific_removed <- combined_maf_dp20_vaf_df_specific %>%
  filter(!Symbol_Variant %in% combined_maf_dp20_vaf_df_specific_f$Symbol_Variant)

combined_maf_dp20_vaf_df_specific_removed %>%
  write_csv2(here(output_dir, "combined_maf_dp20_vaf_df_specific_removed.csv"))

# Check extra variants removed
combined_maf_dp20_vaf_df_specific_removed %>%
  filter(Symbol_Variant %in% remove_extra_variants) %>%
  select(Patient, all_of(variant_eval_columns)) %>% view()

combined_maf_dp20_vaf_df_specific_removed %>%
  select(Patient, all_of(variant_eval_columns)) %>% view()

combined_maf_dp20_vaf_df_specific_removed %>%
  select(Patient, all_of(variant_eval_columns)) %>%
  factor_summary()

combined_maf_dp20_vaf_df_specific_f %>%
  select(all_of(variant_eval_columns)) %>%
  factor_summary()

```

## Keep variants with evidence for functional impact

```{r, fig.width = 11, fig.height = 13}

# Add variant filter columns
combined_maf_specific@data$Variant_filter <- combined_maf_specific@data$Symbol_Variant %in%
  combined_maf_dp20_vaf_df_specific_f$Symbol_Variant

# Same column needs to be added to silent datas lot, otherwise checkTypos on column names fails
combined_maf_specific@maf.silent$Variant_filter <- FALSE

# Subset based on this filter column
combined_maf_specific_f <- subsetMaf(
  combined_maf_specific,
  tsb = combined_maf_specific@data$Tumor_Sample_Barcode %>% unique(),
  query = "Variant_filter == TRUE")

combined_maf_dp20_vaf@data$Symbol_Variant <- sprintf(
  "%s_%s", combined_maf_dp20_vaf@data$SYMBOL,
  combined_maf_dp20_vaf@data$HGVSp_Short)

combined_maf_specific@data$Variant_filter <- combined_maf_specific@data$Symbol_Variant %in%
  combined_maf_dp20_vaf_df_specific_f$Symbol_Variant

combined_maf_specific <- subsetMaf(
  combined_maf_dp20_vaf,
  tsb = combined_maf_dp20_vaf@clinical.data$Tumor_Sample_Barcode,
  genes = genes_to_assess)

# This seems to be dropping samples, the ones without
# any hits on the genes queried
nrow(combined_maf_dp20_vaf@clinical.data)
nrow(combined_maf_specific_f@clinical.data)

missing_samples <- combined_maf_dp20_vaf@clinical.data$Tumor_Sample_Barcode[
  !combined_maf_dp20_vaf@clinical.data$Tumor_Sample_Barcode %in%
    combined_maf_specific_f@clinical.data$Tumor_Sample_Barcode]

combined_maf_specific_f <- combined_maf_dp20_vaf

# Add variant filter columns
combined_maf_specific_f@data$Variant_filter <- combined_maf_specific_f@data$Symbol_Variant %in%
  combined_maf_dp20_vaf_df_specific_f$Symbol_Variant

# Same column needs to be added to silent datas lot, otherwise checkTypos on column names fails
combined_maf_specific_f@maf.silent$Variant_filter <- FALSE

combined_maf_specific_f@data$Keep_sample_post_f <- combined_maf_specific_f@data$Tumor_Sample_Barcode %in%
  missing_samples
combined_maf_specific_f@maf.silent$Keep_sample_post_f <- TRUE

# Subset based on this filter column
combined_maf_specific_f <- subsetMaf(
  combined_maf_specific_f,
  query = "Variant_filter == TRUE | Keep_sample_post_f == TRUE")
                                                       
```

# Oncoplots

```{r, fig.width = 16, fig.height = 45}
length(genes_to_assess)
oncoplot(combined_maf_specific_f, genes = genes_to_assess[1:250], 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, genes = genes_to_assess[251:500], 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, genes = genes_to_assess[501:750], 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, genes = genes_to_assess[751:1000], 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, genes = genes_to_assess[1001:1250], 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, genes = genes_to_assess[1251:1395], 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

```

```{r, fig.width = 11, fig.height = 13}

oncoplot(combined_maf_specific_f, genes = genes_to_assess_shortlist, 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, genes = genes_to_assess_shortlist, 
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE)

table(unique(combined_maf_dp20_vaf_df$Patient)) %>%
  as.data.frame() %>% filter(Freq > 1)

# Dev off only works from the console, and not from the rmd chunk. Copy and run
# do not run from chunks and knit
if (FALSE) {
pdf(file = here(output_dir, "_oncoplot_specific_genes_response.pdf"), width = 10, height = 14)
oncoplot(combined_maf_specific_f, genes = genes_to_assess_shortlist, 
         clinicalFeatures = c("Response", "Trial"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE,
         removeNonMutated = FALSE
         )
dev.off()

pdf(file = here(output_dir, "_oncoplot_specific_genes_trial.pdf"), width = 10, height = 14)
oncoplot(combined_maf_specific_f, genes = genes_to_assess_shortlist, 
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         topBarData = "TMB",
         drawColBar = TRUE,
         sortByAnnotation = TRUE,
         removeNonMutated = FALSE)
dev.off()
}
```

# Oncogenic Pathways - Functional filtering

```{r}
oncoplot(combined_maf_specific_f, pathways = maftools_sigpw_pathways %>%
           filter(Pathway == "RTK-RAS") %>%
           select(Gene, Pathway),
         selectedPathways = c("RTK-RAS"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, pathways = "smgbp",
         selectedPathways = c("RTK_signaling"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)
```

# Mutation matrix

```{r Mutation matrix}
genes_to_assess[!genes_to_assess %in% combined_maf_dp20_vaf_df_specific_f$Hugo_Symbol]

# add missing patients as factor levels, still used in table with 0 counts
combined_maf_dp20_vaf_df_specific_f$Patient <- factor(
  combined_maf_dp20_vaf_df_specific_f$Patient,
  levels = unique(combined_maf_dp20_vaf_df$Patient))

missing_patients <- dna_metadata_f$Patient[!dna_metadata_f$Patient %in%
                                           combined_maf_dp20_vaf_df_specific_f$Patient]

missing_patients <- missing_patients[missing_patients %in%
                                       combined_maf_dp20_vaf_df$Patient]

specific_mutation_m <- table(combined_maf_dp20_vaf_df_specific_f$Hugo_Symbol,
      combined_maf_dp20_vaf_df_specific_f$Patient)
dim(specific_mutation_m)

# Chose to label MAT 71 as R for now
annot_df <- dna_metadata_f %>%
  select(Trial, Patient, Response) %>%
  # filter(!(Patient == "MAT_71" & Response == "NR")) %>%
  unique() %>%
  tibble::column_to_rownames("Patient") %>%
  mutate_if(is.numeric, as.factor)

# Check unique mutations
combined_maf_dp20_vaf_df_specific_f_unique <- combined_maf_dp20_vaf_df_specific_f %>%
  select(Source_MAF, Patient, Hugo_Symbol, HGVSp) %>%
  unique()

# Some patients have multi hits in the same gene
multi_hits <- table(combined_maf_dp20_vaf_df_specific_f_unique$Patient,
      combined_maf_dp20_vaf_df_specific_f_unique$Hugo_Symbol) %>%
  as.data.frame() %>%
  filter(Freq > 1) %>%
  setNames(c("Patient", "Hugo_Symbol" , "Freq")) %>%
  mutate(Patient_Symbol = paste0(Patient, "_", Hugo_Symbol))

combined_maf_dp20_vaf_df_specific_f  %>%
  mutate(Patient_Symbol = paste0(Patient, "_", Hugo_Symbol)) %>%
  filter(Patient_Symbol %in% multi_hits$Patient_Symbol) %>%
  select(Source_MAF, Tumor_Sample_Barcode, Patient,
         Hugo_Symbol,Start_Position, End_Position,
         Variant_Classification, Variant_Type, HGVSc) %>%
  view()

specific_mutation_table <- table(
  combined_maf_dp20_vaf_df_specific_f_unique$Hugo_Symbol,
      combined_maf_dp20_vaf_df_specific_f_unique$Patient) %>%
  as.matrix() %>%
  as.data.frame() %>%
  setNames(c("Gene", "Patient", "Freq")) %>%
  mutate(High_moderate_mut = Freq > 0,
         Gene_high_moderate_mut = ifelse(Freq > 0,
                                         sprintf("%smt", Gene),
                                         sprintf("%swt", Gene)))

specific_mutation_char_df <- specific_mutation_table %>%
  select(-c(Freq, High_moderate_mut)) %>%
  pivot_wider(names_from = Gene, values_from = Gene_high_moderate_mut) %>%
  left_join(dna_metadata_f, by = "Patient")

specific_mutation_char_df %>% write_csv2(here(output_dir,
                                         "specific_mutation_char_df.csv"))

specific_mutation_df <- specific_mutation_table %>%
  select(-c(Freq, Gene_high_moderate_mut)) %>%
  pivot_wider(names_from = Gene, values_from = High_moderate_mut) %>%
  left_join(dna_metadata_f, by = "Patient")
 # %>% rename(BRAF_status = BRAF)
n_distinct(combined_maf_dp20_vaf_df$Patient)
n_distinct(dna_metadata_f$Patient)
n_distinct(combined_maf_dp20_vaf_df_specific_f_unique$Patient)

specific_mutation_df %>%
  mutate_if(is.character, as.factor) %>%
  summary()

specific_mutation_df %>% write_csv2(here(output_dir,
                                         "specific_mutation_df.csv"))

specific_mutation_df <- read_csv2(here(output_dir,
                                         "specific_mutation_df.csv"))
specific_mutation_df$Patient <- as.factor(specific_mutation_df$Patient)

```

# Pathway level

```{r}

ddr_ptns_f <- ddr_ptns[
  ddr_ptns %in% colnames(specific_mutation_df)]

immune_escape_genes_f <- immune_escape_genes[
  immune_escape_genes %in% colnames(specific_mutation_df)]

immune_escape_genes_df_f <- immune_escape_genes_df %>%
  mutate(Pathway.general = ifelse(
    Pathway.general == "IFN-? pathway", "IFN-y pathway", Pathway.general)) %>%
  filter(Gene %in% immune_escape_genes_f)

immune_escape_genes_f_hla <- immune_escape_genes_df_f[
  immune_escape_genes_df_f$Pathway.general == "HLA-I", "Gene"] 

immune_escape_genes_f_ap <- immune_escape_genes_df_f[
  immune_escape_genes_df_f$Pathway.general == "Antigen Presentation", "Gene"] 

immune_escape_genes_f_ifny <- immune_escape_genes_df_f[
  immune_escape_genes_df_f$Pathway.general == "IFN-y pathway", "Gene"] 

immune_escape_genes_f_cd58 <- immune_escape_genes_df_f[
  immune_escape_genes_df_f$Pathway.general == "CD58", "Gene"] 

immune_escape_genes_f_pdl1 <- immune_escape_genes_df_f[
  immune_escape_genes_df_f$Pathway.general == "PD-L1", "Gene"] 

immune_escape_genes_f_epigenetic <- immune_escape_genes_df_f[
  immune_escape_genes_df_f$Pathway.general == "Epigenetic", "Gene"] 

genes_to_assess_exp <- unique(
  c(ddr_ptns_f, immune_escape_genes_df_f$Gene))

maftools_sigpw_pathways %>% filter(Pathway == "Cell_Cycle") %>% pull(Gene) %>% unique()
maftools_BP_SMG_pathways %>% filter(Pathway == "Cell cycle") %>% pull(Gene) %>% unique()

maftools_sigpw_pathways %>% filter(Pathway == "NOTCH") %>% pull(Gene) %>% unique()

maftools_BP_SMG_pathways %>% filter(Pathway == "TGFB signaling") %>% pull(Gene) %>% unique()
maftools_sigpw_pathways %>% filter(Pathway == "TGF-beta") %>% pull(Gene) %>% unique()

maftools_BP_SMG_pathways %>% filter(Pathway == "RTK signaling") %>% pull(Gene) %>% unique()
maftools_sigpw_pathways %>% filter(Pathway == "RTK-RAS") %>% pull(Gene) %>% unique()

maftools_BP_SMG_pathways %>% filter(Pathway == "Genome integrity") %>% pull(Gene) %>% unique()
maftools_sigpw_pathways %>% filter(Pathway == "TP53") %>% pull(Gene) %>% unique()

maftools_BP_SMG_pathways %>% filter(Pathway == "Immune signaling") %>% pull(Gene) %>% unique()
maftools_BP_SMG_pathways %>% filter(Pathway == "MAPK signaling") %>% pull(Gene) %>% unique()
maftools_BP_SMG_pathways %>% filter(Pathway == "PI3K signaling") %>% pull(Gene) %>% unique()
maftools_sigpw_pathways %>% filter(Pathway == "PI3K") %>% pull(Gene) %>% unique()


table(maftools_sigpw_pathways$Gene %in% colnames(specific_mutation_df))
maftools_sigpw_pathways %>% filter(!Gene %in% colnames(specific_mutation_df))

sigpw_cell_cycle_pathway <- maftools_sigpw_pathways %>% filter(Pathway == "Cell_Cycle") %>% pull(Gene) %>% unique()
sigpw_cell_cycle_pathway <- sigpw_cell_cycle_pathway[sigpw_cell_cycle_pathway  %in% colnames(specific_mutation_df)]

sigpw_notch_signaling_pathway <- maftools_sigpw_pathways %>% filter(Pathway == "NOTCH") %>% pull(Gene) %>% unique()
sigpw_notch_signaling_pathway <- sigpw_notch_signaling_pathway[sigpw_notch_signaling_pathway  %in% colnames(specific_mutation_df)]

chromatin_histone_pathway <- maftools_BP_SMG_pathways %>% filter(Pathway == "Chromatin histone modifiers") %>% pull(Gene) %>% unique()
chromatin_histone_pathway <- chromatin_histone_pathway[chromatin_histone_pathway  %in% colnames(specific_mutation_df)]

protein_homeostasis_ubiquitination_pathway <- maftools_BP_SMG_pathways %>% filter(Pathway == "Protein homeostasis/ubiquitination") %>% pull(Gene) %>% unique()
protein_homeostasis_ubiquitination_pathway <- protein_homeostasis_ubiquitination_pathway[protein_homeostasis_ubiquitination_pathway  %in% colnames(specific_mutation_df)]

maftools_sigpw_pathways %>% select(-Gene) %>% factor_summary(maxsum = 50)
maftools_BP_SMG_pathways %>% select(-Gene, Cancer_type) %>% unique() %>% factor_summary(maxsum = 50)

```

```{r, fig.height = 10, fig.width = 10}
pathway_level_mutation_df <- specific_mutation_df
pathway_level_mutation_df$ddr_hits <- rowSums(pathway_level_mutation_df[, ddr_ptns_f])

# All GIE Martínez-Jiménez et al. - 2023
pathway_level_mutation_df$immune_hits <- rowSums(pathway_level_mutation_df[, immune_escape_genes_df_f$Gene])
# Specific GIEs
pathway_level_mutation_df$immune_hits_hla <- rowSums(pathway_level_mutation_df[, immune_escape_genes_f_hla]) # HLA-I
pathway_level_mutation_df$immune_hits_ap <- rowSums(pathway_level_mutation_df[, immune_escape_genes_f_ap])
pathway_level_mutation_df$immune_hits_ifny <- rowSums(pathway_level_mutation_df[, immune_escape_genes_f_ifny])
pathway_level_mutation_df$immune_hits_cd58 <- rowSums(pathway_level_mutation_df[, immune_escape_genes_f_cd58])
pathway_level_mutation_df$immune_hits_epi <- rowSums(pathway_level_mutation_df[, immune_escape_genes_f_epigenetic])
pathway_level_mutation_df$dmmr_hits <- rowSums(pathway_level_mutation_df[, dmmr_ptns])

# SMG
pathway_level_mutation_df$SMG_Apoptosis <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "Apoptosis")%>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_Wnt_B_Cat <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "Wnt/B-catenin signaling")%>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_TGFB_signaling <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "TGFB signaling") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_RTK_signaling <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "RTK signaling") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_Genome_integrity <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "Genome integrity") %>% pull(Gene) %>% unique()])

pathway_level_mutation_df$SMG_Cell_cycle <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "Cell cycle") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_Chromatin_histone_mod <- rowSums(pathway_level_mutation_df[, chromatin_histone_pathway])
pathway_level_mutation_df$SMG_Immune_signaling <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "Immune signaling") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_MAPK_signaling <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "MAPK signaling") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SMG_Protein_homeost_ub <- rowSums(pathway_level_mutation_df[, protein_homeostasis_ubiquitination_pathway])
pathway_level_mutation_df$SMG_PI3K_signaling <- rowSums(pathway_level_mutation_df[, maftools_BP_SMG_pathways %>% filter(Pathway == "PI3K signaling") %>% pull(Gene) %>% unique()])


# SIGPW
pathway_level_mutation_df$SIGPW_Cell_cycle <- rowSums(pathway_level_mutation_df[, sigpw_cell_cycle_pathway])
pathway_level_mutation_df$SIGPW_Hippo <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "Hippo") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_MYC <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "MYC") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_NOTCH_signaling <- rowSums(pathway_level_mutation_df[, sigpw_notch_signaling_pathway])
pathway_level_mutation_df$SIGPW_NRF2 <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "NRF2") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_PI3K_signaling <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "PI3K") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_TGFB_signaling <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "TGF-Beta") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_RTK_RAS <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "RTK-RAS") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_TP53 <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "TP53") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_Wnt_B_Cat <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "WNT") %>% pull(Gene) %>% unique()])

pathway_level_mutation_df %>%
  mutate(TP53_only = ddr_hits == 1 & as.numeric(TP53) == ddr_hits) %>%
  pivot_longer(cols = c("ddr_hits", "immune_hits"),
                      names_to = "Category", values_to = "Hits") %>%
  ggplot(aes(x = Category, y = Hits)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(col = Trial),
             position = position_jitter(w = 0.2, h = 0)) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

pathway_level_mutation_df %>%
  mutate(TP53_only = ddr_hits == 1 & as.numeric(TP53) == ddr_hits) %>%
  pivot_longer(cols = c("ddr_hits", "immune_hits"),
                      names_to = "Category", values_to = "Hits") %>%
  ggplot(aes(x = Category, y = Hits)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(col = Trial),
             position = position_jitter(w = 0.2, h = 0)) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  facet_wrap(~Trial)

# SMG

tail(colnames(pathway_level_mutation_df), 30)

# SIGPW
pathway_level_mutation_df$SIGPW_Cell_cycle <- rowSums(pathway_level_mutation_df[, sigpw_cell_cycle_pathway])
pathway_level_mutation_df$SIGPW_Hippo <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "Hippo") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_MYC <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "MYC") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_NOTCH_signaling <- rowSums(pathway_level_mutation_df[, sigpw_notch_signaling_pathway])
pathway_level_mutation_df$SIGPW_NRF2 <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "NRF2") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_PI3K_signaling <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "PI3K") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_TGFB_signaling <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "TGF-Beta") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_RTK_RAS <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "RTK-RAS") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_TP53 <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "TP53") %>% pull(Gene) %>% unique()])
pathway_level_mutation_df$SIGPW_Wnt_B_Cat <- rowSums(pathway_level_mutation_df[, maftools_sigpw_pathways %>% filter(Pathway == "WNT") %>% pull(Gene) %>% unique()])

pathway_level_mutation_df <- pathway_level_mutation_df %>%
  mutate(GIE = immune_hits > 0,
         GIE_IFNy = immune_hits_ifny > 0,
         GIE_AP = immune_hits_ap > 0,
         DDR = ddr_hits > 0,
         SMG_Apoptosis = SMG_Apoptosis > 0,
         SMG_Wnt_B_Cat = SMG_Wnt_B_Cat > 0,
         SMG_TGFB_signaling = SMG_TGFB_signaling > 0,
         SMG_RTK_signaling = SMG_RTK_signaling > 0,
         SMG_Genome_integrity = SMG_Genome_integrity > 0,
         SMG_Cell_cycle = SMG_Cell_cycle > 0,
         SMG_Chromatin_histone_mod = SMG_Chromatin_histone_mod > 0,
         SMG_Immune_signaling = SMG_Immune_signaling > 0,
         SMG_MAPK_signaling = SMG_MAPK_signaling > 0,
         SMG_Protein_homeost_ub = SMG_Protein_homeost_ub > 0,
         SMG_PI3K_signaling  = SMG_PI3K_signaling > 0,
         SIGPW_Cell_cycle = SIGPW_Cell_cycle > 0,
         SIGPW_Hippo = SIGPW_Hippo > 0,
         SIGPW_MYC = SIGPW_MYC > 0,
         SIGPW_NOTCH_signaling = SIGPW_NOTCH_signaling > 0,
         SIGPW_NRF2 = SIGPW_NRF2 > 0, 
         SIGPW_PI3K_signaling = SIGPW_PI3K_signaling > 0,
         SIGPW_TGFB_signaling = SIGPW_TGFB_signaling > 0,
         SIGPW_RTK_RAS = SIGPW_RTK_RAS > 0,
         SIGPW_TP53 = SIGPW_TP53 > 0,
         SIGPW_Wnt_B_Cat = SIGPW_Wnt_B_Cat > 0,

         )

n_variants_dp20_vaf_tmb_status <- n_variants_dp20_vaf_tmb %>%
  select(-Tumor_Sample_Barcode)  %>%
  left_join(pathway_level_mutation_df) %>%
  mutate(
    immune_hits_status = ifelse(
    GIE, "immune_hits", "immune_wt"),
    immune_hits_ifny_status = ifelse(
    GIE_IFNy > 0, "pIFNy_mut", "pIFNy_wt"),
    immune_hits_ap_status = ifelse(
    GIE_AP, "AP_mut", "AP_mut"))

table(pathway_level_mutation_df$Patient) %>%
  as.data.frame() %>% filter(Freq > 1)

pathway_level_mutation_df <- pathway_level_mutation_df %>%
  left_join(
    n_variants_dp20_vaf_tmb %>% select(c("Tumor_Sample_Barcode", colnames(n_variants_dp20_vaf_tmb)[
      !colnames(n_variants_dp20_vaf_tmb) %in% colnames(pathway_level_mutation_df)]))) %>%
  unique() 

pathway_level_mutation_df <- pathway_level_mutation_df %>%
  mutate(
    RAS_RAF = BRAF | KRAS | NRAS | HRAS,
    BRCA1_2 = BRCA1 | BRCA2)

pathway_level_mutation_df %>% write_csv2(here(output_dir,
                                         "pathway_level_mutation_df.csv"))

pathway_level_mutation_df <- read_csv2(here(output_dir,
                                         "pathway_level_mutation_df.csv"))
pathway_level_mutation_df$Patient <- as.factor(
  pathway_level_mutation_df$Patient)

str(pathway_level_mutation_df[, (ncol(pathway_level_mutation_df)-35-1):ncol(pathway_level_mutation_df)])
pathway_level_mutation_df[, (ncol(pathway_level_mutation_df)-35-1):ncol(pathway_level_mutation_df)] %>%
  factor_summary()

pathway_level_mutation_df %>%
  select("GIE", "GIE_IFNy", "GIE_AP", "DDR", "SMG_Wnt_B_Cat", "SMG_TGFB_signaling",
         "SMG_RTK_signaling", "SMG_Genome_integrity", "SMG_Cell_cycle",
         "SMG_Chromatin_histone_mod", "SMG_Immune_signaling", "SMG_MAPK_signaling",
         "SMG_Protein_homeost_ub", "SMG_PI3K_signaling") %>%
  factor_summary()

pathway_level_mutation_df %>% select(c("Patient", Trial, "ddr_hits", ddr_ptns_f)) %>% view()
pathway_level_mutation_df %>% select(c("Patient", Trial, "immune_hits", immune_escape_genes_f)) %>%
  view()
```

# Pathway level correlations

```{r, fig.height = 10, fig.width = 10}
 
pathway_level_mutation_m <- pathway_level_mutation_df %>%
  select(TMB, starts_with("SMG"), starts_with("GIE"), starts_with("SIGPW"), "DDR", "RAS_RAF", "BRCA1_2") %>%
  as.matrix() 

pathway_level_mutation_m %>%
  cor(use = "pairwise.complete.obs", method = "spearman") %>%
  corrplot(order = "hclust", type = "lower", tl.col = "black")

pathway_level_mutation_excl_dmmr_m <- pathway_level_mutation_df %>%
  filter(Trial != "NICHE-MSI") %>%
  select(TMB, starts_with("SMG"), starts_with("GIE"), starts_with("SIGPW"), "DDR", "RAS_RAF", "BRCA1_2") %>%
  as.matrix() 

pathway_level_mutation_excl_dmmr_m %>%
  cor(use = "pairwise.complete.obs", method = "spearman") %>%
  corrplot(order = "hclust", type = "lower", tl.col = "black")


table(pathway_level_mutation_df$SIGPW_Cell_cycle)
table(pathway_level_mutation_df$SIGPW_Cell_cycle, pathway_level_mutation_df$SMG_Cell_cycle)
table(pathway_level_mutation_df$SIGPW_PI3K_signaling, pathway_level_mutation_df$SMG_PI3K_signaling)
table(pathway_level_mutation_df$SIGPW_Wnt_B_Cat, pathway_level_mutation_df$SMG_Wnt_B_Cat)
table(pathway_level_mutation_df$SIGPW_RTK_RAS, pathway_level_mutation_df$SMG_RTK_signaling)

table(pathway_level_mutation_df$RAS_RAF)
table(pathway_level_mutation_df$RAS_RAF, pathway_level_mutation_df$SMG_MAPK_signaling)

```

```{r}
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$ddr_hits > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "DDR_hits", "Count")) %>%
  ggplot(aes(x = Trial, y = Count, fill = DDR_hits)) + geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggsave(here(output_dir, "ddr_hits_trial_freq.png"), height = 5, width = 4)

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$immune_hits > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "immune_hits", "Count")) %>%
  ggplot(aes(x = Trial, y = Count, fill = immune_hits)) + geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggsave(here(output_dir, "immune_hits_trial_freq.png"), height = 5, width = 4)

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$immune_hits > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "immune_hits", "Count")) %>%
  group_by(Trial) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = Trial, y = Percentage, fill = immune_hits)) + geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
  ggsave(here(output_dir, "immune_hits_trial_perc.png"), height = 5, width = 4)

table(pathway_level_mutation_df$immune_hits > 0,
      pathway_level_mutation_df$Response)

table(pathway_level_mutation_df$immune_hits > 0,
      pathway_level_mutation_df$Response) %>%
  as.data.frame() %>%
  setNames(c("immune_hits", "Response", "Count")) %>%
  group_by(Response) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = Response, y = Percentage, fill = immune_hits)) + geom_col() +
  theme_Publication() 
ggsave(here(output_dir, "immune_hit_freq.png"), height = 5, width = 4)

table(pathway_level_mutation_df$immune_hits > 0,
      pathway_level_mutation_df$Response) %>%
  as.data.frame() %>%
  setNames(c("immune_hits", "Response", "Count")) %>%
  group_by(immune_hits) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = immune_hits, y = Percentage, fill = Response)) + geom_col() +
  theme_Publication()+
  scale_fill_manual(values = c(Color_NR, Color_R))
ggsave(here(output_dir, "immune_hit_freq_r.png"), height = 5, width = 4)

table(pathway_level_mutation_df$ddr_hits > 0,
      pathway_level_mutation_df$Response) %>%
  as.data.frame() %>%
  setNames(c("ddr_hits", "Response", "Count")) %>%
  group_by(ddr_hits) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = ddr_hits, y = Percentage, fill = Response)) + geom_col() +
  theme_Publication() +
  scale_fill_manual(values = c(Color_NR, Color_R))
ggsave(here(output_dir, "ddr_hits_freq_r.png"), height = 5, width = 4)

table(pathway_level_mutation_df$ddr_hits > 0,
      pathway_level_mutation_df$Response) %>%
  as.data.frame() %>%
  setNames(c("ddr_hits", "Response", "Count")) %>%
  group_by(Response) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = Response, y = Percentage, fill = ddr_hits)) + geom_col() +
  theme_Publication()
```

# PCA on mutational status

```{r}
specific_mutation_m %>% dim()

specific_mutation_z_norm <- specific_mutation_m %>% t() %>% scale()

head(rowSums(specific_mutation_z_norm))
head(colSums(specific_mutation_z_norm))

pca <- prcomp(specific_mutation_z_norm)

percent_var <- pca$sdev^2 / sum(pca$sdev^2)

percent_var <- percent_var %>%
  as.data.frame() %>%
  setNames("Var_exp") %>%
  rownames_to_column("PC") %>%
  mutate(Cumvar = cumsum(Var_exp), PC = as.numeric(PC))

percent_var %>%
  filter(PC < 50) %>%
  ggplot(aes(x = PC, y = Var_exp)) +
  geom_col() +
  geom_point(aes(x = PC, y = Cumvar)) +
  labs(x = "Principal Component", y = "(Cumulative) Variance Explained") +
  theme_Publication()

percent_var_pc1 <- percent_var[1, 2] * 100
percent_var_pc2 <- percent_var[2, 2] * 100
percent_var_pc3 <- percent_var[3, 2] * 100

n_pcs <- 8
pca_heatmap <- pheatmap(t(pca$x[, 1:n_pcs]),
         # annotation_col = annot_df,
         col = paletteer_d(palette = "RColorBrewer::RdYlBu", direction = -1),
         show_colnames = TRUE,
         cluster_rows = FALSE,
         )
print(pca_heatmap)


pca_data <- as.data.frame(pca$x) %>%
  rownames_to_column("Patient") %>%
  left_join(dna_metadata_f)
  
pca_data %>%
  ggplot(aes(x = PC1, y = PC2, col = Trial)) +
  geom_point(size = 2.5) +
  labs(x = sprintf("PC1: %s%% variance", round(percent_var_pc1, 2)),
       y = sprintf("PC2: %s%% variance", round(percent_var_pc2, 2)),
       col = "Response",
       title = "Genetic landscape - mutational frequencies") +
  theme_Publication()

pca_data %>% write_csv2(here(output_dir, "pca_data_mutational_status.csv"))
saveRDS(pca, here(output_dir, "pca_data_mutational_status.csv.RDS"))

```

# PCA on mutational gene sets

```{r}
pathway_level_mutation_m <- pathway_level_mutation_df %>%
  select(Patient, starts_with("SMG"), starts_with("GIE"), starts_with("SIGPW"), "DDR") %>%
  column_to_rownames("Patient") %>% as.matrix() %>% mutate_all(as.numeric)
colnames(pathway_level_mutation_m)

table(rowSums(pathway_level_mutation_m) == 0)
pathway_level_mutation_m[rowSums(pathway_level_mutation_m) == 0, ]
pathway_level_mutation_df %>% filter(Patient %in% c("BEL30", "BEL34")) %>% view()

pathway_level_mutation_z_norm <- pathway_level_mutation_m[rowSums(pathway_level_mutation_m) != 0, ] %>% scale()
dim(pathway_level_mutation_z_norm)
pathway_level_mutation_z_norm[1:5, 1:5]

head(rowSums(pathway_level_mutation_z_norm))
head(colSums(pathway_level_mutation_z_norm))

pca <- prcomp(pathway_level_mutation_z_norm)

percent_var <- pca$sdev^2 / sum(pca$sdev^2)

percent_var <- percent_var %>%
  as.data.frame() %>%
  setNames("Var_exp") %>%
  rownames_to_column("PC") %>%
  mutate(Cumvar = cumsum(Var_exp), PC = as.numeric(PC))

percent_var %>%
  filter(PC < 50) %>%
  ggplot(aes(x = PC, y = Var_exp)) +
  geom_col() +
  geom_point(aes(x = PC, y = Cumvar)) +
  labs(x = "Principal Component", y = "(Cumulative) Variance Explained") +
  theme_Publication()

percent_var_pc1 <- percent_var[1, 2] * 100
percent_var_pc2 <- percent_var[2, 2] * 100
percent_var_pc3 <- percent_var[3, 2] * 100

n_pcs <- 8
pca_heatmap <- pheatmap(t(pca$x[, 1:n_pcs]),
         # annotation_col = annot_df,
         col = paletteer_d(palette = "RColorBrewer::RdYlBu", direction = -1),
         show_colnames = TRUE,
         cluster_rows = FALSE,
         )
print(pca_heatmap)

pca_data <- as.data.frame(pca$x) %>%
  rownames_to_column("Patient") %>%
  left_join(dna_metadata_f)
  
pca_data %>%
  ggplot(aes(x = PC1, y = PC2, col = Histology)) +
  geom_point(size = 2.5) +
  labs(x = sprintf("PC1: %s%% variance", round(percent_var_pc1, 2)),
       y = sprintf("PC2: %s%% variance", round(percent_var_pc2, 2)),
       title = "Genetic landscape - mutational pathways") +
  theme_Publication()
ggsave(here(output_dir, "pca_data_mutation_pathway.png"), width = 6, height = 6)

pca_data %>%
  ggplot(aes(x = PC1, y = PC2, col = Histology, shape = Response)) +
  geom_point(size = 2.5) +
  labs(x = sprintf("PC1: %s%% variance", round(percent_var_pc1, 2)),
       y = sprintf("PC2: %s%% variance", round(percent_var_pc2, 2)),
       shape = "Response",
       title = "Genetic landscape - mutational pathways") +
  theme_Publication()
ggsave(here(output_dir, "pca_data_mutation_pathway_response.png"), width = 6, height = 6)

pca_data %>% write_csv2(here(output_dir, "pca_data_mutation_pathway_status.csv"))
saveRDS(pca, here(output_dir, "pca_data_mutation_pathway_status.csv.RDS"))

# pca$rotation[, "PC1"]
loadings <- as.data.frame(pca$rotation) %>%
  rownames_to_column("Gene") %>%
  arrange(PC1)

loadings[select_top_extremes(loadings$PC1, 10), ] %>%
  ggplot(aes(x = reorder(Gene, PC1), y = PC1)) +
  geom_col() +
  coord_flip() +
  labs(x = "PC1 Loadings") +
  theme_Publication()

loadings[select_top_extremes(loadings$PC2, 10), ] %>%
  ggplot(aes(x = reorder(Gene, PC2), y = PC2)) +
  geom_col() +
  coord_flip() +
  labs(x = "PC2 Loadings") +
  theme_Publication()
```


# ----------

# AP and IFN muts

```{r, fig.height = 6, fig.width = 6}
oncoplot(combined_maf_dp20_vaf,
         pathways = immune_escape_genes_df %>%
           select(Gene, Pathway.general),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

immune_hits_ap_freq <- table(pathway_level_mutation_df$Trial,
                             pathway_level_mutation_df$immune_hits_ap > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "immune_hits_ap", "Freq")) %>%
  mutate(immune_hits_ap = ifelse(as.logical(immune_hits_ap), "AP_mut", "AP_wt")) %>%
  pivot_wider(names_from = immune_hits_ap, values_from = Freq) %>%
  mutate(Total = AP_wt + AP_mut,
         AP_mut_perc = AP_mut / Total * 100) %>%
  left_join(tmb_response_rate)
immune_hits_ap_freq

table(pathway_level_mutation_df$immune_hits_ap > 0,
      pathway_level_mutation_df$Response) %>%
  as.data.frame() %>%
  setNames(c("immune_hits_ap", "Response", "Count")) %>%
  mutate(immune_hits_ap = ifelse(as.logical(immune_hits_ap), "AP_mut", "AP_wt"))  %>%
  group_by(immune_hits_ap) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = immune_hits_ap, y = Percentage, fill = Response)) + geom_col() +
  theme_Publication()+
  scale_fill_manual(values = c(Color_NR, Color_R))

immune_hits_ap_freq %>%
  ggplot(aes(x = factor(Trial, levels = tmb_response_rate$Trial),
             y = AP_mut_perc, fill = Histology)) +
  geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(x = "")

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$immune_hits_ap > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "immune_hits_ap", "Count")) %>%
  ggplot(aes(x = Trial, y = Count, fill = immune_hits_ap)) + geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggsave(here(output_dir, "immune_hits_ap_trial_freq.png"),
       height = 5, width = 4)

immune_hits_ap_freq %>%
  ggplot(aes(x = median_TMB, y = AP_mut_perc, col = Histology_MSI_MSS)) +
  geom_point() +
  labs(x = "median_TMB (mutations/Mb)", y = "AP mutation %",  col = "") +
  theme_Publication()

immune_hits_ap_freq %>%
  ggplot(aes(x = median_TMB, y = AP_mut_perc, col = Histology_MSI_MSS)) +
  geom_point() +
  labs(x = "median_TMB (mutations/Mb)", y = "AP mutation %",  col = "") +
  scale_x_continuous(trans = "log", breaks = c(1, 10, 20, 30, 40, 50)) +
  theme_Publication()
ggsave(here(output_dir, "ap_mut_freq_tmb.png"), height = 5, width = 6)


table(pathway_level_mutation_df$Trial,
      pathway_level_mutation_df$immune_hits_ifny > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "immune_hits_ifny", "Count")) %>%
  ggplot(aes(x = Trial, y = Count, fill = immune_hits_ifny)) + geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggsave(here(output_dir, "immune_hits_ifny_trial_freq.png"),
       height = 5, width = 4)

immune_hits_ifny_freq <- table(
  pathway_level_mutation_df$Trial,
  pathway_level_mutation_df$immune_hits_ifny > 0) %>%
  as.data.frame() %>%
  setNames(c("Trial", "immune_hits_ifny", "Freq")) %>%
  mutate(immune_hits_ifny = ifelse(as.logical(immune_hits_ifny), "pIFNy_mut", "pIFNy_wt")) %>%
  pivot_wider(names_from = immune_hits_ifny, values_from = Freq) %>%
  mutate(Total = pIFNy_mut + pIFNy_wt,
         pIFNy_mut_perc = pIFNy_mut / Total * 100) %>%
  left_join(tmb_response_rate)

immune_hits_ifny_freq %>%
  ggplot(aes(x = median_TMB, y = pIFNy_mut_perc, col = Histology_MSI_MSS)) +
  geom_point() +
  labs(x = "median_TMB (mutations/Mb)", y = "pIFNy mutation %",  col = "") +
  theme_Publication() +
  scale_x_continuous(trans = "log", breaks = c(1, 10, 20, 30, 40, 50))
ggsave(here(output_dir, "pifny_mut_freq_tmb.png"), height = 5, width = 6)

n_variants_dp20_vaf_tmb_status %>%
  left_join(pathway_level_mutation_df) %>%
  ggplot(aes(x = immune_hits_status, y = TMB,
             fill = immune_hits_status)) +
  geom_boxplot() +
  theme_Publication() +
  labs(fill = "")
ggsave(here(output_dir, "immune_hits_status_tmb.png"), height = 5, width = 4)

 n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = immune_hits_status, y = TMB,
             fill = immune_hits_status)) +
  geom_boxplot() +
  theme_Publication() + facet_wrap(~Trial) + labs() +
  labs(fill = "")
 
n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = GIE_AP, y = TMB,
             fill = GIE_AP)) +
  geom_boxplot() +
  theme_Publication()  +
  labs(fill = "")

 n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = GIE_AP, y = TMB,
             fill = GIE_AP)) +
  geom_boxplot() +
  theme_Publication() + facet_wrap(~Trial) + labs() +
  labs(fill = "")

n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = GIE_IFNy, y = TMB,
             fill = GIE_IFNy)) +
  geom_boxplot() +
  theme_Publication() +
  labs(fill = "")

 n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = GIE_IFNy, y = TMB,
             fill = GIE_IFNy)) +
  geom_boxplot() +
  theme_Publication() + facet_wrap(~Trial) + labs() +
  labs(fill = "")
 
```

```{r, fig.width = 5, fig.height = 5}
table(pathway_level_mutation_df$Trial)
table(pathway_level_mutation_df$TMB > 10,
      pathway_level_mutation_df$immune_hits > 0,
      pathway_level_mutation_df$Response) %>%
  as.data.frame() %>%
  setNames(c("TMB", "immune_hits", "Response", "Count")) %>%
  mutate(TMB = ifelse(as.logical(TMB), "TMB-H", "TMB-L")) %>%
  group_by(TMB, immune_hits) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = immune_hits, y = Percentage, fill = Response)) + geom_col() +
  theme_Publication() +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  facet_wrap(~TMB)

pathway_level_mutation_df_non_dmmr <- pathway_level_mutation_df %>% filter(!Trial %in% c("NICHE-MSI"))
table(pathway_level_mutation_df_non_dmmr$TMB > 10,
      pathway_level_mutation_df_non_dmmr$immune_hits > 0,
      pathway_level_mutation_df_non_dmmr$Response) %>%
  as.data.frame() %>%
  setNames(c("TMB", "immune_hits", "Response", "Count")) %>%
  mutate(TMB = ifelse(as.logical(TMB), "TMB-H", "TMB-L")) %>%
  group_by(TMB, immune_hits) %>%
  mutate(Total = sum(Count), Percentage = Count / Total * 100)  %>% 
  ggplot(aes(x = immune_hits, y = Percentage, fill = Response)) + geom_col() +
  theme_Publication() +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  facet_wrap(~TMB) +
  labs(title = "Excluding dMMR CC")

```

# Mutation percentages - associations

```{r}
# MSI
mut_ci_counts <- mapply(FUN = counts_CI_for_frequency_tables,
                            column_name = genes_to_assess_shortlist, value = TRUE,
                            MoreArgs = list(data = specific_mutation_df))

mut_ci_counts <- as.data.frame(do.call(rbind, mut_ci_counts)) %>%
  mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper))

# R
mut_ci_counts_r <- mapply(
  FUN = counts_CI_for_frequency_tables, column_name = genes_to_assess_shortlist,
  value = TRUE, MoreArgs = list(data = specific_mutation_df %>%
                                filter(Response == "R")))

mut_ci_counts_r <- as.data.frame(
  do.call(rbind, mut_ci_counts_r)) %>%
  mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper))

# NR
mut_ci_counts_nr <- mapply(
  FUN = counts_CI_for_frequency_tables, column_name = genes_to_assess_shortlist,
  value = TRUE, MoreArgs = list(data = specific_mutation_df %>%
                                filter(Response == "NR")))

mut_ci_counts_nr <- as.data.frame(
  do.call(rbind, mut_ci_counts_nr)) %>%
  mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper))

mut_ci_counts_c <- mut_ci_counts_nr %>%
  mutate(Response = "NR") %>%
  rownames_to_column("Gene") %>%
  rbind(mut_ci_counts_r %>%
  mutate(Response = "R") %>%
  rownames_to_column("Gene"))

mut_ci_counts_c %>%
  write_csv2(here(output_dir, "mut_ci_counts_c.csv"))

```

```{r, fig.width = 16, fig.height = 8}
mut_ci_counts_c %>%
  ggplot(aes(x = Gene, y = percentage_of_total, fill = Response)) +
  geom_col(position = position_dodge()) +
  geom_errorbar(aes(ymin = CI_lower, ymax = CI_upper),
                position = position_dodge(width = 1), lwd = 0.25, width = 0.5) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(y = "Percentage (95% CI) of (non)responders with mutations") +
  scale_fill_manual(values = c(Color_NR, Color_R))

```

```{r, fig.width = 16, fig.height = 8}
get_wt_mt_response <- function(specific_mutation_df, genes_to_assess){
  mt_wt_response_ci <- data.frame(matrix(nrow = 0, ncol = 6))

  for (current_gene in genes_to_assess) {
    if (!current_gene %in% colnames(specific_mutation_df)) {
      next
    }
  
    specific_mutation_df_mut <- specific_mutation_df %>%
      filter(specific_mutation_df[, current_gene])
  
    specific_mutation_df_wt <- specific_mutation_df %>%
      filter(!specific_mutation_df[, current_gene])
  
    mut_ci_counts_r <- counts_CI_for_frequency_tables(
      specific_mutation_df_mut, "Response", "R")
  
    mut_ci_counts_r <- as.data.frame(
    do.call(rbind, mut_ci_counts_r)) %>%
    mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper),
           Response = "R", Gene = sprintf("%smt", current_gene))
  
    mut_ci_counts_nr <- counts_CI_for_frequency_tables(
      specific_mutation_df_mut, "Response", "NR")
  
      mut_ci_counts_nr <- as.data.frame(
    do.call(rbind, mut_ci_counts_nr)) %>%
    mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper),
           Response = "NR", Gene = sprintf("%smt", current_gene))
  
    wt_ci_counts_r <- counts_CI_for_frequency_tables(
      specific_mutation_df_wt, "Response", "R")
  
    wt_ci_counts_r <- as.data.frame(
    do.call(rbind, wt_ci_counts_r)) %>%
    mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper),
           Response = "R", Gene = sprintf("%swt", current_gene))
  
    wt_ci_counts_nr <- counts_CI_for_frequency_tables(
      specific_mutation_df_wt, "Response", "NR")
  
    wt_ci_counts_nr <- as.data.frame(
    do.call(rbind, wt_ci_counts_nr)) %>%
    mutate(CI_interval = sprintf("%s - %s", CI_lower, CI_upper),
           Response = "NR", Gene = sprintf("%swt", current_gene))
    
    mt_wt_response_ci <- mt_wt_response_ci %>%
      rbind(mut_ci_counts_r, mut_ci_counts_nr, wt_ci_counts_r, wt_ci_counts_nr)
  }

  mt_wt_response_ci <- mt_wt_response_ci %>%
    mutate(Gene_response_n = sprintf("%s (%s)", Gene, Number_of_values),
           Gene_name = substring(Gene, 1, nchar(Gene) - 2)) %>%
    group_by(Gene) %>%
    mutate(Gene_status_n = sprintf("%s (%s)", Gene, sum(Number_of_values)))

  mt_wt_response_ci
}

```

```{r, fig.height = 6, fig.width = 22}

if (filter_and_parse_vcfs) {
  mt_wt_response_ci <- get_wt_mt_response(specific_mutation_df,
                                                genes_to_assess) # genes_to_assess_exp

  mt_wt_response_ci %>%
    write_csv2(here(output_dir, "mt_wt_response_ci.csv"))
  
  mt_wt_response_ci_excl_dmmr <- get_wt_mt_response(specific_mutation_df %>%
                                                      filter(Trial != "NICHE-MSI"),
                                                genes_to_assess) # genes_to_assess_exp

  mt_wt_response_ci_excl_dmmr %>%
    write_csv2(here(output_dir, "mt_wt_response_ci_excl_dmmr.csv"))
}

mt_wt_response_ci <- read_csv2(
  here(output_dir, "mt_wt_response_ci.csv"))

mt_wt_response_ci_excl_dmmr <- read_csv2(
  here(output_dir, "mt_wt_response_ci_excl_dmmr.csv"))

mt_wt_response_ci[1:50, ] %>%
  ggplot(aes(x = Gene_status_n, y = percentage_of_total, fill = Response)) +
  geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(y = "Percentage (95% CI) of \n(non) Responders") +
  scale_fill_manual(values = c(Color_NR, Color_R))

mt_wt_response_ci[1:124, ] %>%
  ggplot(aes(x = Gene_response_n, y = percentage_of_total, fill = Response)) +
  geom_col(position = position_dodge()) +
  geom_errorbar(aes(ymin = CI_lower, ymax = CI_upper),
                position = position_dodge(width = 1), lwd = 0.25, width = 0.5) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(y = "Percentage (95% CI) of \n(non) Responders") +
  scale_fill_manual(values = c(Color_NR, Color_R))

mt_wt_response_ci_delta <- mt_wt_response_ci %>%
  ungroup() %>%
  filter(Response == "R") %>%
  mutate(mut_wt = ifelse(str_detect(Gene, "mt$"), "mutant", "wildtype")) %>%
  select(Response, Number_of_values, percentage_of_total, mut_wt, Gene_name) %>%
  pivot_wider(names_from = mut_wt, values_from = c(Number_of_values, percentage_of_total)) %>%
  filter(Number_of_values_mutant > 10) %>%
  mutate(Response_rate_delta = percentage_of_total_mutant - percentage_of_total_wildtype) %>%
  arrange(Response_rate_delta)

mt_wt_response_ci_excl_dmmr_delta <- mt_wt_response_ci_excl_dmmr %>%
  ungroup() %>%
  filter(Response == "R") %>%
  mutate(mut_wt = ifelse(str_detect(Gene, "mt$"), "mutant", "wildtype")) %>%
  select(Response, Number_of_values, percentage_of_total, mut_wt, Gene_name) %>%
  pivot_wider(names_from = mut_wt, values_from = c(Number_of_values, percentage_of_total)) %>%
  filter(Number_of_values_mutant > 10) %>%
  mutate(Response_rate_delta = percentage_of_total_mutant - percentage_of_total_wildtype) %>%
  arrange(Response_rate_delta)

```

```{r, fig.width = 5, fig.height = 5}
genes_to_assess_exp

for (current_gene in genes_to_assess_exp) {
  if (!current_gene %in% colnames(specific_mutation_df)) {
    next
  }
  
  # Skip genes where a total is below X, not very informative
  check_total <- mt_wt_response_ci %>%
    filter(Gene_name == current_gene) %>%
    group_by(Gene) %>% summarize(Total = sum(Total))
  
  if (any(check_total$Total < 20)) {
    next
  }
  
  cat(current_gene)
  cat("  \n")
  
  plot <- mt_wt_response_ci %>%
    filter(Gene_name == current_gene) %>%
    ggplot(aes(x = Gene_response_n, y = percentage_of_total, fill = Response)) +
  geom_col(position = position_dodge()) +
  geom_errorbar(aes(ymin = CI_lower, ymax = CI_upper),
                position = position_dodge(width = 1), lwd = 0.25, width = 0.5) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(y = "Percentage (95% CI) of \n(non) Responders") +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
    facet_wrap(~Gene, scales = "free")
  print(plot)

  plot <- mt_wt_response_ci %>%
    filter(Gene_name == current_gene) %>%
    ggplot(aes(x = Gene_status_n, y = percentage_of_total, fill = Response)) +
  geom_col() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(x = "Gene", y = "Percentage of \n(non) Responders") +
  scale_fill_manual(values = c(Color_NR, Color_R))
  print(plot)
}
 
```

# ----------

# POLE

```{r POLE}
POLE_mutations <- combined_maf_dp20_vaf_df_specific_f %>%
  filter(Hugo_Symbol == "POLE") 

POLE_mutations %>% write_csv2(here(output_dir, "POLE_mutations.csv"))

POLE_mutations %>% select(all_of(variant_eval_columns)) %>% factor_summary()
POLE_mutations %>% select(all_of(variant_eval_columns)) %>% view()

n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = POLE, y = TMB, col = POLE)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "POLE mutation") +
  expand_limits(y = 0)

n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = POLD1, y = TMB, col = POLD1)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "POLD1 mutation") +
  expand_limits(y = 0)

current_gene <- "POLE"
plot <- mt_wt_response_ci %>%
  filter(Gene_name == current_gene) %>%
  ggplot(aes(x = Gene_response_n, y = percentage_of_total, fill = Response)) +
geom_col(position = position_dodge()) +
geom_errorbar(aes(ymin = CI_lower, ymax = CI_upper),
              position = position_dodge(width = 1), lwd = 0.25, width = 0.5) +
theme_Publication() +
theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
labs(y = "Percentage (95% CI) of \n(non) Responders") +
scale_fill_manual(values = c(Color_NR, Color_R)) +
  facet_wrap(~Gene, scales = "free")
print(plot)

plot <- mt_wt_response_ci %>%
  filter(Gene_name == current_gene) %>%
  ggplot(aes(x = Gene_status_n, y = percentage_of_total, fill = Response)) +
geom_col() +
theme_Publication() +
theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
labs(x = "Gene", y = "Percentage of \n(non) Responders") +
scale_fill_manual(values = c(Color_NR, Color_R))
print(plot)

table(specific_mutation_df$Trial, specific_mutation_df$Response, specific_mutation_df$POLE)

plotProtein(gene = "POLE")

n_variants_dp20_vaf_tmb_status %>% filter(POLE) %>% select(TMB, Trial) %>% factor_summary()

n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = Trial, y = TMB, fill = Trial, group = Trial)) +
  geom_boxplot(outlier.shape = NA) + geom_point(aes(group = Trial)) +
  theme_Publication() + guides(fill = FALSE) + 
  labs(x = "") +
  scale_fill_manual(values = trial_colors) + facet_wrap(~POLE) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))


n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = Trial, y = TMB, fill = Trial, group = Trial)) +
  geom_boxplot(outlier.shape = NA) + geom_point(aes(group = Trial)) +
  theme_Publication() + guides(fill = FALSE) + 
  labs(x = "") +
  scale_fill_manual(values = trial_colors) + facet_wrap(~POLE) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

# POLD1

```{r POLD1}
POLD1_mutations <- combined_maf_dp20_vaf_df_specific_f %>%
  filter(Hugo_Symbol == "POLD1") 

POLD1_mutations %>% write_csv2(here(output_dir, "POLD1_mutations.csv"))
POLD1_mutations %>% select(all_of(variant_eval_columns)) %>% factor_summary()
POLD1_mutations %>% select(all_of(variant_eval_columns)) %>% view()

lollipopPlot(
  maf = combined_maf_specific_f,
  gene = 'POLE',
  AACol = 'HGVSp',
  showMutationRate = TRUE
)

plotProtein(gene = "POLD1")

lollipopPlot(
  maf = combined_maf_specific_f,
  gene = 'POLD1',
  AACol = 'HGVSp',
  showMutationRate = TRUE
)

n_variants_dp20_vaf_tmb_status %>% filter(POLD1) %>% select(TMB, Trial) %>% factor_summary()

n_variants_dp20_vaf_tmb_status %>%
  ggplot(aes(x = Trial, y = TMB, fill = Trial, group = Trial)) +
  geom_boxplot(outlier.shape = NA) + geom_point(aes(group = Trial)) +
  theme_Publication() + guides(fill = FALSE) + 
  labs(x = "") +
  scale_fill_manual(values = trial_colors) + facet_wrap(~POLD1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

# RAS BRAF co-mutations

```{r}
oncoplot(combined_maf_specific_f,
        genes = c("BRAF", "KRAS", "NRAS", "HRAS"),
        clinicalFeatures = c("Response", "Trial"),
         topBarData = "TMB",
         annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

```

```{r, fig.height = 6.5, fig.width = 4}

# no moderate high impact NRAS mutations, only modifier
specific_mutation_m[c("BRAF", "KRAS", "NRAS"), ] %>%
  as.data.frame() %>%
  setNames(c("Gene", "Patient", "Freq")) %>%
  group_by(Gene) %>%
  arrange(-Freq) %>%
  ggplot(aes(y = reorder(Patient, Freq), x = Gene, fill = Freq)) +
  geom_tile() +
  labs(y = "Patient")

table(specific_mutation_df$BRAF)
table(specific_mutation_df$NRAS)
table(specific_mutation_df$KRAS)

table(specific_mutation_df$KRAS, specific_mutation_df$BRAF) %>%
    `colnames<-`(c("BRAFwt", "BRAFmt")) %>%
    `rownames<-`(c("KRASwt", "KRASmt")) 

table(specific_mutation_df$KRAS, specific_mutation_df$BRAF) %>%
    `colnames<-`(c("BRAFwt", "BRAFmt")) %>%
    `rownames<-`(c("KRASwt", "KRASmt")) %>%
  prop.table() %>%
  round(2)

```

# BRAF mutation breakdown

```{r}
lollipopPlot(
  maf = combined_maf_specific_f,
  gene = 'BRAF',
  AACol = 'HGVSp',
  showMutationRate = TRUE
)



BRAF_mutations <- combined_maf_dp20_vaf_df_specific_f %>%
  filter(Hugo_Symbol == "BRAF") 

BRAF_mutations %>% write_csv2(here(output_dir, "BRAF_mutations.csv"))

BRAF_mutations %>% select(all_of(variant_eval_columns)) %>% factor_summary()
BRAF_mutations %>% select(all_of(variant_eval_columns)) %>% view()

# colnames(BRAF_mutations)
table(BRAF_mutations$HGVSp_Short)

BRAF_mutations <- BRAF_mutations %>%
  mutate(BRAF_type = paste0("BRAF", HGVSp_Short)
    )

table(BRAF_mutations$BRAF_type, BRAF_mutations$Trial)
table(BRAF_mutations$BRAF_type, BRAF_mutations$Response)
```

# RAS mutation breakdown

```{r ras breakdown}
lollipopPlot(
  maf = combined_maf_specific_f,
  gene = 'KRAS',
  AACol = 'HGVSp',
  showMutationRate = TRUE
)

KRAS_mutations <- combined_maf_dp20_vaf_df_specific_f %>%
  filter(Hugo_Symbol == "KRAS") 

KRAS_mutations %>% write_csv2(here(output_dir, "KRAS_mutations.csv"))

KRAS_mutations %>% select(Patient, all_of(variant_eval_columns)) %>% factor_summary()
KRAS_mutations %>% select(Trial, Patient, all_of(variant_eval_columns)) %>% view()
# colnames(BRAF_mutations)
table(KRAS_mutations$HGVSp_Short)
table(KRAS_mutations$Trial)

KRAS_mutations <- KRAS_mutations %>%
  mutate(KRAS_type = paste0("KRAS", HGVSp_Short)
    )

# table(KRAS_mutations$KRAS_type, KRAS_mutations$Trial)
table(KRAS_mutations$KRAS_type, KRAS_mutations$Response)

```

# ----------

# B2M

```{r B2M}
lollipopPlot(
  maf = combined_maf_specific_f,
  gene = 'B2M',
  AACol = 'HGVSp',
  showMutationRate = TRUE
)

B2M_mutations <- combined_maf_dp20_vaf_df_specific_f %>%
  filter(Hugo_Symbol == "B2M") 

B2M_mutations %>% write_csv2(here(output_dir, "KRAS_mutations.csv"))

B2M_mutations %>% select(Patient, all_of(variant_eval_columns)) %>% factor_summary()
B2M_mutations %>% select(Trial, Patient, all_of(variant_eval_columns)) %>% view()
table(B2M_mutations$HGVSp_Short)
table(B2M_mutations$Trial)

B2M_mutations <- B2M_mutations %>%
  mutate(B2M_type = paste0("B2M", HGVSp_Short)
    )

# table(KRAS_mutations$KRAS_type, KRAS_mutations$Trial)
table(B2M_mutations$B2M_type, B2M_mutations$Response)
```

# ----------

# Compare mutation proportions

```{r}
mutation_prop_full_TCGA <- read.delim(
  here("data", "external", "cBioportal_TCGA_Pancancer_atlas",
       "Mutated_Genes_Full_10443.txt")) %>%
  mutate(Freq_n = parse_number(Freq)) %>% arrange(-Freq_n)

mutation_prop_matched_TCGA <- read.delim(
  here("data", "external", "cBioportal_TCGA_Pancancer_atlas",
       "Mutated_Genes_Matched_Histology_3054.txt")) %>%
  mutate(Freq_n = parse_number(Freq)) %>% arrange(-Freq_n)

mutation_prop_matched_samstein <- read.delim(
  here("data", "external", "cBioportal_samtein_et_al_2019",
       "Mutated_Genes_Matched_Histology_828.txt")) %>%
  mutate(Freq_n = parse_number(Freq)) %>% arrange(-Freq_n)

mutation_prop_full_TCGA %>%
  select(-c(Gene, Freq)) %>%
  mutate_if(is.character, as.factor) %>%
  summary()

mutation_prop_matched_TCGA %>%
  select(-c(Gene, Freq)) %>%
  mutate_if(is.character, as.factor) %>%
  summary()

mutation_prop_matched_samstein %>%
  select(-c(Gene, Freq)) %>%
  mutate_if(is.character, as.factor) %>%
  summary()

freq_cutoff <- 5
mutation_prop_full_TCGA_f <- mutation_prop_full_TCGA %>%
  filter(Is.Cancer.Gene..source..OncoKB. == "Yes", Freq_n >= freq_cutoff)

mutation_prop_matched_TCGA_f <- mutation_prop_matched_TCGA %>%
  filter(Is.Cancer.Gene..source..OncoKB. == "Yes", Freq_n >= freq_cutoff)

mutation_prop_matched_samstein_f <- mutation_prop_matched_samstein %>%
  filter(Is.Cancer.Gene..source..OncoKB. == "Yes", Freq_n >= freq_cutoff)

mutation_prop_full_TCGA_f %>%
  select(-c(Gene, Freq)) %>%
  mutate_if(is.character, as.factor) %>%
  summary()

mutation_prop_matched_TCGA_f %>%
  select(-c(Gene, Freq)) %>%
  mutate_if(is.character, as.factor) %>%
  summary()

# Genes frequently mutated but not included
mutation_prop_full_TCGA_f$Gene [!mutation_prop_full_TCGA_f$Gene %in% genes_to_assess]
mutation_prop_matched_TCGA_f$Gene [!mutation_prop_matched_TCGA_f$Gene %in% genes_to_assess]

mutation_prop_matched_samstein_f$Gene[!mutation_prop_matched_samstein_f$Gene %in% genes_to_assess]

consider_to_assess <- unique(c(mutation_prop_full_TCGA_f$Gene, mutation_prop_matched_TCGA_f$Gene, mutation_prop_matched_samstein_f$Gene))
consider_to_assess[consider_to_assess %in% genes_to_assess]
consider_to_assess[!consider_to_assess %in% genes_to_assess]

genes_to_assess_exp <- genes_to_assess[
  genes_to_assess %in% colnames(specific_mutation_df)]


specific_mutation_df_freq <- specific_mutation_df %>%
  pivot_longer(cols = c(genes_to_assess_exp),
names_to = "Gene", values_to = "Status") %>%
  filter(Status == 1)

specific_mutation_df_freq <- table(specific_mutation_df_freq$Gene) %>%
  as.data.frame() %>%
setNames(c("Gene", "Freq")) %>%
  mutate(Total = nrow(specific_mutation_df),
         Proportion = round(Freq / Total * 100, 2)) %>%
  arrange(-Proportion)

specific_mutation_df_freq

external_proportions <- mutation_prop_full_TCGA %>%
  select(Gene, Freq_n, Is.Cancer.Gene..source..OncoKB.) %>%
  setNames(c("Gene", "Perc_full_TCGA", "OncoKB_Canger_gene")) %>%
  full_join(
    mutation_prop_matched_TCGA %>%
  select(Gene, Freq_n, Is.Cancer.Gene..source..OncoKB.) %>%
  setNames(c("Gene", "Perc_matched_TCGA", "OncoKB_Canger_gene"))
  ) %>%
  full_join(
    mutation_prop_matched_samstein %>%
  select(Gene, Freq_n, Is.Cancer.Gene..source..OncoKB.) %>%
  setNames(c("Gene", "Perc_matched_samstein", "OncoKB_Canger_gene"))
  )

external_proportions %>%
  write_csv2(here(output_dir, "external_mut_proportions.csv"))

nki_vs_ext_prop <- specific_mutation_df_freq %>%
  setNames(c("Gene", "Freq_NKI", "Total_NKI", "Perc_NKI")) %>%
  left_join(external_proportions)

# TODO compare plots
nki_vs_ext_prop_l <- nki_vs_ext_prop %>%
  pivot_longer(cols = c(Perc_NKI, Perc_full_TCGA, Perc_matched_TCGA,
                        Perc_matched_samstein), values_to = "Perc",
               names_to = "Source") %>%
  mutate(Gene_source = paste0(Gene, "_", Source))

nki_vs_ext_prop_l[1:50, ] %>%
  ggplot(aes(x = Gene, y = Perc,
             fill = Source, group = Source)) +
  geom_col(position = position_dodge2()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  theme_Publication() +
  coord_flip() +
 theme(legend.position = "bottom", legend.direction = "vertical")

nki_vs_ext_prop %>%
  write_csv2(here(output_dir, "nki_vs_ext_prop.csv"))
nki_vs_ext_prop <- read.csv2(here(output_dir, "nki_vs_ext_prop.csv"))

```

# ----------

# Get pathway proportions oncoplot return fractions function

# Oncoplot R NR split

```{r, fig.width = 14, fig.height = 14}
responder_tsb <- combined_maf_specific_f@clinical.data %>% filter(Response == "R") %>% pull(Tumor_Sample_Barcode)
nresponder_tsb <- combined_maf_specific_f@clinical.data %>% filter(Response == "NR") %>% pull(Tumor_Sample_Barcode)

combined_maf_r <- combined_maf_specific_f %>% subsetMaf(tsb = responder_tsb)
combined_maf_nr <- combined_maf_specific_f %>% subsetMaf(tsb = nresponder_tsb)

# GIE Martínez-Jiménez
oncoplot(combined_maf_r, pathways = immune_escape_genes_df %>%
           select(Gene, Pathway.general),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

oncoplot(combined_maf_nr, pathways = immune_escape_genes_df %>%
           select(Gene, Pathway.general),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

# SMG BP
oncoplot(combined_maf_r, pathways = "smgbp", 
         selectedPathways = c("Cell_cycle", "MAPK_signaling", "RTK_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

oncoplot(combined_maf_nr, pathways = "smgbp", 
         selectedPathways = c("Cell_cycle", "MAPK_signaling", "RTK_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

oncoplot(combined_maf_r, pathways = "smgbp", 
         selectedPathways = c("Genome_integrity",
                              "TGFB_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

oncoplot(combined_maf_nr, pathways = "smgbp", 
         selectedPathways = c("Genome_integrity",
                              "TGFB_signaling"),
         gene_mar = 8, fontSize = 0.6,
         clinicalFeatures = c("Trial", "Response"),
         annotationColor = annotation_color_list,
         sortByAnnotation = TRUE,
         topBarData = "TMB")

```

# Oncoplot functions return and export fractions

```{r}
custom_pathway_load <- function(maf, pathways = NULL) {
  if (is.null(pathways)) {
    pathdb <- system.file("extdata", "BP_SMGs.txt.gz", package = "maftools")
    pathdb = data.table::fread(input = pathdb, skip = "Gene")
    pathdb = pathdb[!duplicated(Gene)][,.(Gene, Pathway)]
  }
  else {
    pathdb = data.table::setDT(pathways) # convert to data table, 
    # adds internal pointer attribute used for rest of the function
    pathdb = pathdb[!duplicated(Gene)][,.(Gene, Pathway)]

    }
    
  pathdb$Pathway = gsub(pattern = " ", replacement = "_", x = pathdb$Pathway)
  pathdb$Pathway = gsub(pattern = "/", replacement = "_", x = pathdb$Pathway)

  pathdb_size = pathdb[,.N,Pathway]
  pathdb = split(pathdb, as.factor(pathdb$Pathway))

  altered_pws = lapply(pathdb, function(pw){
    x = suppressMessages(try(genesToBarcodes(maf = maf, genes = pw$Gene)))
    if(class(x) == "try-error"){
      pw_genes = NULL
    }else{
      pw_genes = names(genesToBarcodes(maf = maf, genes = pw$Gene, justNames = TRUE, verbose = FALSE))
    }
    pw_genes
  })

  mut_load = lapply(altered_pws, function(x){
    if(is.null(x)){
      nsamps =  0
    }else{
      nsamps = length(unique(as.character(unlist(
        genesToBarcodes(
          maf = maf,
          genes = x,
          justNames = TRUE
        )
      ))))
    }
    nsamps
  })

  altered_pws = as.data.frame(t(data.frame(lapply(altered_pws, length))))
  data.table::setDT(x = altered_pws, keep.rownames = TRUE)
  colnames(altered_pws) = c("Pathway", "n_affected_genes")
  altered_pws$Pathway = gsub(pattern = "\\.", replacement = "-", x = altered_pws$Pathway)
  altered_pws = merge(pathdb_size, altered_pws, all.x = TRUE)

  altered_pws[, fraction_affected := n_affected_genes/N]
  altered_pws$Mutated_samples = unlist(mut_load)
  nsamps = as.numeric(maf@summary[ID == "Samples", summary])
  altered_pws[,Fraction_mutated_samples := Mutated_samples/nsamps]
  altered_pws = altered_pws[order(Fraction_mutated_samples, fraction_affected, decreasing = TRUE)]

  altered_pws = altered_pws[!n_affected_genes %in% 0]
  altered_pws
  }
 
# custom_pathway_load(combined_maf_dp20_vaf)
# custom_pathway_load(combined_maf_dp20_vaf, pathways = oncoplot_custom_pathways)
```

```{r}

oncoplot_return_fractions <- function(maf,
  top = 20,
  minMut = NULL,
  genes = NULL,
  altered = FALSE,
  drawRowBar = TRUE,
  drawColBar = TRUE,
  leftBarData = NULL,
  leftBarLims = NULL,
  rightBarData = NULL,
  rightBarLims = NULL,
  topBarData = NULL,
  logColBar = FALSE,
  includeColBarCN = TRUE,
  clinicalFeatures = NULL,
  annotationColor = NULL,
  annotationDat = NULL,
  pathways = NULL,
  selectedPathways = NULL,
  draw_titv = FALSE,
  showTumorSampleBarcodes = FALSE,
  barcode_mar = 4,
  barcodeSrt = 90,
  gene_mar = 5,
  anno_height = 1,
  legend_height = 4,
  sortByAnnotation = FALSE,
  groupAnnotationBySize = TRUE,
  annotationOrder = NULL,
  sortByMutation = FALSE,
  keepGeneOrder = FALSE,
  GeneOrderSort = TRUE,
  sampleOrder = NULL,
  additionalFeature = NULL,
  additionalFeaturePch = 20,
  additionalFeatureCol = "gray70",
  additionalFeatureCex = 0.9,
  genesToIgnore = NULL,
  removeNonMutated = TRUE,
  fill = TRUE,
  cohortSize = NULL,
  colors = NULL,
  bgCol = "#CCCCCC",
  borderCol = "white",
  annoBorderCol = NA,
  numericAnnoCol = NULL,
  drawBox = FALSE,
  fontSize = 0.8,
  SampleNamefontSize = 1,
  titleFontSize = 1.5,
  legendFontSize = 1.2,
  annotationFontSize = 1.2,
  sepwd_genes = 0.5,
  sepwd_samples = 0.25,
  writeMatrix = FALSE,
  colbar_pathway = FALSE,
  showTitle = TRUE,
  titleText = NULL,
  cBioPortal = FALSE,
  path_order = NULL
) {
  {
    if (is.null(cohortSize)) {
        totSamps = as.numeric(maf@summary[3, summary])
    }
    else {
        totSamps = cohortSize
    }
    if (!is.null(genes)) {
        om = createOncoMatrix(m = maf, g = genes, add_missing = fill, 
            cbio = cBioPortal)
        numMat = om$numericMatrix
        mat_origin = om$oncoMatrix
        if (!is.null(pathways)) {
            stop("Please use genes or pathways.")
        }
    }
    else if (!is.null(pathways)) {
        if (is(pathways, "data.frame")) {
            pathwayLoad = custom_pathway_load(maf = maf, pathways = pathways)
            pathways = data.table::copy(pathways)
            colnames(pathways)[1:2] = c("Gene", "Pathway")
            data.table::setDT(x = pathways)
            pathways = pathways[!duplicated(Gene)][, .(Gene, 
                Pathway)]
            if (!is.null(selectedPathways)) {
                pathways = pathways[Pathway %in% selectedPathways]
            }
            

        }
        else if (file.exists(pathways)) {
            pathways = data.table::fread(file = pathways)
            colnames(pathways)[1:2] = c("Gene", "Pathway")
            data.table::setDT(x = pathways)
            pathways = pathways[!duplicated(Gene)][, .(Gene, 
                Pathway)]
            if (!is.null(selectedPathways)) {
                pathways = pathways[Pathway %in% selectedPathways]
            }
        }
        else {
            pathways = system.file("extdata", "BP_SMGs.txt.gz", 
                package = "maftools")
            pathways = data.table::fread(file = pathways, skip = "Gene")
            pathways = pathways[!duplicated(Gene)][, .(Gene, 
                Pathway)]
            pathways$Pathway = gsub(pattern = " ", replacement = "_", 
                x = pathways$Pathway)
            pathwayLoad = pathway_load(maf = maf)
            if (is.null(selectedPathways)) {
                # message("Drawing upto top 3 mutated pathways")
                print(pathwayLoad)
                # if (ncol(pathwayLoad) >= 3) {
                #   pathways = pathways[Pathway %in% pathwayLoad[1:3, 
                #     Pathway]]
                # }
                # else {
                #   pathways = pathways[Pathway %in% pathwayLoad[, 
                #     Pathway]]
                # }
              # 
            }
            else {
                pathways = pathways[Pathway %in% selectedPathways]
            }
        }
        genes = as.character(pathways$Gene)
        om = createOncoMatrix(m = maf, g = genes, cbio = cBioPortal)
        numMat = om$numericMatrix
        mat_origin = om$oncoMatrix
        if (length(genes[!genes %in% rownames(numMat)]) > 0) {
            genes = genes[genes %in% rownames(numMat)]
        }
    }
    else if (!is.null(minMut)) {
        if (altered) {
            genes = getGeneSummary(x = maf)[order(AlteredSamples, 
                decreasing = TRUE)][, .(Hugo_Symbol, AlteredSamples)]
        }
        else {
            genes = getGeneSummary(x = maf)[order(MutatedSamples, 
                decreasing = TRUE)][, .(Hugo_Symbol, MutatedSamples)]
        }
        colnames(genes)[2] = "mutload"
        genes[, `:=`(fractMutated, mutload/totSamps)]
        if (minMut > 1) {
            genes = genes[mutload >= minMut, Hugo_Symbol]
        }
        else {
            genes = genes[fractMutated >= minMut, Hugo_Symbol]
        }
        om = createOncoMatrix(m = maf, g = genes, cbio = cBioPortal)
        numMat = om$numericMatrix
        mat_origin = om$oncoMatrix
    }
    else {
        if (altered) {
            genes = getGeneSummary(x = maf)[order(AlteredSamples, 
                decreasing = TRUE)][1:top, Hugo_Symbol]
        }
        else {
            genes = getGeneSummary(x = maf)[1:top, Hugo_Symbol]
        }
        om = createOncoMatrix(m = maf, g = genes, cbio = cBioPortal)
        numMat = om$numericMatrix
        mat_origin = om$oncoMatrix
    }
    if (!is.null(genesToIgnore)) {
        numMat = numMat[!rownames(numMat) %in% genesToIgnore, 
            ]
        mat_origin = mat_origin[!rownames(mat_origin) %in% genesToIgnore, 
            ]
    }
    tsbs = levels(getSampleSummary(x = maf)[, Tumor_Sample_Barcode])
    if (!removeNonMutated) {
        tsb.include = matrix(data = 0, nrow = nrow(numMat),
                             ncol = length(tsbs[!tsbs %in% 
            colnames(numMat)]))
        colnames(tsb.include) = tsbs[!tsbs %in% colnames(numMat)]
        rownames(tsb.include) = rownames(numMat)
        numMat = cbind(numMat, tsb.include)
        mat_origin = cbind(mat_origin, tsb.include)
    }
    if (keepGeneOrder) {
        if (fill) {
            numMat = numMat[genes, , drop = FALSE]
        }
        else {
            if (GeneOrderSort) {
                numMat = sortByGeneOrder(m = numMat, g = genes)
            }
            else {
                numMat = numMat[genes, , drop = FALSE]
            }
        }
        mat = mat_origin[rownames(numMat), , drop = FALSE]
        mat = mat[, colnames(numMat), drop = FALSE]
    }
    if (sortByMutation) {
        numMat_temp = sortByMutation(numMat = numMat, maf = maf)
        numMat = numMat[rownames(numMat_temp), colnames(numMat_temp), 
            drop = FALSE]
    }
    samp_sum = data.table::copy(getSampleSummary(x = maf))
    samp_sum[, `:=`(total, NULL)]
    if (cBioPortal) {
        samp_sum = data.table::melt(data = samp_sum, id.vars = "Tumor_Sample_Barcode")
        colnames(samp_sum)[2] = "Variant_Classification"
        vc = c("Nonstop_Mutation", "Frame_Shift_Del", "Missense_Mutation", 
            "Nonsense_Mutation", "Splice_Site", "Frame_Shift_Ins", 
            "In_Frame_Del", "In_Frame_Ins")
        vc.cbio = c("Truncating", "Truncating", "Missense", "Truncating", 
            "Truncating", "Truncating", "In-frame", "In-frame")
        names(vc.cbio) = vc
        samp_sum[, `:=`(Variant_Classification_temp, vc.cbio[as.character(samp_sum$Variant_Classification)])]
        samp_sum$Variant_Classification_temp = ifelse(test = is.na(samp_sum$Variant_Classification_temp), 
            yes = as.character(samp_sum$Variant_Classification), 
            no = samp_sum$Variant_Classification_temp)
        samp_sum[, `:=`(Variant_Classification, as.factor(as.character(Variant_Classification_temp)))]
        samp_sum[, `:=`(Variant_Classification_temp, NULL)]
        samp_sum = samp_sum[, sum(value), .(Tumor_Sample_Barcode, 
            Variant_Classification)]
        samp_sum = data.table::dcast(data = samp_sum, formula = Tumor_Sample_Barcode ~ 
            Variant_Classification, value.var = "V1")
    }
    if ("CNV_total" %in% colnames(samp_sum)) {
        samp_sum[, `:=`(CNV_total, NULL)]
    }
    if (!includeColBarCN) {
        suppressWarnings(samp_sum[, `:=`(Amp, NULL)])
        suppressWarnings(samp_sum[, `:=`(Del, NULL)])
    }
    data.table::setDF(x = samp_sum, rownames = as.character(samp_sum$Tumor_Sample_Barcode))
    samp_sum = samp_sum[, -1, drop = FALSE]
    if (!is.null(clinicalFeatures)) {
        if (is.null(annotationDat)) {
            annotation = parse_annotation_dat(annotationDat = maf, 
                clinicalFeatures = clinicalFeatures)
        }
        else {
            annotation = parse_annotation_dat(annotationDat = annotationDat, 
                clinicalFeatures = clinicalFeatures)
        }
        annotation = annotation[colnames(numMat), , drop = FALSE]
        if (sortByAnnotation) {
            numMat = sortByAnnotation(numMat = numMat, maf = maf, 
                anno = annotation, annoOrder = annotationOrder, 
                group = groupAnnotationBySize, isNumeric = FALSE)
        }
    }
   
    gene_sum = apply(numMat, 1, function(x) length(x[x != 0]))
    percent_alt = paste0(round(100 * (apply(numMat, 1, function(x) length(x[x != 
        0]))/totSamps)), "%")
    nonmut_samps = colnames(numMat)[!colnames(numMat) %in% rownames(samp_sum)]
    if (length(nonmut_samps) > 0) {
        samp_sum_missing_samps = matrix(data = 0, nrow = length(nonmut_samps), 
            ncol = ncol(samp_sum))
        colnames(samp_sum_missing_samps) = colnames(samp_sum)
        rownames(samp_sum_missing_samps) = nonmut_samps
        samp_sum = rbind(samp_sum, samp_sum_missing_samps)
    }
    
    if (!is.null(pathways)) {
        # vc_codes = c(vc_codes, `99` = "pathway")
        temp_dat = data.table::data.table(Gene = rownames(numMat), 
            percent_alt)
        temp_dat = merge(temp_dat, pathways, all.x = TRUE)
        temp_dat[is.na(temp_dat)] = "Unknown"
        temp_dat$pct_alt = as.numeric(gsub(pattern = "%", replacement = "", 
            x = temp_dat$percent_alt))
        temp_dat = split(temp_dat, as.factor(as.character(temp_dat$Pathway)))
        temp_dat_freq = lapply(temp_dat, function(x) {
            data.table::data.table(n = nrow(x), max_alt = max(x$pct_alt))
        })
        temp_dat_freq = data.table::rbindlist(l = temp_dat_freq, 
            use.names = TRUE, fill = TRUE, idcol = "Pathway")
        temp_dat_ord = temp_dat_freq[order(max_alt, n, decreasing = TRUE)][, 
            Pathway]
        temp_dat = lapply(temp_dat, function(x) {
            x[order(pct_alt, decreasing = TRUE)]
        })[temp_dat_ord]
        if ("Unknown" %in% names(temp_dat)) {
            temp_dat_ord = c(grep(pattern = "Unknown", x = names(temp_dat), 
                invert = TRUE, value = TRUE), "Unknown")
            temp_dat = temp_dat[temp_dat_ord]
        }
        if (!is.null(path_order)) {
            path_order = intersect(x = path_order, names(temp_dat))
            temp_dat = temp_dat[path_order]
            if (length(temp_dat) == 0) {
                stop("None of the pathways match the provided order!")
            }
        }
        nm = lapply(seq_along(temp_dat), function(i) {
            x = numMat[temp_dat[[i]]$Gene, , drop = FALSE]
            x = rbind(x, apply(x, 2, function(y) ifelse(test = sum(y) == 
                0, yes = 0, no = 99)))
            rownames(x)[nrow(x)] = names(temp_dat)[i]
            x
        })
        numMat = do.call(rbind, nm)
        mat_origin_path = rownames(numMat)[!rownames(numMat) %in% 
            rownames(mat_origin)]
        mat_origin_path = numMat[mat_origin_path, , drop = FALSE]
        mat_origin_path[mat_origin_path == 0] = ""
        mat_origin_path[mat_origin_path == "99"] = "pathway"
        mat_origin_path = mat_origin_path[, colnames(mat_origin), 
            drop = FALSE]
        mat_origin = rbind(mat_origin, mat_origin_path)
        mat_origin = mat_origin[rownames(numMat), colnames(numMat), 
            drop = FALSE]
    }
   
    # return(invisible(rownames(nm)))
    return(list(pathwayLoad, temp_dat))
}
}

try_get_fraction_res <- function(maf, pathway) {
  tryCatch(
        {
          fraction_res <- oncoplot_return_fractions(maf, pathways = pathway)
          fraction_res
        },
        error = function(cond) {
          message("Here's the original error message:")
          message(conditionMessage(cond))
          return(NULL)
        }
  )
}

export_pathway_proportions <- function(maf, pathway_dir = "pathway_proportions", pathways = NULL, file_suffix = "", Group_name = "") {
  
  if (is.null(pathways)) {
    pathways = "smgbp"
    # These are default from oncoplot, and each gene is only present in 1 pathway
    fraction_res <- oncoplot_return_fractions(maf, pathways = pathways)
    pathway_proportions  <- fraction_res[[1]] %>% mutate(Group = Group_name)
    pathway_gene_proportions  <- do.call(rbind, fraction_res[[2]]) %>% mutate(Group = Group_name)
    } else {
      
    pathway_proportions <- data.frame(matrix(nrow = 0, ncol = 0))
    pathway_gene_proportions <- data.frame(matrix(nrow = 0, ncol = 0))

    # Needs to do this for each pathway at a time, genes can only be assigned to 1 pathway, not multiple
    for (current_pathway in unique(pathways$Pathway)) {
      # print(current_pathway)
      fraction_res <- try_get_fraction_res(maf, pathway = pathways %>% filter(Pathway == current_pathway))
      
      if (is.null(fraction_res)) {
        custom_proportion_df = data.frame(Pathway = current_pathway, N = NA, n_affected_genes = 0, fraction_affected = 0, Mutated_samples = 0, Fraction_mutated_samples = 0)
          pathway_proportions <- pathway_proportions %>% rbind(custom_proportion_df)
          custom_gene_proportion_df <- data.frame(Gene = "NA", percent_alt = "0%", Pathway = current_pathway, pct_alt = 0)
          pathway_gene_proportions <- pathway_gene_proportions %>% rbind(custom_gene_proportion_df)
      } else {
        pathway_proportions  <- pathway_proportions %>% rbind(fraction_res[[1]])
        pathway_gene_proportions  <- pathway_gene_proportions %>% rbind(do.call(rbind, fraction_res[[2]]))
        }
      }
    
    pathway_proportions <- pathway_proportions %>% mutate(Group = Group_name)
    pathway_gene_proportions <- pathway_gene_proportions %>% mutate(Group = Group_name)

    }
  
  pathway_proportions %>%
    write_csv2(here(output_dir, pathway_dir, sprintf("pathway_proportions_%s.csv", file_suffix)))
  pathway_gene_proportions %>%
    write_csv2(here(output_dir, pathway_dir, sprintf("pathway_gene_proportions_%s.csv", file_suffix)))
}


get_pathway_proportion_trial_mafs <- function(
    maf,
    pathways = NULL,
    pathway_dir = "pathway_proportions",
    trial_list = c("Combined", "BELLINI", "IMCISION", "MATISSE",
                   "NABUCCO", "NICHE-MSS", "OpACIN_neo", "PRADO"),
    dna_metadata = dna_metadata_f) {
  
  for (current_trial in trial_list) {
    current_maf = maf
    
    print(current_trial)
    if (current_trial != "Combined") {
        filtered_metadata <- dna_metadata %>% filter(Trial == current_trial)
        current_maf <- subsetMaf(maf = current_maf,
                            tsb = filtered_metadata$Tumor_Sample_Barcode)
        }
    
    # General mutation frequencies
    export_pathway_proportions(maf = current_maf, pathway_dir = pathway_dir, pathways = pathways,
                               file_suffix = current_trial,
                               Group_name = current_trial)
    
    # R x NR splits

    responder_tsb <- current_maf@clinical.data %>% filter(Response == "R") %>% pull(Tumor_Sample_Barcode)
    nresponder_tsb <- current_maf@clinical.data %>% filter(Response == "NR") %>% pull(Tumor_Sample_Barcode)
    
    print(current_trial, "R")

    current_maf_r <- current_maf %>% subsetMaf(tsb = responder_tsb)

    export_pathway_proportions(maf = current_maf_r,
                               pathway_dir = pathway_dir,
                               pathways = pathways,
                               file_suffix = sprintf("%s_R", current_trial),
                               Group_name = sprintf("%s_R", current_trial))
    print(current_trial, "NR")

    current_maf_nr <- current_maf %>% subsetMaf(tsb = nresponder_tsb)

    export_pathway_proportions(maf = current_maf_nr,
                               pathway_dir = pathway_dir,
                               pathways = pathways,
                               file_suffix = sprintf("%s_NR", current_trial),
                               Group_name = sprintf("%s_NR", current_trial))
    }
}

```

# Parse and export combined files

```{r}
parse_pathway_proportions <- function(pathway_dir) {
  # Combine data from all these files
  pathway_data_combined <- data.frame(matrix(nrow = 0, ncol = 0))
  pathway_gene_data_combined <- data.frame(matrix(nrow = 0, ncol = 0))
  
  pathway_dir_files <- list.files(pathway_dir, full.names = TRUE)
  
  pathway_dir_files <- pathway_dir_files[!pathway_dir_files %in% c(here(pathway_dir, "pathway_gene_data_combined.csv"),here(pathway_dir, "pathway_data_combined.csv"))]
  
  pathway_proportion_files <- pathway_dir_files[!str_detect(pathway_dir_files, "pathway_gene_proportions")]
  pathway_gene_proportion_files <- pathway_dir_files[str_detect(pathway_dir_files, "pathway_gene_proportions")]
  
  for (current_file in pathway_proportion_files){
    # print(current_file)
    current_pathway_proportion <- read.csv2(current_file)
    pathway_data_combined = pathway_data_combined %>% rbind(current_pathway_proportion)
    
  }
  
  for (current_file in pathway_gene_proportion_files){
    # print(current_file)
    current_pathway_gene_proportion <- read.csv2(current_file)
    pathway_gene_data_combined = pathway_gene_data_combined %>% rbind(current_pathway_gene_proportion)
    
  }

  pathway_data_combined %>% write_csv2(here(pathway_dir, "pathway_data_combined.csv"))
  pathway_gene_data_combined %>% write_csv2(here(pathway_dir, "pathway_gene_data_combined.csv"))

  
}

```

# ----------

# Pathway proportions

```{r}
msigdb_pathways <- readRDS(here("data", "processed", "PanCancer_pathway_mutations", "all_pathways", "msigdb_all_pathways.rds"))
length(msigdb_pathways)
head(names(msigdb_pathways), 100)

msigdb_pathways_f <- msigdb_pathways # [c("")]
msigdb_pathways_f <- stack(msigdb_pathways_f) %>%
  setNames(c("Gene", "Pathway"))

# Analysis Petros signatures
list.files(here("data", "processed", "PanCancer_pathway_mutations", "all_pathways"))

non_response_pathways <- read_csv(here("data", "processed", "PanCancer_pathway_mutations", "Non_responders", "Non_response_pathways__stats.csv"))

non_response_pathways$Pathway %in% names(msigdb_pathways)

oncoplot_nr_pathways <- stack(msigdb_pathways[non_response_pathways$Pathway]) %>%
  setNames(c("Gene", "Pathway"))

table(oncoplot_nr_pathways$Pathway) %>% as.data.frame() %>% arrange(-Freq)
table(oncoplot_nr_pathways$Pathway) %>% as.data.frame() %>% summary()

# oncoplot_nr_pathways %>% select(Pathway) %>% unique() %>% write_csv2(here(output_dir, "oncoplot_nr_pathways_map.csv"))
oncoplot_nr_pathways_map <- read_csv2(here(output_dir, "oncoplot_nr_pathways_map.csv"))

oncoplot_nr_pathways <- oncoplot_nr_pathways %>% left_join(oncoplot_nr_pathways_map)
oncoplot_nr_pathways %>% write_csv2(here(output_dir, "oncoplot_nr_pathways.csv"))

# SMG
# Genome_integrity, MAPK_signaling, Cell_cycle, Chromatin_other,
# Chromatin_SWI/SNF_complex, Transcription_factor, RNA_abundance
# Chromatin_histone_modifiers, RTK_signaling
# Wnt/B-catening_signaling PI3K_signaling
# NOTCH_signaling Protein_homeostatis/ubiquitination
# Histone_modification  TGFB_signaling
# TOR_signaling Metabolism Immune_signaling
# Splicing Epigenetics_DNA_modifiers
#  NFKB_signaling Apoptosis

```

# All pathways

```{r}
if (filter_and_parse_vcfs) {
    
  dir.create(here(output_dir, "pathway_proportions_msigdb"))
  
  get_pathway_proportion_trial_mafs(combined_maf_dp20_vaf,
                                    pathway_dir = "pathway_proportions_msigdb",
                                    pathways = msigdb_pathways_f)
  
  # Get NICHE-MSI separately, breaks with NR = 0
  # For now ignored, function already outputs the total and R frequencies before
  get_pathway_proportion_trial_mafs(combined_maf_dp20_vaf,
                                    pathway_dir = "pathway_proportions_msigdb",
                                    trial_list = c("NICHE-MSI"),
                                    pathways = msigdb_pathways_f)
  
  # Combine data from all these files
  parse_pathway_proportions(here(output_dir, "pathway_proportions_msigdb"))
}

pathway_data_msigdb_combined <- read.csv2(here(output_dir, "pathway_proportions_msigdb", "pathway_data_combined.csv")) %>% left_join(trial_group_map)
pathway_gene_data_msigdb_combined <- read.csv2(here(output_dir, "pathway_proportions_msigdb", "pathway_gene_data_combined.csv")) %>% left_join(trial_group_map)

pathway_data_msigdb_combined_trial_prop <- pathway_data_msigdb_combined %>%
  filter(!str_detect(Group, "_R|_NR"))

pathway_data_msigdb_combined_trial_prop %>%
  ggplot(aes(x = Fraction_mutated_samples, fill = Trial, alpha = 0.5)) +
  geom_density() + theme_Publication()

pathway_data_msigdb_combined_trial_prop %>%
  ggplot(aes(x = Fraction_mutated_samples, fill = Trial, alpha = 0.5)) +
  geom_density() + theme_Publication() + facet_wrap(~Trial)
```

# Number of gene sets over mutation threshold

```{r}
check_proportion_thresholds <- function(df, sequence = seq(0, 1, by = 0.01)) {
  result <- data.frame(threshold = numeric(), num_rows = numeric())
  
  for (threshold in sequence) {
    num_rows <- sum(apply(df, 1, function(row) all(row > threshold)))
    result <- rbind(result, data.frame(threshold = threshold, num_rows = num_rows))
  }
  
  return(result)
}

pathway_data_msigdb_combined_trial_prop_wide <- pathway_data_msigdb_combined_trial_prop %>%
  filter(Trial != "Combined") %>%
  select(Pathway, Trial, Fraction_mutated_samples) %>%
  pivot_wider(names_from = Trial, values_from = Fraction_mutated_samples)

pathway_data_msigdb_combined_trial_prop_wide[is.na(pathway_data_msigdb_combined_trial_prop_wide)] <- 0

pathway_proportion_thresholds_broad <- check_proportion_thresholds(
  df = pathway_data_msigdb_combined_trial_prop_wide,
  sequence = seq(0, 1, by = 0.1))
pathway_proportion_thresholds_broad

pathway_proportion_thresholds_broad %>%
  filter(threshold > 0) %>%
  ggplot(aes(x = threshold, y = num_rows)) +
  geom_point() + theme_Publication() +
  labs(x = "Proportion threshold", y = "Number of pathways")

mutation_proportions <- top_genes_n_samples %>%
    select("Gene", starts_with("Prop")) %>%
    select(-Prop_all) %>% column_to_rownames("Gene")


pathway_data_msigdb_combined %>%
  filter(!str_detect(Group, "_R|_NR")) %>%
  filter(Fraction_mutated_samples > 0.33) %>%
  pull(Trial) %>%
  table()

pathway_data_msigdb_combined %>%
  filter(!str_detect(Group, "_R|_NR")) %>%
  filter(fraction_affected > 0.2) %>%
  filter(Fraction_mutated_samples > 0.33) %>%
  pull(Trial) %>%
  table()

pathway_data_msigdb_combined %>%
    filter(!str_detect(Group, "_R|_NR")) %>%
    # filter(fraction_affected > 0.2) %>%
    filter(Fraction_mutated_samples > 0.33) %>% filter(Trial == "BELLINI")

minimum_mutation_threshold <- 0.33
# minimum_gene_affected_threshold <- 0.2
pathway_data_msigdb_combined_trial_prop_wide[rowSums(pathway_data_msigdb_combined_trial_prop_wide > minimum_mutation_threshold) == ncol(pathway_data_msigdb_combined_trial_prop_wide), ]

pathways_above_threshold <- pathway_data_msigdb_combined_trial_prop_wide[
  rowSums(pathway_data_msigdb_combined_trial_prop_wide > minimum_mutation_threshold) == ncol(pathway_data_msigdb_combined_trial_prop_wide), ]

```

```{r}
pathway_data_msigdb_response <- pathway_data_msigdb_combined %>%
  filter(str_detect(Group, "_R|_NR")) %>% 
  select(-c(n_affected_genes, fraction_affected, Group, Mutated_samples)) %>%
  pivot_wider(names_from = Response, values_from = Fraction_mutated_samples) %>%
  mutate(Response_delta = R - NR)
  
pathway_data_msigdb_response %>% select(R, NR, Trial, Response_delta) %>% factor_summary()

pathway_data_msigdb_response %>% ggplot(aes(x = Response_delta, fill = Trial, alpha = 0.5)) + geom_density() + theme_Publication()
pathway_data_msigdb_response %>%
  filter(!Trial == "NICHE-MSI") %>%
  ggplot(aes(x = Response_delta, fill = Trial, alpha = 0.5)) +
  geom_density() + theme_Publication()

pathway_data_msigdb_response %>%
  filter(!Trial == "NICHE-MSI") %>%
  filter(Pathway %in% pathways_above_threshold$Pathway) %>%
  ggplot(aes(x = Response_delta, fill = Trial, alpha = 0.5)) +
  geom_density() + theme_Publication()

pathway_data_msigdb_response %>%
  filter(!Trial == "NICHE-MSI")  %>%
  filter(Response_delta < -0.1) %>% view()

pathway_data_msigdb_response %>%
  filter(!Trial == "NICHE-MSI") %>%
  filter(Pathway %in% pathways_above_threshold$Pathway) %>%
  filter(Response_delta < -0.1)
```

# Get pathway proportions smgbp

```{r, fig.height = 12, fig.width = 14}
# Careful! Oncoplot pathways is using and returning smgbp pathways, even if you indicate pathways = "sigpw"
# The name of pathways is matching smgbp instead of sigpw... this cannot be used reliably and caused a lot of confusion before
# loaded maftools_sigpw_pathways separately

# This does not work, although the pathway list should explicitly be "sigpw", it's using the name from smgbp, "RTK_signaling" (!)

dir.create(here(output_dir, "pathway_proportions_smgbp"))
get_pathway_proportion_trial_mafs(combined_maf_specific_f,
                                  pathway_dir = "pathway_proportions_smgbp",
                                  pathways = "smgbp")

# Get NICHE-MSI separately, breaks with NR = 0
# For now ignored, function already outputs the total and R frequencies before
get_pathway_proportion_trial_mafs(combined_maf_specific_f,
                                  pathway_dir = "pathway_proportions_smgbp",
                                  trial_list = c("NICHE-MSI"), pathways = "smgbp")

parse_pathway_proportions(here(output_dir, "pathway_proportions_smgbp"))

pathway_smgbp_data_combined <- read.csv2(here(output_dir, "pathway_proportions_smgbp", "pathway_data_combined.csv"))%>% left_join(trial_group_map)
pathway_smgbp_gene_data_combined <- read.csv2(here(output_dir, "pathway_proportions_smgbp", "pathway_gene_data_combined.csv")) %>% left_join(trial_group_map)

pathway_smgbp_data_combined %>% filter(Trial == "MATISSE")
pathway_smgbp_gene_data_combined %>% filter(Trial == "MATISSE") %>% filter(Pathway == "Cell_cycle")

pathway_smgbp_gene_data_combined$Gene[!pathway_smgbp_gene_data_combined$Gene %in% oncoKB_cancerGeneList$Hugo.Symbol]
pathway_smgbp_gene_data_combined$Gene[!pathway_smgbp_gene_data_combined$Gene %in% genes_to_assess]


pathway_smgbp_data_combined %>%
  # mutate(Pathway = ifelse(Pathway == "Protein_homeostasis_ubiquitination", "PTN Homeost. Ubiquitin.", Pathway)) %>%
  filter(!str_detect(Group, "_R|_NR")) %>%
  filter(! Pathway %in% c("Other", "Other_signaling")) %>%
  ggplot(aes(x = fct_rev(Group), y = Fraction_mutated_samples, fill = Group)) + geom_col() +
  facet_wrap(~Pathway) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() + scale_fill_manual(values = trial_colors_combined) +
  guides(fill = FALSE) + labs(x = "", title = "SMGBP pathways")
```

# Get pathway proportions sigpw

```{r, fig.height = 12, fig.width = 14}
# Careful! Oncoplot pathways is using and returning smgbp pathways, even if you indicate pathways = "sigpw"
# The name of pathways is matching smgbp instead of sigpw... this cannot be used reliably and caused a lot of confusion before
# loaded maftools_sigpw_pathways separately

# This does not work, although the pathway list should explicitly be "sigpw", it's using the name from smgbp, "RTK_signaling" (!)
# get_pathway_proportion_trial_mafs(combined_maf_specific_f,
#                                   pathway_dir = "pathway_proportions_sigpw",
#                                   pathways = "sigpw")

dir.create(here(output_dir, "pathway_proportions_sigpw"))
get_pathway_proportion_trial_mafs(combined_maf_specific_f,
                                  pathway_dir = "pathway_proportions_sigpw",
                                  pathways = maftools_sigpw_pathways %>%
                                    select(Gene, Pathway))

# Get NICHE-MSI separately, breaks with NR = 0
# For now ignored, function already outputs the total and R frequencies before
get_pathway_proportion_trial_mafs(combined_maf_specific_f,
                                  pathway_dir = "pathway_proportions_sigpw",
                                  trial_list = c("NICHE-MSI"),
                                  pathways = maftools_sigpw_pathways %>%
                                    select(Gene, Pathway))

parse_pathway_proportions(here(output_dir, "pathway_proportions_sigpw"))


pathway_sigpw_gene_data_combined <- read.csv2(here(output_dir, "pathway_proportions_sigpw", "pathway_gene_data_combined.csv")) %>% left_join(trial_group_map)
pathway_sigpw_data_combined <- read.csv2(here(output_dir, "pathway_proportions_sigpw", "pathway_data_combined.csv"))%>% left_join(trial_group_map)

pathway_sigpw_data_combined %>% filter(Trial == "MATISSE")
pathway_sigpw_gene_data_combined %>% filter(Trial == "MATISSE") %>% filter(Pathway == "Cell_Cycle")

# pathway_sigpw_gene_data_combined %>%
#   select(Gene, Pathway) %>%
#   unique() %>%
#   write_csv2(here(output_dir, "sigpw_gene_list.csv"))

maftools_sigpw_pathways$Gene[!maftools_sigpw_pathways$Gene %in% oncoKB_cancerGeneList$Hugo.Symbol]
maftools_sigpw_pathways$Gene[!maftools_sigpw_pathways$Gene %in% genes_to_assess]

pathway_data_combined$Group <- factor(pathway_data_combined$Group,
                                      levels = c(trial_list, sprintf("%s_R", trial_list), sprintf("%s_NR", trial_list)))
pathway_data_combined <- pathway_data_combined %>% left_join(trial_group_map)

pathway_data_combined$Trial <- factor(pathway_data_combined$Trial,
                                      levels = trial_list)

pathway_sigpw_data_combined %>%
  # mutate(Pathway = ifelse(Pathway == "Protein_homeostasis_ubiquitination", "PTN Homeost. Ubiquitin.", Pathway)) %>%
  filter(!str_detect(Group, "_R|_NR")) %>%
  filter(! Pathway %in% c("Other", "Other_signaling")) %>%
  ggplot(aes(x = fct_rev(Group), y = Fraction_mutated_samples, fill = Group)) + geom_col() +
  facet_wrap(~Pathway) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() + scale_fill_manual(values = trial_colors_combined) +
  guides(fill = FALSE) + labs(x = "", title = "SIGPW pathways")

```

# Select/defining Oncoplot custom pathways

```{r}
maftools_BP_SMG_pathways %>%
  select(Gene, Pathway) %>%
  unique() %>%
  arrange(Pathway) %>% 
  pull(Pathway) %>%
  table()

### BUG warning
# Each gene is only assigned to 1 pathway, so result is different when using them together or separately... 
# It does not work to plot them separately. Each gene will still be assigned to only 1 pathway in the collection
# Even if you plot 1 pathway, it's innacurate if those genes were used in the rest of the collection
# This is a bug, and should be explicit that gene should belong more than once. 
# Feeding each pathway at once or using a completely exclusive collection to prevent this
oncoplot_custom_pathways <- immune_escape_genes_df %>%
           select(Gene, Pathway.general) %>%
  setNames(c("Gene", "Pathway")) %>%
  filter(!Pathway %in% c("PD-L1", "CD58", "Epigenetic")) %>%
  rbind(
    maftools_BP_SMG_pathways %>%
      select(Gene, Pathway) %>%
      unique() %>%
      filter(Pathway %in% c("Cell cycle", "Chromatin histone modifiers",
                                "Genome integrity", "Immune signaling",
                                "MAPK signaling",
                                "Protein homeostasis/ubiquitination",
                                "RTK signaling", "Wnt/B-catenin signaling")) %>%
      mutate(Pathway = paste0("SMG_", Pathway))) %>%
  rbind(maftools_sigpw_pathways %>%
          select(Gene, Pathway) %>% 
          unique() %>%
      mutate(Pathway = paste0("SIGPW_", Pathway)))

oncoplot_custom_pathways$Pathway <- gsub(" ", "_", oncoplot_custom_pathways$Pathway)
oncoplot_custom_pathways$Pathway <- gsub("-", "_", oncoplot_custom_pathways$Pathway)
oncoplot_custom_pathways$Pathway <- gsub("\\/", "_", oncoplot_custom_pathways$Pathway)

oncoplot_custom_pathways$Pathway %>% unique()
oncoplot_custom_pathways$Pathway %>% n_distinct()
oncoplot_custom_pathways$Gene %>% table() %>% as.data.frame() %>% filter(Freq > 1) %>% n_distinct()
```

## Careful, there is a oncoplot bug when using overlapping pathways

```{r}
### BUG warning
# Each gene is only assigned to 1 pathway, so result is different when using them together or separately... 
# It does not work to plot them separately. Each gene will still be assigned to only 1 pathway in the collection
# Even if you plot 1 pathway, it's innacurate if those genes were used in the rest of the collection
# This is a bug, and should be explicit that gene should belong more than once. 
# Feeding each pathway at once or using a completely exclusive collection to prevent this

# current_maf <- combined_maf_specific_f

# Result is different, because each gene is only assigned to 1 pathway
# SIGPW_TP53 has an innacurate readout
oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways,
         selectedPathways = c("SMG_Genome_integrity", "SIGPW_TP53"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# Still does not work, because genes from SIGPW_TP53 are already in "SMG_Genome integrity"
# Even though this is kind of hidden...
oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways,
         selectedPathways = c("SIGPW_TP53"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# This works, because only the SIGPW_TP53 is in the pathway list, so all genes are used
oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways %>% filter(Pathway == "SIGPW_TP53"),
         selectedPathways = c("SIGPW_TP53"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# Result is different, because each gene is only assigned to 1 pathway
oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways,
         selectedPathways = c("SMG_Cell_cycle", "SIGPW_Cell_Cycle"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways,
         selectedPathways = c("SMG_Cell_cycle"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# Still does not work, because genes from SIGPW_Cell_Cycle are already in SMG_Cell
oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways,
         selectedPathways = c("SIGPW_Cell_Cycle"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

# This works, because only the SIGPW_Cell_Cycle is in the pathway list, so all genes are used
oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways %>% filter(Pathway == "SIGPW_Cell_Cycle"),
         selectedPathways = c("SIGPW_Cell_Cycle"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

oncoplot(combined_maf_specific_f, pathways = oncoplot_custom_pathways %>% filter(Pathway == "SIGPW_PI3K"),
         selectedPathways = c("SIGPW_PI3K"),
         clinicalFeatures = c("Trial", "Response"),
         topBarData = "TMB",
         # annotationColor = annotation_color_list,
         gene_mar = 8, fontSize = 0.6,
         sortByAnnotation = TRUE)

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$SIGPW_PI3K_signaling)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$PIK3CA)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$PIK3CA) %>% prop.table(margin = 1) %>% round(2)

```

# Get pathway proportions custom list

```{r}
dir.create(here(output_dir, "pathway_proportions"))

oncoplot_custom_pathways <- oncoplot_custom_pathways %>%
  rbind(data.frame(Gene = "TP53", Pathway = "TP53")) %>%
  rbind(data.frame(Gene = c("BRCA1", "BRCA2"), Pathway = c("BRCA1_2", "BRCA1_2"))) %>%
  rbind(data.frame(Gene = c("KRAS", "NRAS", "HRAS", "BRAF"), Pathway = c(rep("RAS_RAF", 4))))

oncoplot_custom_pathways
oncoplot_custom_pathways$Pathway %>% unique()
oncoplot_custom_pathways %>% factor_summary()

oncoplot_custom_pathways %>% filter(Pathway == "SIGPW_Cell_Cycle")

get_pathway_proportion_trial_mafs(combined_maf_specific_f,
                                  pathway_dir = "pathway_proportions",
                                  pathways = oncoplot_custom_pathways)

# Get NICHE-MSI separately, breaks with NR = 0
# For now ignored, function already outputs the total and R frequencies before
get_pathway_proportion_trial_mafs(combined_maf_specific_f,
                                  pathway_dir = "pathway_proportions",
                                  trial_list = c("NICHE-MSI"),
                                  pathways = oncoplot_custom_pathways)

# Combine data from all these files
parse_pathway_proportions(here(output_dir, "pathway_proportions"))

pathway_data_combined <- read.csv2(here(output_dir, "pathway_proportions",  "pathway_data_combined.csv")) %>% left_join(trial_group_map)
pathway_gene_data_combined <- read.csv2(here(output_dir, "pathway_proportions", "pathway_gene_data_combined.csv")) %>% left_join(trial_group_map)

pathway_data_combined$Group <- factor(pathway_data_combined$Group,
                                      levels = c(trial_list, sprintf("%s_R", trial_list), sprintf("%s_NR", trial_list)))

pathway_data_combined$Trial <- factor(pathway_data_combined$Trial,
                                      levels = trial_list)

pathway_gene_data_combined %>%
  select(Gene, Pathway) %>%
  unique() %>%
  filter(!is.na(Gene)) %>%
  write_csv2(here(output_dir, "custom_gene_list.csv"))

custom_gene_list <- read.csv2(here(output_dir, "custom_gene_list.csv"))
custom_gene_list$Gene[!custom_gene_list$Gene %in% genes_to_assess]
custom_gene_list$Gene[!custom_gene_list$Gene %in% oncoKB_cancerGeneList$Hugo.Symbol]

pathway_data_combined$Pathway %>% unique()

```

```{r, fig.height = 12, fig.width = 14}

pathway_data_combined_plot <- pathway_data_combined %>%
  # mutate(Pathway = ifelse(Pathway == "Protein_homeostasis_ubiquitination", "PTN Homeost. Ubiquitin.", Pathway)) %>%
  filter(!str_detect(Group, "_R|_NR")) %>%
  filter(! Pathway %in% c("Other", "Other_signaling")) %>%
  ggplot(aes(x = fct_rev(Group), y = Fraction_mutated_samples, fill = Group)) + geom_col() +
  facet_wrap(~Pathway) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() + scale_fill_manual(values = trial_colors_combined) +
  guides(fill = FALSE) + labs(x = "")
pathway_data_combined_plot

graph2pdf(x = pathway_data_combined_plot,
          file = here(output_dir, "_pathway_data_combined_plot.pdf"),
          colormodel = "rgb",
          font = "Arial",
          height = 9,
          width = 11)

```

```{r}
responder_pathway_proportion <- read.csv2(
  here(output_dir, "pathway_proportions", "pathway_proportions_Combined_R.csv"))
nresponder_pathway_proportion <- read.csv2(
  here(output_dir, "pathway_proportions", "pathway_proportions_Combined_NR.csv"))

responder_pathway_proportion %>%
  mutate(Response = "R") %>%
  rbind(nresponder_pathway_proportion %>%
          mutate(Response = "NR")) %>%
  ggplot(aes(x = Pathway, y = Fraction_mutated_samples, fill = Response)) +
  geom_col(position = "dodge") + theme_Publication()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  guides(fill = FALSE) +
  labs(x = "", y = "Fraction mutated")

```

```{r, fig.height = 12, fig.width = 14}
pathway_data_combined_response_plot <- pathway_data_combined %>%
  filter(! Trial %in% c("NICHE-MSI")) %>%
  filter(str_detect(Group, "_R|_NR")) %>%
  mutate(Response = ifelse(str_detect(Group, "_R"), "R", "NR")) %>%
  filter(! Pathway %in% c("Other", "Other_signaling")) %>%
  ggplot(aes(x = Pathway, y = Fraction_mutated_samples, fill = Response)) +
  geom_col(position = "dodge") +
  theme_Publication()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  guides(fill = FALSE) +
  labs(x = "", y = "Fraction mutated") + facet_wrap(~Trial, ncol = 4) +
  coord_flip()
pathway_data_combined_response_plot

graph2pdf(x = pathway_data_combined_response_plot,
          file = here(output_dir, "_pathway_data_combined_response_plot.pdf"),
          colormodel = "rgb",
          font = "Arial",
          height = 9,
          width = 11)

```

```{r, fig.height = 10, fig.width = 10}
pathway_data_combined %>%
  filter(! Trial %in% c("NICHE-MSI")) %>%
  filter(str_detect(Group, "_R|_NR")) %>%
  mutate(Response = ifelse(str_detect(Group, "_R"), "R", "NR")) %>%
  select(Pathway, N, Fraction_mutated_samples, Trial, Response) %>%
  mutate(Fraction_mutated_samples = ifelse(
    is.na(Fraction_mutated_samples), 0 ,Fraction_mutated_samples)) %>%
  pivot_wider(names_from = Response, values_from = Fraction_mutated_samples) %>%
  mutate(R = ifelse(is.na(R), 0 , R), NR = ifelse(is.na(NR), 0 , NR))  %>%
  mutate(mean_diff = R - NR) %>%
  filter(!Pathway %in% c("Other", "Other_signaling")) %>%
  ggplot(aes(x = Pathway, y = mean_diff)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(col = Trial, size = 1)) +
  theme_Publication()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_color_manual(values = trial_colors_combined) +
  guides(fill = FALSE, size = FALSE) +
  labs(x = "", y = "Delta R - NR mutation frequency") +
  coord_flip() + 
  geom_hline(yintercept = 0, linetype = "dashed")

pathway_data_combined_response_mean_diff <- pathway_data_combined %>%
  filter(! Trial %in% c("NICHE-MSI")) %>%
  filter(str_detect(Group, "_R|_NR")) %>%
  mutate(Response = ifelse(str_detect(Group, "_R"), "R", "NR")) %>%
  select(Pathway, Fraction_mutated_samples, Trial, Response) %>%
  mutate(Fraction_mutated_samples = ifelse(
    is.na(Fraction_mutated_samples), 0 ,Fraction_mutated_samples)) %>%
  pivot_wider(names_from = Response, values_from = Fraction_mutated_samples) %>%
  mutate(R = ifelse(is.na(R), 0 , R), NR = ifelse(is.na(NR), 0 , NR))  %>%
  mutate(mean_diff = R - NR) 

pathway_data_combined_response_mean_diff %>%
  write_csv2(here(output_dir, "pathway_data_combined_response_mean_diff.csv"))

pathway_data_combined_response_mean_diff %>% filter(Trial %in% c("IMCISION")) %>% view()

pathway_data_combined_response_mean_diff %>%
  filter(!Pathway %in% c("Other", "Other_signaling")) %>%
  ggplot(aes(x = Pathway, y = mean_diff)) +
  geom_point(aes(col = Trial, size = 1)) +
  theme_Publication()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_color_manual(values = trial_colors_combined) +
  guides(fill = FALSE, size = FALSE) +
  labs(x = "", y = "Delta R - NR mutation frequency") +
  coord_flip() + 
  geom_hline(yintercept = 0, linetype = "dashed") +
  facet_wrap(~Trial, ncol = 4)

pathway_data_combined_response_mean_diff_plot <- pathway_data_combined_response_mean_diff %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR),
               values_to = "Fraction_mutated_samples", names_to = "Response") %>%
  ggplot(aes(x = Pathway, y = Fraction_mutated_samples, fill = Response)) +
  geom_col(alpha = 0.75)   +
  facet_wrap(~Trial, ncol = 4) +
  coord_flip()  + theme_Publication() +
  geom_point(aes(x = Pathway, y = mean_diff, col = Trial, size = 0.25), shape = 17) +
  facet_wrap(~Trial, ncol = 4) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_color_manual(values = trial_colors_combined) +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  guides(fill = FALSE, size = FALSE, alpha = 0.5) +
  labs(x = "", y = "Mutation frequency (NR vs R delta)") + 
  geom_hline(yintercept = 0, linetype = "dashed")
pathway_data_combined_response_mean_diff_plot

graph2pdf(x = pathway_data_combined_response_mean_diff_plot,
          file = here(output_dir, "_pathway_data_combined_response_mean_diff_plot.pdf"),
          colormodel = "rgb",
          font = "Arial",
          height = 8,
          width = 10)
```

# Check

```{r}
# pathway_level_mutation_df_non_dmmr
dim(pathway_level_mutation_df)
for (pathway in pathway_level_mutation_df %>%
  select(starts_with("SMG"), starts_with("GIE"), starts_with("SIGPW"), "DDR") %>% colnames()) {
  print(pathway)
  fisher_res <- fisher.test(table(pathway_level_mutation_df_non_dmmr$Response, pathway_level_mutation_df_non_dmmr[, pathway, drop = TRUE]))
  print(fisher_res)
}
```

# Specific Cohort Oncoplots

```{r, fig.height = 14, fig.width = 14}
oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "BELLINI") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("RNA_abundance", "MAPK_signaling", "Genome_integrity"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "BELLINI") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("Chromatin_other", "Chromatin_SWI/SNF_complex", "Chromatin_histone_modifiers"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "IMCISION") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("NOTCH_signaling", "RTK_signaling"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "IMCISION") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("Chromatin_other", "Chromatin_SWI/SNF_complex", "Chromatin_histone_modifiers"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "MATISSE") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("RNA_abundance", "NFKB_signaling"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "NABUCCO") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("Apoptosis", "RTK_signaling"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "NICHE-MSS") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("MAPK_signaling", "Protein_homeostasis/ubiquitination", "Genome_integrity"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "OpACIN_neo") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("MAPK_signaling", "RTK_signaling"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "OpACIN_neo") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("Chromatin_other", "Chromatin_SWI/SNF_complex", "Chromatin_histone_modifiers"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "PRADO") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("MAPK_signaling", "RTK_signaling"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "PRADO") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = "smgbp",
         selectedPathways = c("Chromatin_other", "Chromatin_SWI/SNF_complex", "Chromatin_histone_modifiers"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)
```


```{r, fig.height = 14, fig.width = 14}
oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "IMCISION") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = oncoplot_custom_pathways %>% filter(Pathway == "SIGPW_PI3K"),
         selectedPathways = c("SIGPW_PI3K"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot(subsetMaf(combined_maf_specific_f, tsb = dna_metadata_f %>% filter(Trial == "MATISSE") %>% pull(Tumor_Sample_Barcode)), 
         clinicalFeatures = c("Response", "Trial"),
         pathways = oncoplot_custom_pathways %>% filter(Pathway == "IFN_y_pathway"),
         selectedPathways = c("IFN_y_pathway"),
         annotationColor = annotation_color_list,
         topBarData = "TMB", logColBar = FALSE, sortByAnnotation = TRUE, removeNonMutated = FALSE)

oncoplot_custom_pathways$Pathway %>% unique()

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$SIGPW_PI3K_signaling)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$PIK3CA)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$PIK3CA) %>% prop.table(margin = 1) %>% round(2)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$GIE_IFNy) %>% prop.table(margin = 1) %>% round(2)

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$GIE_IFNy) %>% prop.table(margin = 1) %>% round(2)

```

# ----------
# Cytoband coordinates

```{r, fig.height = 6, fig.width = 6}
cytoband_df <- read.delim(
  here("data", "external", "cytoBand.txt"), header = F) %>%
  setNames(c("chr", "startpos", "endpos", "cytoband", "giemsa"))

cytoband_df <- cytoband_df %>%
  mutate(chr_arm = paste0(chr, ifelse(str_detect(cytoband, "p"), "p", "q")))

cytoband_pq_df <- cytoband_df %>%
  group_by(chr, chr_arm) %>%
  summarize(startpos = min(startpos), endpos = max(endpos)) %>%
  mutate(chr_arm_size = endpos - startpos,
         chr_arm_size_mb = chr_arm_size / 1e6)

chr_levels <- c(paste0("chr", c(1:22)), "chrX", "chrY")

sum(cytoband_pq_df$chr_arm_size)
sum(cytoband_pq_df$chr_arm_size_mb)

cytoband_pq_df %>% ggplot(aes(x = chr_arm_size_mb)) +
  geom_histogram() + theme_Publication()

cytoband_pq_df %>%
  select(chr, chr_arm, chr_arm_size_mb) %>%
  mutate(chr_arm = str_sub(chr_arm,-1,-1)) %>%
  ggplot(aes(x = factor(chr, levels = rev(chr_levels)),
             y = chr_arm_size_mb, fill = chr_arm)) +
  geom_col() + theme_Publication() + coord_flip() +
  labs(x = "", y = "Chr arm sizes (Mb)")

cytoband_pq_df %>%
  select(chr, chr_arm, chr_arm_size_mb) %>%
  mutate(chr_arm = str_sub(chr_arm,-1,-1)) %>%
  ggplot(aes(x = factor(chr, levels = rev(chr_levels)),
             y = chr_arm_size_mb, fill = chr_arm)) +
  geom_col() + theme_Publication() + coord_flip() +
  labs(x = "", y = "Chr arm sizes (Mb)") +
  facet_wrap(~chr_arm)

cytoband_pq_df %>%
  write_csv2(here(output_dir, "cytoband_pq_df.csv"))

chr_arms_exclude <- c("chrYp", "chrYq", "chrXp", "chrXp", "chr13p",
                      "chr14p", "chr15p", "chr21p", "chr22p")

chromosome_sizes <- cytoband_pq_df %>% group_by(chr) %>%
  summarize(chr_size = sum(chr_arm_size), chr_size_mb = sum(chr_arm_size_mb))

```

# ----------

# Parse FACETS data

```{r}
trial_folders <- c("PRADO/DNAseq_output", "OpACIN_neo/DNAseq_output",
                   "NICHE/DNAseq_output", "NICHE/DNAseq_output_MSI_expansion",
                   "NABUCCO/DNAseq_output", "NABUCCO2/DNAseq_output",
                   "MATISSE/DNAseq_output", "IMCISION/DNAseq_output",
                   "BELLINI/DNAseq_output", "BELLINI/DNAseq_output_BELLINI3")
###
# out: data frame of segment summaries pre and post clustering of
#       segments. The columns are: ‘chrom’ the chromosome to which
#       the segment belongs; ‘seg’ the segment number; ‘num.mark’ the
#       number of SNPs in the segment; ‘nhet’ the number of SNPs that
#       are deemed heterozygous; ‘cnlr.median’ the median log-ratio
#       of the segment; ‘mafR’ the log-odds-ratio summary for the
#       segment; ‘segclust’ the segment cluster to which segment
#       belongs; ‘cnlr.median.clust’ the median log-ratio of the
#       segment cluster; ‘mafR.clust’ the log-odds-ratio summary for
#       the segment cluster; ‘cf’ the cellular fraction of the
#       segment; ‘tcn’ the total copy number of the segment; ‘lcn’
#       the minor copy number of the segment.
# cncf: dataframe consisting of the columns of segmentation output as
#       well as cellular fraction (cf), total (tcn) and lesser (lcn)
#       copy number of each segment and their em counterpart (with
#       .em suffix)
###

parse_facets <- FALSE

if (parse_facets) {
  facets_data <- data.frame(matrix(0, ncol = 11, nrow = 0))

  # Loop over trials
  # current_trial <- trial_folders[1]
  for (current_folder in trial_folders){
    print(current_folder)
  
    facets_folder <- here("data", "raw", "Pancancer_DNA_preprocessed_Pedro",
                           current_folder, "FACETS", "fitted")
    
    samples <- list.files(facets_folder, full.names = FALSE)
    samples <- samples[str_detect(samples, "_fitted.csv$")]
    current_sample <- samples[1]
    
    # Loop over sample folders
    for (current_sample in samples) {
      # print(current_sample)
  
      CF_samples <- str_split(current_sample, ".snppile.csv.gz_fitted.csv") %>%
        unlist()
      CF_samples <- CF_samples[1]
      CF_samples <- str_split(CF_samples, "-vs-") %>% unlist()
      
      CF_samples_tumor <- CF_samples[1]
      CF_samples_tumor <- str_split(CF_samples_tumor, "_") %>% unlist()
      CF_samples_tumor <- CF_samples_tumor[str_detect(CF_samples_tumor, "CF|Nr")]
      CF_sample_gl <- CF_samples[2]
      CF_sample_gl <- str_split(CF_sample_gl, "_") %>% unlist()
      CF_sample_gl <- CF_sample_gl[str_detect(CF_sample_gl, "CF|Nr")]

      sample_facets <- read.delim(
        here(facets_folder, current_sample),
        header = T, sep = ",") %>%
        mutate(filename = current_sample,
               CF_sample_tumor = CF_samples_tumor,
               CF_sample_gl = CF_samples[2])
      
      if (!(c("emflags") %in% colnames(sample_facets))) {
        sample_facets <- sample_facets %>%
          mutate(emflags = NA) %>%
          relocate(emflags, .after = "dipLogR")
      }
  
      facets_data <- facets_data %>% rbind(sample_facets)
    }
  }
  facets_data %>% write_csv2(here(output_dir, "facets_data.csv"))
}

facets_data <- read_csv2(here(output_dir, "facets_data.csv"))

facets_data_df <- facets_data %>% left_join(dna_metadata_f,
                                            by = "CF_sample_tumor")

facets_data_df <- facets_data_df %>%
  filter(!(Trial %in% "NICHE-MSI" & Tumor_nr == "Tumor_2"))

facets_data_samples <- facets_data$CF_sample_tumor %>% unique()
facets_data_samples[!facets_data_samples %in% dna_metadata_f$CF_sample_tumor]

# Filter for samples contained in maf file, filtered metadata
facets_data_df <- facets_data_df %>% filter(CF_sample_tumor %in% dna_metadata_f$CF_sample_tumor)

facets_data_df$Patient %>% n_distinct()

WGD_ploidy_cutoff <- 2.75

facets_data_metrics <- facets_data_df %>%
  select(Patient, CF_sample_tumor, Cohort, Trial, Response, Response_detail,
         Histology, Source_MAF, Tumor_Sample_Barcode, Purity, Ploidy,
         emflags) %>%
  unique() %>%
  filter(!is.na(Trial)) %>%
  left_join(n_variants_dp20_vaf_tmb %>%
              select(CF_sample_tumor, TMB, WES_target_size)) %>%
  mutate(ploidy_WGD = ifelse(Ploidy > WGD_ploidy_cutoff, "WGD", "not_WGD"))

# Set unreliable purity calls to NA. This was duplicated for some patients
facets_data_metrics$Patient %>% table() %>% as.data.frame() %>% filter(Freq > 1)
facets_data_metrics %>%
  filter(Patient %in% c("MAT_20", "MAT_54", "MAT_71", "MAT_8")) %>% view()
facets_data_metrics %>%
  filter(emflags == "Insufficient information to estimate purity. Likely diplod or purity too low.") %>% view()

facets_data_metrics <- facets_data_metrics %>%
  mutate(Purity = ifelse(Patient %in% c("MAT_20", "MAT_54", "MAT_8"),
                         NA, Purity)) %>%
  unique()

facets_data_metrics$Patient %>% table() %>% as.data.frame() %>% filter(Freq > 1)

facets_data_metrics %>%
  mutate_if(is.character, as.factor) %>%
  summary(maxsum = 10)

facets_data_metrics$ploidy_WGD %>% table() %>% prop.table() %>% round(2)

facets_data %>% filter(!is.na(emflags)) %>% select(CF_sample_tumor, emflags) %>% unique() %>% factor_summary()
```

# FACETS metrics and Ploidy WGD
```{r}
facets_data_metrics %>%
  ggplot(aes(x = Purity, y = Ploidy, col = Trial)) +
  geom_point() +
  theme_Publication()

facets_data_metrics %>%
  ggplot(aes(x = Purity, y = Ploidy, col = Response)) +
  geom_point() +
  theme_Publication() +
  facet_wrap(~Trial)

facets_data_metrics %>%
  ggplot(aes(x = Purity, y = TMB, col = Trial)) +
  geom_point() +
  theme_Publication()

facets_data_metrics %>%
  ggplot(aes(x = Purity, y = TMB, col = Response)) +
  geom_point() +
  theme_Publication() +
  facet_wrap(~Trial)

facets_data_metrics %>%
  ggplot(aes(x = Ploidy, y = TMB, col = Trial)) +
  geom_point() +
  theme_Publication() +
  scale_color_manual(values = trial_colors) +
  theme(text = element_text(family = "Arial"))

facets_data_metrics %>%
  ggplot(aes(x = Ploidy, y = TMB, col = Response)) +
  geom_point() +
  theme_Publication() +
  facet_wrap(~Trial)

facets_data_metrics %>%
  filter(Trial == "NABUCCO") %>%
  ggplot(aes(x = Ploidy, y = TMB, col = Response)) +
  geom_point() +
  theme_Publication() +
  facet_wrap(~Trial)

facets_data_metrics %>%
  ggplot(aes(x = Ploidy, y = TMB, col = Response)) +
  geom_point() +
  theme_Publication() +
  facet_wrap(~Trial, scales = "free_y")

facets_data_metrics %>%
  ggplot(aes(x = Ploidy)) +
  geom_density() +
  theme_Publication() +
  geom_vline(xintercept = WGD_ploidy_cutoff, linetype = "dashed")

facets_data_metrics %>%
  ggplot(aes(x = Ploidy, col = Trial)) +
  geom_density() +
  theme_Publication() +
  geom_vline(xintercept = WGD_ploidy_cutoff, linetype = "dashed")

facets_data_metrics %>%
  filter(! Trial %in% c("NICHE-MSI")) %>%
  ggplot(aes(x = Ploidy, col = Trial)) +
  geom_density() +
  theme_Publication() +
  geom_vline(xintercept = WGD_ploidy_cutoff, linetype = "dashed")

facets_data_metrics %>% dim()
facets_data_metrics %>% write_csv2(here(output_dir, "facets_data_metrics.csv"))
facets_data_metrics <- read_csv2(here(output_dir, "facets_data_metrics.csv"))

```

```{r}
table(facets_data_metrics$ploidy_WGD)
table(facets_data_metrics$Response)

table(facets_data_metrics$ploidy_WGD,
      facets_data_metrics$Response)

 table(facets_data_metrics$ploidy_WGD,
      facets_data_metrics$Response) %>%
  as.data.frame() %>% setNames(c("ploidy_WGD", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Response_rate = round(R / Total, 2))
 
 table(facets_data_metrics$ploidy_WGD,
      facets_data_metrics$Response) %>%
  as.data.frame() %>% setNames(c("ploidy_WGD", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Response_rate = round(R / Total, 2)) %>%
  ggplot(aes(x = ploidy_WGD,
             y = Response_rate)) + geom_col() +
  theme_Publication() + labs(x = "")
 
facets_data_metrics_tmb_low <- facets_data_metrics %>% filter(TMB < 10)
 
table(facets_data_metrics_tmb_low$ploidy_WGD,
      facets_data_metrics_tmb_low$Response)
 
table(facets_data_metrics_tmb_low$ploidy_WGD,
      facets_data_metrics_tmb_low$Response) %>%
  as.data.frame() %>% setNames(c("ploidy_WGD", "Response", "Freq")) %>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Response_rate = round(R / Total, 2)) %>%
  ggplot(aes(x = ploidy_WGD,
             y = Response_rate)) + geom_col() +
  theme_Publication() + labs(x = "")
 
table(facets_data_metrics$Trial,
      facets_data_metrics$ploidy_WGD)

table(facets_data_metrics$Trial,
      facets_data_metrics$ploidy_WGD) %>%
  as.data.frame() %>% setNames(c("Trial", "WGD_status", "Freq"))%>%
  pivot_wider(names_from = WGD_status, values_from = Freq) %>%
  mutate(Total = WGD + not_WGD, WGD_prop = round(WGD / Total, 2))

table(facets_data_metrics$Trial,
      facets_data_metrics$ploidy_WGD) %>%
  as.data.frame() %>% setNames(c("Trial", "WGD_status", "Freq"))%>%
  pivot_wider(names_from = WGD_status, values_from = Freq) %>%
  mutate(Total = WGD + not_WGD, WGD_prop = round(WGD / Total, 2)) %>%
  ggplot(aes(x = reorder(Trial, WGD_prop), y = WGD_prop)) +
  geom_col() + theme_Publication() + coord_flip() +
  labs(x = "")

trial_wgd_response <- table(facets_data_metrics$Trial,
      facets_data_metrics$ploidy_WGD,
      facets_data_metrics$Response) %>%
  as.data.frame() %>% setNames(c("Trial", "WGD", "Response", "Freq"))%>%
  pivot_wider(names_from = Response, values_from = Freq) %>%
  mutate(Total = R + NR, Response_rate = round(R / Total, 2))

trial_wgd_response %>%
  ggplot(aes(x = factor(Trial, levels = c(unique(tmb_response_rate$Trial))),
             y = Response_rate, fill = WGD)) +
  geom_col(position = "dodge2") +
  theme_Publication() +
  labs(x = "", y = "Response rate") +
  scale_y_continuous(breaks = seq(0, 1, by = 0.25)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = c("#00BA38", "#619CFF"))

```

# FACETS segments
```{r}
facets_data %>%
  mutate(size = end - start,
         size_mb = size / 1e6) %>%
  ggplot(aes(x = log10(size))) +
  geom_density() + theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

facets_data %>%
  mutate(size = end - start,
         size_mb = size / 1e6) %>%
  ggplot(aes(x = size_mb)) +
  geom_density() + theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

facets_data %>%
  mutate(size = end - start,
         size_mb = size / 1e6) %>%
  ggplot(aes(x = log10(size), y = log10(num.mark))) +
  geom_pointdensity()+ theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

facets_data %>%
  mutate(size = end - start,
         size_mb = size / 1e6) %>%
  ggplot(aes(x = size_mb, y = log10(num.mark))) +
  geom_pointdensity()+ theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

facets_data %>%
  ggplot(aes(x = log10(num.mark), y = log10(nhet))) +
  geom_pointdensity()+ theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

facets_data %>%
  ggplot(aes(x = log10(nhet), y = log2(tcn.em + 1))) +
  geom_pointdensity() + theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))
  
facets_data %>%
  mutate(size = end - start,
         size_mb = size / 1e6) %>%
  ggplot(aes(y = log2(tcn.em + 1), x = size_mb)) +
  geom_pointdensity()+ theme_Publication() +
  guides(colour = guide_colourbar(barwidth = 10))

```

```{r}
facets_data %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "NEUTR") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "DEL") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "HEMIZYG") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "LOH") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "LOH") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "DUP") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()
facets_data %>% filter(type == "DUP-LOH") %>% mutate_if(is.character, as.factor) %>% select(-c(CF_sample_tumor, CF_sample_gl, filename)) %>% summary()

```

# Intersect segments Chr arms and cytobands

```{r FACETS parse intervals}
intervals_df <- facets_data_df %>%
  select(chrom, start, end, cnlr.median, tcn.em, type, num.mark, nhet, CF_sample_tumor, Patient) %>%
  mutate(chr = paste0("chr", chrom)) %>%
  rename(startpos = start, endpos = end) %>%
  mutate(cn = tcn.em) %>%
  mutate(CF_sample = CF_sample_tumor)

parse_facets_arms <- FALSE
if (parse_facets_arms) {
  facets_c_arms <- data.frame()
  for (i in 1:nrow(intervals_df)){
    current_chr <- intervals_df$chr[i]
    current_cytoband_pq_df <- cytoband_pq_df %>% filter(chr == current_chr)
  
    # Initialize an empty list to store split intervals
    split_intervals <- list()
    
    interval_start <- intervals_df$startpos[i]
    interval_end <- intervals_df$endpos[i]
    
    # Filter cytoband_df to find overlapping intervals
    overlapping_rows <- current_cytoband_pq_df[
      (current_cytoband_pq_df$startpos <= interval_end) & (current_cytoband_pq_df$endpos >= interval_start), ]
    
    if (nrow(overlapping_rows) == 0) {
        result_df <- data.frame(
        start = interval_start,
        end = interval_end,
        chr_arm = overlap_name,
        Patient = intervals_df$Patient[i],
        CF_sample = intervals_df$CF_sample[i],
        cnlr.median = intervals_df$cnlr.median[i],
        cn = intervals_df$cn[i],
        type = intervals_df$type[i]
        )
  
        facets_c_arms <- facets_c_arms %>% rbind(result_df)
        next
    }
    
    # Split overlapping intervals
    for (j in 1:nrow(overlapping_rows)) {
      overlap_start <- max(interval_start, overlapping_rows$startpos[j])
      overlap_end <- min(interval_end, overlapping_rows$endpos[j])
      overlap_name <- overlapping_rows$chr_arm[j]
      
      split_intervals[[length(split_intervals) + 1]] <- data.frame(
        start = overlap_start,
        end = overlap_end,
        chr_arm = overlap_name,
        Patient = intervals_df$Patient[i],
        CF_sample = intervals_df$CF_sample[i],
        cnlr.median = intervals_df$cnlr.median[i],
        cn = intervals_df$cn[i],
        type = intervals_df$type[i]
        )
    }
      
    # Combine split intervals into a new dataframe
    result_df <- do.call(rbind, split_intervals)
    facets_c_arms <- facets_c_arms %>% rbind(result_df)
  }
  
  facets_c_arms <- facets_c_arms %>% unique()
  
  # Some chr arms are not used (p arms of 13, 14, 15, 21 and 22)
  # Exclude chr Y and X for aneuploidy calculation
  cytoband_pq_df$chr_arm[!cytoband_pq_df$chr_arm %in% facets_c_arms$chr_arm]
  
  facets_c_arms <- facets_c_arms %>%
    mutate(Patient = as.factor(Patient)) %>%
    left_join(dna_metadata_f %>%
                select(Patient, Trial, Response) %>%
                unique() ) %>%
    relocate(Patient, CF_sample, Trial, Response, .before = start)
  facets_c_arms <- facets_c_arms %>% unique()
  # %>%               filter(!(Trial_nr == 145 & Response == "R"))
  facets_c_arms %>% write_csv2(here(output_dir, "facets_c_arms.csv"))
}

parse_facets_cytobands <- FALSE
if (parse_facets_cytobands) {
  facets_c_cytobands <- data.frame()
  for (i in 1:nrow(intervals_df)){
    current_chr <- intervals_df$chr[i]
    current_cytoband_df <- cytoband_df %>% filter(chr == current_chr)
  
    # Initialize an empty list to store split intervals
    split_intervals <- list()
    
    interval_start <- intervals_df$startpos[i]
    interval_end <- intervals_df$endpos[i]
    
    # Filter cytoband_df to find overlapping intervals
    overlapping_rows <- current_cytoband_df[
      (current_cytoband_df$startpos <= interval_end) & (current_cytoband_df$endpos >= interval_start), ]
    
    if (nrow(overlapping_rows) == 0) {
        result_df <- data.frame(
        start = interval_start,
        end = interval_end,
        cytoband = NA,
        chr_arm = overlap_name,
        Patient = intervals_df$Patient[i],
        CF_sample = intervals_df$CF_sample[i],
        cnlr.median = intervals_df$cnlr.median[i],
        cn = intervals_df$cn[i],
        type = intervals_df$type[i]
        )
  
        facets_c_cytobands <- facets_c_cytobands %>% rbind(result_df)
        next
    }
    
    # Split overlapping intervals
    for (j in 1:nrow(overlapping_rows)) {
      overlap_start <- max(interval_start, overlapping_rows$startpos[j])
      overlap_end <- min(interval_end, overlapping_rows$endpos[j])
      overlap_arm <- overlapping_rows$chr_arm[j]
      overlap_name <- overlapping_rows$cytoband[j]

      split_intervals[[length(split_intervals) + 1]] <- data.frame(
        start = overlap_start,
        end = overlap_end,
        cytoband = overlap_name,
        chr_arm = overlap_arm,
        Patient = intervals_df$Patient[i],
        CF_sample = intervals_df$CF_sample[i],
        cnlr.median = intervals_df$cnlr.median[i],
        cn = intervals_df$cn[i],
        type = intervals_df$type[i]
        )
    }
      
    # Combine split intervals into a new dataframe
    result_df <- do.call(rbind, split_intervals)
    facets_c_cytobands <- facets_c_cytobands %>% rbind(result_df)
  }
  
  facets_c_cytobands <- facets_c_cytobands %>% unique()
  
  # Some chr arms are not used (p arms of 13, 14, 15, 21 and 22)
  # Exclude chr Y and X for aneuploidy calculation
  cytoband_pq_df$chr_arm[!cytoband_pq_df$chr_arm %in% facets_c_cytobands$chr_arm]
  
  facets_c_cytobands <- facets_c_cytobands %>%
    mutate(Patient = as.factor(Patient)) %>%
    left_join(dna_metadata_f %>%
                select(Patient, Trial, Response) %>%
                unique() ) %>%
    relocate(Patient, CF_sample, Trial, Response, .before = start)
  facets_c_cytobands <- facets_c_cytobands %>% unique()
  # %>%               filter(!(Trial_nr == 145 & Response == "R"))
  facets_c_cytobands %>% write_csv2(here(output_dir, "facets_c_cytobands.csv"))
}

facets_c_arms <- read_csv2(here(output_dir, "facets_c_arms.csv"))
facets_c_cytobands <- read_csv2(here(output_dir, "facets_c_cytobands.csv"))

facets_c_arms <- facets_c_arms %>% mutate(segment_size = end - start,
                                          segment_size_mb = segment_size / 1e6)

facets_c_arms_summary <- facets_c_arms %>%
  select(c("Trial", "chr_arm", "cnlr.median", "cn", "type", "segment_size_mb")) %>%
  mutate(chr_arm_to_exclude = chr_arm %in% chr_arms_exclude)

facets_c_arms_summary %>% filter(cn == 0) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary %>% filter(cn == 1) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary %>% filter(cn == 2) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary %>% filter(cn > 2) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary %>% filter(cn >3) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary %>% filter(cn >5) %>% mutate_if(is.character, as.factor) %>% summary()

facets_c_arms_summary_e <- facets_c_arms_summary %>%
  filter(!chr_arm_to_exclude)

facets_c_arms_summary_e %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn == 0) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn == 1) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn == 2) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn >2) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn >3) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn >5) %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_summary_e %>% filter(cn >10) %>% mutate_if(is.character, as.factor) %>% summary()

```

# Arm coverage

```{r, fig.height = 4, fig.width = 4}
facets_c_arms_coverage <- facets_c_arms %>%
  mutate(covered_chr_arm = end - start) %>%
  group_by(chr_arm, Patient, CF_sample) %>% 
  summarise_at(vars(covered_chr_arm), 
               list(covered_chr_arm = sum)) %>%
  left_join(cytoband_pq_df %>%
              select(chr_arm, chr_arm_size, chr_arm_size_mb), by = "chr_arm") %>%
  mutate(covered_chr_arm_perc = covered_chr_arm / chr_arm_size * 100)

facets_c_arms_coverage %>%
  write_csv2(here(output_dir, "facets_c_arms_coverage.csv"))

facets_c_arms_coverage %>% summary()

facets_c_arms_coverage %>% group_by(chr_arm) %>% 
  summarise_at(vars(covered_chr_arm, chr_arm_size, covered_chr_arm_perc), 
               list(mean = mean)) %>%
  summary()

facets_c_arms_coverage %>% group_by(chr_arm) %>% 
  summarise_at(vars(covered_chr_arm, chr_arm_size, covered_chr_arm_perc), 
               list(mean = mean)) %>% filter(covered_chr_arm_perc_mean < 80) %>%
  mutate_if(is.character, as.factor) %>%
  summary()

facets_c_arms %>%
 filter(!chr_arm %in% chr_arms_exclude) %>% ggplot(aes(x = cn)) +
  geom_histogram() + theme_Publication() + scale_y_log10() 

facets_c_arms %>% ggplot(aes(x = log2(cn + 1))) +
  geom_histogram() + theme_Publication() + scale_y_log10() 

facets_c_arms %>%
  filter(!chr_arm %in% chr_arms_exclude) %>%
  ggplot(aes(y = log2(cn + 1), x = segment_size_mb)) +
  geom_point() + theme_Publication() 

# Trim to 4
facets_c_arms %>%
  mutate(cn = ifelse(cn > 4, 4, cn)) %>%
  ggplot(aes(x = cn, fill = Trial)) +
  geom_density( alpha = 0.5) +
  theme_Publication() + facet_wrap(~Trial)
```

# FACETS arm status

```{r, fig.height = 4, fig.width = 4}
segment_status_n_map <- data.frame(
  segment_status = c("deep_del", "loss", "normal", "gain", "amp"), cn = c(0, 1, 2, 3, 4))

facets_c_arms_p <- facets_c_arms %>%
  mutate(cn = ifelse(cn > 4, 4, cn)) %>%
  mutate(covered_chr_arm_status = end - start) %>%
  group_by(chr_arm, Patient, CF_sample, cn) %>% 
  summarise_at(vars(covered_chr_arm_status), 
               list(covered_chr_arm_status = sum)) %>%
  ungroup() %>%
  left_join(facets_c_arms_coverage) %>%
  mutate(covered_chr_arm_status_perc = covered_chr_arm_status / covered_chr_arm * 100) %>%
  left_join(segment_status_n_map)

facets_c_arms_p %>% write_csv2(here(output_dir, "facets_c_arms_p.csv"))

# Filter out chr arms
facets_c_arms_p <- facets_c_arms_p %>%
  filter(!chr_arm %in% c("chrYp", "chrYq", "chr13p", "chr14p", "chr15p", "chr21p", "chr22p")) %>%
  mutate(sample_chr_arm = paste0(CF_sample, "_", chr_arm))

facets_c_arms_p %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_p %>% mutate_if(is.character, as.factor) %>% filter(segment_status == "normal") %>% summary()
facets_c_arms_p %>% mutate_if(is.character, as.factor) %>% filter(segment_status != "normal") %>% summary()

facets_c_arms_p %>% mutate_if(is.character, as.factor) %>% filter(covered_chr_arm_perc < 80) %>% summary()
facets_c_arms_p %>% mutate_if(is.character, as.factor) %>% filter(covered_chr_arm_status_perc < 80) %>% summary()

# Change logic, take <=20% normal, as arm can have different status between gains, amp and losses and still count as aneuploid
facets_c_arms_p_normal <- facets_c_arms_p %>%
  filter(segment_status == "normal") %>%
  filter(covered_chr_arm_status_perc > 20)
         
facets_c_arms_p_normal %>%
  filter(segment_status == "normal") %>%
  mutate_if(is.character, as.factor) %>%
  summary()

facets_c_arms_p_normal %>%
  filter(segment_status != "normal") %>%
  mutate_if(is.character, as.factor) %>%
  summary()

facets_c_arms_p_normal <- facets_c_arms_p %>%
  filter(sample_chr_arm %in% facets_c_arms_p_normal$sample_chr_arm)

facets_c_arms_p_altered <- facets_c_arms_p   %>%
  filter(!sample_chr_arm %in% c(facets_c_arms_p_normal$sample_chr_arm))

# Normal segments in altered chr arms represent less than 20% of the arm
facets_c_arms_p_altered %>%
  filter(segment_status == "normal") %>% mutate_if(is.character, as.factor)  %>% summary()

facets_c_arms_p_altered %>%
  filter(segment_status != "normal") %>% mutate_if(is.character, as.factor)  %>% summary()

facets_c_arms_p_altered %>% mutate_if(is.character, as.factor) %>% summary()
facets_c_arms_p_altered %>% mutate_if(is.character, as.factor) %>% filter(covered_chr_arm_status_perc > 80) %>% summary()
facets_c_arms_p_altered %>% mutate_if(is.character, as.factor) %>% filter(covered_chr_arm_status_perc < 80) %>% summary()

# Take weighted average CN for chr with alterations in >80%, can be in different states
facets_c_arms_p_altered_weighted <- facets_c_arms_p_altered %>%
  group_by(sample_chr_arm, Patient, CF_sample, chr_arm) %>%
  summarize(cn = round(weighted.mean(cn, covered_chr_arm_status), 2)) %>%
  ungroup()

facets_c_arms_wide <- facets_c_arms_p_normal  %>%
  filter(segment_status == "normal") %>%
  select(chr_arm, Patient, CF_sample, cn) %>%
  rbind(facets_c_arms_p_altered_weighted %>%
  select(chr_arm, Patient, CF_sample, cn))%>%
  pivot_wider(values_from = cn, names_from = chr_arm) %>%
  relocate(c("chr1p", "chr1q", "chr2p", "chr2q", "chr3p", "chr3q",
             "chr4p", "chr4q","chr5p", "chr5q","chr6p", "chr6q",
             "chr7p", "chr7q", "chr8p", "chr8q", "chr9p", "chr9q",
             "chr10p", "chr10q", "chr11p", "chr11q","chr11p", "chr11q",
             "chr12p", "chr12q","chr13q", "chr14q", "chr15q",
             "chr16p", "chr16q", "chr17p", "chr17q",
             "chr18p", "chr18q","chr19p", "chr19q","chr20p", "chr20q",
             "chr21q", "chr22q","chrXp", "chrXq"),
           .after = CF_sample) %>%
  ungroup()

facets_c_arms_wide$Patient[!facets_c_arms_wide$Patient %in% dna_metadata_f$Patient]
```

# GII score

```{r}

# gii Adapted from Lee et al. Chromosomal instability confers intrinsic multidrug resistance 10.1158/0008-5472.CAN-10-3604
# > gii.fun = function(seg.mat) {
# + gii <- c()
# + bp <- c()
# + for (i in unique(seg.mat[, 1])) {
# + s <- subset(seg.mat, seg.mat[, 1] == i)
# + med <- weighted.median(as.numeric(s[, 6]), w = as.numeric(s[,
# + 4]) - as.numeric(s[, 3]))
# + b <- sum(as.numeric(s[, 4]) - as.numeric(s[, 3]))
# + a <- s[as.numeric(s[, 6]) > med | as.numeric(s[, 6] <
# + med), ]
# + gii <- c(gii, sum(as.numeric(a[, 4]) - as.numeric(a[,
# + 3]))/b)
# + bp <- c(bp, nrow(a))
# + }
# + gii2 <- cbind(gii, bp)
# + rownames(gii2) <- unique(seg.mat[, 1])
# + return(gii2)
# + }

#  Adapted from Lee et al. Chromosomal instability confers intrinsic multidrug resistance 10.1158/0008-5472.CAN-10-3604
# The function calculates the GII and BP for each segment in seg.mat and returns a matrix with these values. The GII is a measure based on the weighted median and the total length of segments, while BP counts the number of breakpoints.
adapted_gii = function(facets_c_arms) {
  facets_cnvs_df <- facets_c_arms %>% mutate(sample = CF_sample, segment_cn = cn)
  genome_integrity_index <- c()
  n_breakpoints <- c()
  median_ploidy <- c()
  for (i in unique(facets_cnvs_df$sample)) {
    s <- subset(facets_cnvs_df, facets_cnvs_df[, "sample"] == i)
    med <- weightedMedian(s$segment_cn, w = s$segment_size) # median ploidy across profile
    b <- sum(s$segment_size) # total profile size
    a <- s[(s$segment_cn > med | s$segment_cn < med), ] # filter for altered segments above or below median ploidy
    genome_integrity_index[i] <- c(sum(as.numeric(a$segment_size)/b))
    n_breakpoints[i] <- c(nrow(a)) # n altered segments, used as breakpoints
    median_ploidy[i] <- c(med)
  }
  
  gii2 <- cbind(genome_integrity_index, n_breakpoints, median_ploidy)
  rownames(gii2) <- unique(facets_cnvs_df$sample)
  return(gii2)
}

adapted_gii_df <- adapted_gii(facets_c_arms) %>%
  as.data.frame() %>%
  rownames_to_column("CF_sample")

adapted_gii_df %>% summary()
adapted_gii_df %>% write_csv2(here(output_dir, "adapted_gii_df.csv"))

```

# wgII score

```{r}
# Two definitions, one in % of bases affected and one in % of SNPs affected. Both would be called aberrant ploidy based on a difference of 0.6 from the median profile

# Replication stress links structural and numerical cancer chromosomal instability Burrel et al. 2013 # https://doi.org/10.1038/nature11935
# The genome instability index (GII)33 is the percentage of SNPs across the genome present at an aberrant copy number, relative to the baseline ploidy of the sample. We adapted the GII to account for variation in chromosome size, so that large chromosomes do not have a greater effect on the score than small chromosomes: the percentage of aberrant SNPs for each chromosome was calculated separately, and the mean percentage aberration was then calculated across all 22 chromosomes. To define a threshold for CIN− versus CIN+, the weighted GII was calculated for the cell lines. A threshold of 0.2 accurately distinguished CIN+ from CIN−, as previously defined28. The same threshold was then applied to the TCGA cohort of tumours.

# Loss and gains were defined as copy numbers deviating from the ploidy, as estimated by ASCAT, by more than 0.6, similarly to the original ASCAT publication 
# Weighted Genomic Instability Index (wGII)  https://pathsocjournals.onlinelibrary.wiley.com/doi/10.1002/path.4214
# The wGII score 31 is computed on each chromosome as the proportion of bases whose copy number deviates from the ploidy value of the sample given by ASCAT by more than 0.6. The # sample score is the sum of the chromosomal scores divided by the number of analysed chromosomes (n = 22).

# wgII_score segments
wgII_scores <- facets_c_arms %>%
  # filter(!chr %in% c("X", "Y")) %>%
  unique() %>%
  left_join(facets_data_metrics %>%
              mutate(CF_sample = CF_sample_tumor) %>%
              select(CF_sample, Ploidy) %>%
              unique()) %>% 
  mutate(segment_cn = cn,
         segment_delta = abs(segment_cn - Ploidy)) %>%
  relocate(c(Ploidy, segment_cn, segment_delta, segment_size),
           .after = Response)

# wgII_score segments
# % base definition, using segment size
# Also % snp definition, based on number of snps falling within aberrant segments
wgII_scores_0.6 <- wgII_scores %>%
  mutate(count_for_wgII = segment_delta > 0.6) %>% 
  left_join(cytoband_pq_df %>% select(chr_arm, chr)) %>%
  group_by(chr, Patient, CF_sample) %>%
  mutate(altered_segment_delta = sum(
    segment_size * count_for_wgII, na.rm = TRUE),
         chr_size_covered = sum(segment_size)) %>%
  ungroup()

wgII_scores_0.6_chr <- wgII_scores_0.6 %>%
  select(Patient, CF_sample, chr, chr_size_covered, altered_segment_delta) %>%
  unique() %>%
  left_join(chromosome_sizes) %>%
  mutate(perc_chr_covered = chr_size_covered / chr_size) %>%
  mutate(perc_covered_altered = altered_segment_delta / chr_size_covered)

wgII_scores_0.6_sample <- wgII_scores_0.6_chr %>%
  group_by(Patient, CF_sample) %>%
  summarize(wgII_0.6_bases = mean(perc_covered_altered)) %>%
  ungroup()

wgII_scores_0.6_sample %>% factor_summary()

wgII_scores %>%
  write_csv2(here(output_dir, "wgII_scores.csv"))

wgII_scores_0.6_chr %>%
  write_csv2(here(output_dir, "wgII_scores_0.6_chr.csv"))

wgII_scores_0.6_sample %>%
  write_csv2(here(output_dir, "wgII_scores_0.6_sample.csv"))

```

# Calculate Aneuploidy score

```{r, fig.height = 4, fig.width = 4}
facets_aneuploidy_score_m <- facets_c_arms_wide %>%
  select(-c(Patient, chrXp, chrXq)) %>%
  column_to_rownames("CF_sample") %>%
  as.matrix()
  
dim(facets_aneuploidy_score_m)
facets_aneuploidy_score_m[is.na(facets_aneuploidy_score_m)] <- 2
facets_aneuploidy_score_m[facets_aneuploidy_score_m != 2] <- 1
facets_aneuploidy_score_m[facets_aneuploidy_score_m == 2] <- 0

facets_aneuploidy_score <- facets_aneuploidy_score_m %>%
  rowSums() %>%
  as.data.frame() %>%
  setNames("Aneuploidy_score") %>%
  rownames_to_column("CF_sample")

facets_scna_loss_m <- facets_c_arms_wide %>%
  select(-c(Patient, chrXp, chrXq)) %>%
  column_to_rownames("CF_sample") %>%
  as.matrix()
  
dim(facets_scna_loss_m)
cnvkit_cnvs_scna_loss_m[is.na(facets_scna_loss_m)] <- 2
facets_scna_loss_m[facets_scna_loss_m < 2] <- 1
facets_scna_loss_m[facets_scna_loss_m > 2] <- 0
facets_scna_loss_m[facets_scna_loss_m == 2] <- 0

facets_scna_loss <- facets_scna_loss_m %>%
  rowSums() %>%
  as.data.frame() %>%
  setNames("scna_loss") %>%
  rownames_to_column("CF_sample")

facets_aneuploidy_score_df <- facets_c_arms_wide %>%
  left_join(facets_aneuploidy_score) %>%
  left_join(facets_scna_loss) %>%
  mutate(CF_sample_tumor = CF_sample) %>%
  left_join(dna_metadata_f %>%
              select(CF_sample_tumor, Response, Trial, Cohort)) %>%
  relocate(c(Response, Response, Trial, Cohort), .after = CF_sample)

facets_aneuploidy_score_df <- facets_aneuploidy_score_df %>%
  left_join(adapted_gii_df)%>%
  left_join(wgII_scores_0.6_sample) %>%
  mutate(CIN = ifelse(wgII_0.6_bases > 0.2, "CIN+", "CIN-"))

facets_aneuploidy_score_df$CIN %>% table()
table(facets_aneuploidy_score_df$Trial, facets_aneuploidy_score_df$CIN)
table(facets_aneuploidy_score_df$Trial, facets_aneuploidy_score_df$CIN) %>%
  prop.table(margin = 1) %>% round(2)

facets_aneuploidy_score_df %>%
  write_csv2(here(output_dir, "facets_aneuploidy_score_df.csv"))
```

### Compare aneuploidy scores

```{r, fig.height = 4, fig.width = 4}
facets_aneuploidy_score_df <- read_csv2(
  here(output_dir, "facets_aneuploidy_score_df.csv"))

cnvkit_cnvs_aneuploidy_score_df <- read.csv2(
  here(output_dir, "cnvkit_cnvs_aneuploidy_score_df.csv"))

facets_aneuploidy_score_df %>% filter(is.na(Response))
n_distinct(facets_aneuploidy_score_df$CF_sample)

aneuploidy_score_df_c <- cnvkit_cnvs_aneuploidy_score_df %>%
  select(Patient, Trial, Response, CF_sample_tumor,
         Aneuploidy_score, scna_loss) %>%
  left_join(facets_aneuploidy_score_df %>%
              select(Aneuploidy_score, scna_loss, CF_sample_tumor,
                     genome_integrity_index, wgII_0.6_bases),
            by = "CF_sample_tumor", suffix = c("_cnvkit", "_facets")) 

aneuploidy_score_df_c %>%
  ggplot(aes(x = Aneuploidy_score_cnvkit,
             y = Aneuploidy_score_facets, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors) +
  geom_abline(slope = 1, intercept = 0) +
  guides(color = guide_legend(nrow = 4))

aneuploidy_score_df_c %>%
  ggplot(aes(x = Aneuploidy_score_cnvkit,
             y = Aneuploidy_score_facets, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors) +
  geom_abline(slope = 1, intercept = 0) +
  guides(color = guide_legend(nrow = 4)) +
  facet_wrap(~Response)

aneuploidy_score_df_c %>%
  filter(Trial == "NICHE-MSS") %>%
  ggplot(aes(x = Aneuploidy_score_cnvkit,
             y = Aneuploidy_score_facets, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors) +
  geom_abline(slope = 1, intercept = 0) +
  guides(color = guide_legend(nrow = 4)) +
  facet_wrap(~Response)

aneuploidy_score_df_c %>%
  ggplot(aes(x = scna_loss_cnvkit,
             y = scna_loss_facets, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors) +
  geom_abline(slope = 1, intercept = 0) +
  guides(color = guide_legend(nrow = 4))


aneuploidy_score_df_c %>%
  filter(!is.na(Trial)) %>%
  ggplot(aes(x = Aneuploidy_score_facets,
             y = genome_integrity_index, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors)+
  guides(color = guide_legend(nrow = 4))

aneuploidy_score_df_c %>%
  filter(!is.na(Trial)) %>%
  ggplot(aes(x = Aneuploidy_score_facets,
             y = wgII_0.6_bases, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors) +
  guides(color = guide_legend(nrow = 4))

aneuploidy_score_df_c %>%
  filter(!is.na(Trial)) %>%
  ggplot(aes(x = genome_integrity_index,
             y = wgII_0.6_bases, col = Trial)) +
  geom_point() + theme_Publication() +
  scale_colour_manual(values = trial_colors) +
  guides(color = guide_legend(nrow = 4))

aneuploidy_score_df_c %>%
  select(Aneuploidy_score_cnvkit, Aneuploidy_score_facets, wgII_0.6_bases,
         genome_integrity_index) %>% as.matrix() %>%
  cor(use = "complete.obs")

aneuploidy_score_df_c %>%
  select(Aneuploidy_score_cnvkit, Aneuploidy_score_facets, wgII_0.6_bases,
         genome_integrity_index) %>% as.matrix() %>%
  cor(use = "complete.obs") %>% corrplot() 
```

# Aneuploidy Response

```{r, fig.height = 4, fig.width = 4}

facets_aneuploidy_score_df_excl_dmmr <- facets_aneuploidy_score_df %>%
  filter(!Trial %in% c("NICHE-MSI"))

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Trial, y = Aneuploidy_score, fill = Trial)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") + coord_flip() +
  scale_fill_manual(values = trial_colors)

stat.test <- compare_means(formula = Aneuploidy_score ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df,
                           paired = FALSE)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Response, y = Aneuploidy_score)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 42,
                     label = "wilcoxon, p = {p.format}") 


stat.test <- compare_means(formula = Aneuploidy_score ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df_excl_dmmr,
                           paired = FALSE)

facets_aneuploidy_score_df_excl_dmmr %>%
  ggplot(aes(x = Response, y = Aneuploidy_score)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "", title = "Excluding dMMR") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 42,
                     label = "wilcoxon, p = {p.format}") 

stat.test <- compare_means(formula = Aneuploidy_score ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df, paired = FALSE)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Response, y = Aneuploidy_score)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 42,
                     label = "wilcoxon, p = {p.format}") 

```

# wgII Response

```{r, fig.height = 4, fig.width = 4}
facets_aneuploidy_score_df %>%
  ggplot(aes(x = Trial, y = wgII_0.6_bases, fill = Trial)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") + coord_flip() +
  scale_fill_manual(values = trial_colors)

stat.test <- compare_means(formula = wgII_0.6_bases ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df, paired = FALSE)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Response, y = wgII_0.6_bases)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 1,
                     label = "wilcoxon, p = {p.format}") 


stat.test <- compare_means(formula = wgII_0.6_bases ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df_excl_dmmr,
                           paired = FALSE)

facets_aneuploidy_score_df_excl_dmmr %>%
  ggplot(aes(x = Response, y = wgII_0.6_bases)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "", title = "Excluding dMMR") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 1,
                     label = "wilcoxon, p = {p.format}") 

table(facets_aneuploidy_score_df$Response, facets_aneuploidy_score_df$CIN)
fisher.test(table(facets_aneuploidy_score_df$Response,
                  facets_aneuploidy_score_df$CIN))

table(facets_aneuploidy_score_df_excl_dmmr$Response,
      facets_aneuploidy_score_df_excl_dmmr$CIN)
fisher.test(table(facets_aneuploidy_score_df_excl_dmmr$Response,
                  facets_aneuploidy_score_df_excl_dmmr$CIN))

```

```{r, fig.height = 6, fig.width = 6}
facets_aneuploidy_score_df %>%
  ggplot(aes(x = Response, y = Aneuploidy_score, fill = Response)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  facet_wrap(~Trial) +
  scale_fill_manual(values = c(Color_NR, Color_R))
```

```{r, fig.height = 6, fig.width = 6}
facets_aneuploidy_score_df %>%
  left_join(n_variants_dp20_vaf_tmb) %>%
  ggplot(aes(x = Aneuploidy_score, y = TMB, col = Trial)) +
  geom_point() +
  theme_Publication() +
  scale_color_manual(values = trial_colors)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Aneuploidy_score, fill = Trial)) +
  geom_density(aes(alpha = 0.5)) +
  theme_Publication() +
  guides(alpha = FALSE) +
  scale_fill_manual(values = trial_colors)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Aneuploidy_score, fill = Trial)) +
  geom_histogram(aes()) +
  theme_Publication() +
  guides(alpha = FALSE) +
  facet_wrap(~Trial, scale = "free_y") +
  scale_fill_manual(values = trial_colors) +
  expand_limits(y = c(0, 10))
```

# SCNA loss burden

```{r, fig.height = 4, fig.width = 4}
stat.test <- compare_means(formula = scna_loss ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df, paired = FALSE)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Response, y = scna_loss)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 30,
                     label = "wilcoxon, p = {p.format}") 

stat.test <- compare_means(formula = scna_loss ~ Response,
                           method = "wilcox.test",
                           data = facets_aneuploidy_score_df %>%
  filter(!Trial %in% c("NICHE-MSI")), paired = FALSE)

facets_aneuploidy_score_df %>%
  filter(!Trial %in% c("NICHE-MSI")) %>%
  ggplot(aes(x = Response, y = scna_loss)) +
  geom_boxplot(aes(fill = Response)) +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "", title = "Excluding dMMR") +
  scale_fill_manual(values = c(Color_NR, Color_R))  +
  stat_pvalue_manual(stat.test, y.position = 30,
                     label = "wilcoxon, p = {p.format}") 

facets_aneuploidy_score_df %>%
  ggplot(aes(x = Trial, y = scna_loss, fill = Trial)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") + coord_flip() +
  scale_fill_manual(values = trial_colors)
```

```{r, fig.height = 6, fig.width = 6}
facets_aneuploidy_score_df %>%
  ggplot(aes(x = Response, y = scna_loss, fill = Response)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  facet_wrap(~Trial) +
  scale_fill_manual(values = c(Color_NR, Color_R))
```

```{r, fig.width = 6, fig.height = 6}
facets_aneuploidy_score_df %>%
  filter(Aneuploidy_score > 15, scna_loss < 5) %>%
  factor_summary()


facets_aneuploidy_score_df  %>%
  ggplot(aes(x = Response, y = Aneuploidy_score, fill = Response)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  facet_wrap(~Trial) +
  scale_fill_manual(values = c(Color_NR, Color_R))

facets_aneuploidy_score_df %>%
  filter(scna_loss < 5) %>%
  ggplot(aes(x = Response, y = Aneuploidy_score, fill = Response)) +
  geom_boxplot() + geom_point() +
  theme_Publication() +
  guides(fill = "none") +
  labs(x = "") +
  facet_wrap(~Trial) +
  scale_fill_manual(values = c(Color_NR, Color_R))

```

```{r, fig.height = 6, fig.width = 6}
facets_aneuploidy_score_df %>%
  left_join(n_variants_dp20_vaf_tmb) %>%
  ggplot(aes(x = scna_loss, y = TMB, col = Trial)) +
  geom_point() +
  theme_Publication() +
  scale_color_manual(values = trial_colors)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = scna_loss, fill = Trial)) +
  geom_density(aes(alpha = 0.5)) +
  theme_Publication() +
  guides(alpha = FALSE) +
  scale_fill_manual(values = trial_colors)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = scna_loss, fill = Trial)) +
  geom_histogram(aes()) +
  theme_Publication() +
  guides(alpha = FALSE) +
  facet_wrap(~Trial) +
  scale_fill_manual(values = trial_colors)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = scna_loss, y = Aneuploidy_score, col = Trial)) +
  geom_point() +
  theme_Publication() +
  guides(alpha = FALSE) +
  facet_wrap(~Trial) +
  scale_color_manual(values = trial_colors) +
  geom_abline(slope = 1, intercept = 0)

facets_aneuploidy_score_df %>%
  ggplot(aes(x = scna_loss, y = Aneuploidy_score)) +
  geom_pointdensity() +
  theme_Publication() +
  guides(alpha = FALSE) +
  geom_abline(slope = 1, intercept = 0) +
  guides(colour = guide_colourbar(barwidth = 10))
```

# Arms Response Frequency

```{r, fig.width = 19, fig.height = 11}
cnvkit_cnvs_c_arms_p <- cnvkit_cnvs_c_arms_p %>%
  left_join(dna_metadata_f %>%
              mutate(CF_sample = CF_sample_tumor) %>%
              select(Patient, Trial, CF_sample, CF_sample_tumor) %>%
              unique())

chr_arm_status_df <- table(cnvkit_cnvs_c_arms_p$CF_sample_tumor,
                               cnvkit_cnvs_c_arms_p$Trial,
                           cnvkit_cnvs_c_arms_p$chr_arm,
              cnvkit_cnvs_c_arms_p$segment_status) %>%
  as.data.frame() %>%
  setNames(c("CF_sample_tumor", "Trial", "chr_arm", "segment_status",  "Freq")) %>%
  filter(Freq > 0)

chr_arm_status_prop_df <- chr_arm_status_df %>%
  group_by(Trial, chr_arm, segment_status) %>%
  summarize(Freq = sum(Freq)) %>%
  group_by(chr_arm, Trial) %>%
  add_count(name = "Total_segments", wt = Freq) %>%
  mutate(Proportion = Freq / Total_segments)

chr_arm_levels <- c("chr1p", "chr1q", "chr2p", "chr2q", "chr3p", "chr3q",
             "chr4p", "chr4q","chr5p", "chr5q","chr6p", "chr6q",
             "chr7p", "chr7q", "chr8p", "chr8q", "chr9p", "chr9q",
             "chr10p", "chr10q", "chr11p", "chr11q", 
             "chr12p", "chr12q","chr13q", "chr14q", "chr15q",
             "chr16p", "chr16q", "chr17p", "chr17q",
             "chr18p", "chr18q", "chr19p", "chr19q", "chr20p", "chr20q",
             "chr21q", "chr22q","chrXp", "chrXq")

# TODO these are stacked proportions per trial, so added proportions do not make sense
# chr_arm_status_prop_df %>%
#   filter(Trial == "BELLINI") %>%
#   ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion)) +
#   geom_col() +
#   facet_wrap(~segment_status) +
#   theme_Publication() + labs(x = "chr_arm", y = "Proportion of Arm segments") +
#     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

chr_arm_status_prop_df %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Freq, fill = Trial)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Freq of Arm segments") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = trial_colors)

# Proportions per trial
chr_arm_status_prop_df %>%
  filter(segment_status == "amp") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Proportion, fill = Trial)) +
  geom_col() +
  facet_wrap(~Trial, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of Arm segments amplifications") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = trial_colors)

chr_arm_status_prop_df %>%
  filter(segment_status == "gain") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion, fill = Trial)) +
  geom_col() +
  facet_wrap(~Trial, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of Arm segments gains") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = trial_colors)

chr_arm_status_prop_df %>%
  filter(segment_status == "loss") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion, fill = Trial)) +
  geom_col() +
  facet_wrap(~Trial, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of Arm segments losses") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = trial_colors)

chr_arm_status_prop_df %>%
  filter(segment_status == "deep_del") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion, fill = Trial)) +
  geom_col() +
  facet_wrap(~Trial, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of Arm segments Deep deletions") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))+
  scale_fill_manual(values = trial_colors)

chr_arm_status_prop_df %>%
  filter(segment_status == "gain") %>%
  arrange(-Proportion)

chr_arm_status_prop_df %>%
  filter(segment_status == "gain") %>%
  arrange(-Proportion)

chr_arm_status_prop_df %>%
  filter(segment_status == "gain") %>%
  arrange(-Proportion) %>% 
  ggplot(aes(x = reorder(chr_arm, Proportion), y = Proportion)) +
  geom_col() +
  coord_flip()+
  theme_Publication() +
  labs(x = "chr_arm", y = "Proportion of segments gains")

chr_arm_status_prop_df %>%
  filter(segment_status == "loss") %>%
  arrange(-Proportion)

chr_arm_status_prop_df %>%
  filter(segment_status == "loss") %>%
  arrange(-Proportion) %>% 
  ggplot(aes(x = reorder(chr_arm, Proportion), y = Proportion)) +
  geom_col() +
  coord_flip() +
  theme_Publication() +
  labs(x = "chr_arm", y = "Proportion of segments losses")
```

```{r, fig.width = 16, fig.height = 9}
chr_arm_status_response_df <- chr_arm_status_df %>%
  left_join(dna_metadata_f %>%
              select(CF_sample_tumor, Response) %>% unique()) %>%
  group_by(Response, chr_arm, segment_status) %>%
  summarize(Freq = sum(Freq)) %>%
  group_by(chr_arm, Response) %>%
  add_count(name = "Total_arm_segments", wt = Freq) %>%
  mutate(Proportion = Freq / Total_arm_segments)

chr_arm_status_response_df %>%
  filter(Response == "R") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of arm segments") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

chr_arm_status_response_df %>%
  filter(Response == "NR") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of arm segments") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

chr_arm_status_delta_df <- chr_arm_status_response_df %>%
  mutate(chr_arm_status = sprintf("%s_%s", chr_arm, segment_status)) %>%
  select(chr_arm_status, segment_status, Response, Proportion) %>%
  pivot_wider(names_from = Response, values_from = Proportion) %>%
  mutate(R = ifelse(is.na(R), 0 , R)) %>%
  mutate(NR = ifelse(is.na(NR), 0 , NR)) %>%
  mutate(Response_delta = R - NR)

chr_arm_status_delta_df %>%
  write_csv2(here(output_dir, "chr_arm_status_delta_df.csv"))

chr_arm_status_delta_df[
  select_top_extremes(chr_arm_status_delta_df$Response_delta, top_n = 20),
  ] %>% arrange(Response_delta)

chr_arm_status_delta_df_altered <- chr_arm_status_delta_df %>% filter(!segment_status == "normal")
chr_arm_status_delta_df_altered %>% filter(chr_arm == "chr9p")
chr_arm_status_delta_df_altered[
  select_top_extremes(chr_arm_status_delta_df_altered$Response_delta, top_n = 20),
  ] %>% arrange(Response_delta)


```

```{r, fig.width = 16, fig.height = 16}
chr_arm_status_levels <- c(
  paste0(chr_arm_levels, "_amp"),
  paste0(chr_arm_levels, "_gain"),
  paste0(chr_arm_levels, "_loss"),
  paste0(chr_arm_levels, "_deep_del"),
  paste0(chr_arm_levels, "_normal"))

chr_arm_status_delta_df %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR), values_to = "Freq", names_to = "Response") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Freq, fill = Response)) + geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + coord_flip() 

chr_arm_status_delta_df$chr_arm <- factor(chr_arm_status_delta_df$chr_arm,
                                          levels = chr_arm_levels)

chr_arm_status_delta_df %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR), values_to = "Freq", names_to = "Response") %>%
  mutate(Percentage = Freq * 100) %>%
  ggplot(aes(x = chr_arm,
             y = Percentage, fill = Response)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + coord_flip()

chr_arm_status_delta_df %>%
  filter(segment_status != "normal") %>%
  mutate(Response_delta = 100 * Response_delta) %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Response_delta, fill = Response_delta > 0)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free_y") +
  theme_Publication() + coord_flip() +
  scale_fill_manual(values = c(Color_NR, Color_R))

chr_arm_status_delta_df %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR), values_to = "Freq", names_to = "Response") %>%
  mutate(Percentage = Freq * 100) %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Percentage, fill = Response)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free_y") +
  theme_Publication() + coord_flip() +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  labs(x = "")

chr_arm_status_delta_df %>%
  mutate(Response_delta = 100 * Response_delta)  %>%
  ggplot(aes(x = factor(chr_arm_status,
                        levels = (chr_arm_status_levels)), y = Response_delta,
                    fill = Response_delta > 0)) +
  geom_col() + theme_Publication() + coord_flip() +
  facet_wrap(~segment_status, scales = "free_y") +
  labs(x = "", y = "Response Delta (Arm segments % in R - NR)") +
  scale_fill_manual(values = c(Color_NR, Color_R))

```

# Condense gains and amplifications, and (deep) dels

```{r, fig.width = 16, fig.height = 9}

chr_arm_status_simplified_response_df <- chr_arm_status_df %>%
  left_join(dna_metadata_f %>%
              select(CF_sample_tumor, Response) %>% unique()) %>%
  mutate(segment_status = as.character(segment_status)) %>%
  mutate(segment_status = ifelse(segment_status == "amp", "gain", segment_status)) %>%
  mutate(segment_status = ifelse(segment_status == "deep_del", "loss", segment_status)) %>%
  group_by(Response, chr_arm, segment_status) %>%
  summarize(Freq = sum(Freq)) %>%
  group_by(chr_arm, Response) %>%
  add_count(name = "Total_arm_segments", wt = Freq) %>%
  mutate(Proportion = Freq / Total_arm_segments) %>%
  ungroup()

chr_arm_status_simplified_response_df %>%
  filter(Response == "R") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of arm segments") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

chr_arm_status_simplified_response_df %>%
  filter(Response == "NR") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels), y = Proportion)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + labs(x = "chr_arm", y = "Proportion of arm segments") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

chr_arm_status_simplified_delta_df <- chr_arm_status_simplified_response_df %>%
  mutate(chr_arm_status = sprintf("%s_%s", chr_arm, segment_status)) %>%
  select(chr_arm, chr_arm_status, segment_status, Response, Proportion) %>%
  pivot_wider(names_from = Response, values_from = Proportion) %>%
  mutate(R = ifelse(is.na(R), 0 , R)) %>%
  mutate(NR = ifelse(is.na(NR), 0 , NR)) %>%
  mutate(Response_delta = R - NR)

chr_arm_status_simplified_delta_df %>%
  write_csv2(here(output_dir, "chr_arm_status_simplified_delta_df.csv"))

chr_arm_status_simplified_delta_df[
  select_top_extremes(chr_arm_status_simplified_delta_df$Response_delta, top_n = 20),
  ] %>% arrange(Response_delta)

chr_arm_status_simplified_delta_df_altered <- chr_arm_status_simplified_delta_df %>% filter(!segment_status == "normal")
chr_arm_status_simplified_delta_df_altered %>% filter(chr_arm == "chr9p")
chr_arm_status_delta_df_altered[
  select_top_extremes(chr_arm_status_delta_df_altered$Response_delta, top_n = 20),
  ] %>% arrange(Response_delta)

```

```{r, fig.width = 16, fig.height = 16}
chr_arm_status_simplified_levels <- c(
  paste0(chr_arm_levels, "_gain"),
  paste0(chr_arm_levels, "_loss"),
  paste0(chr_arm_levels, "_normal"))

chr_arm_status_simplified_delta_df %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR), values_to = "Freq", names_to = "Response") %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Freq, fill = Response)) + geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + coord_flip() 

chr_arm_status_simplified_delta_df$chr_arm <- factor(chr_arm_status_simplified_delta_df$chr_arm, levels = chr_arm_status_simplified_levels)

chr_arm_status_simplified_delta_df %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR), values_to = "Freq", names_to = "Response") %>%
  mutate(Percentage = Freq * 100) %>%
  ggplot(aes(x = chr_arm,
             y = Percentage, fill = Response)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free") +
  theme_Publication() + coord_flip()

chr_arm_status_simplified_delta_df %>%
  filter(segment_status != "normal") %>%
  mutate(Response_delta = 100 * Response_delta) %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Response_delta, fill = Response_delta > 0)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free_y") +
  theme_Publication() + coord_flip() +
  scale_fill_manual(values = c(Color_NR, Color_R))

chr_arm_status_simplified_delta_df %>%
  mutate(NR = -NR) %>%
  pivot_longer(cols = c(R, NR), values_to = "Freq", names_to = "Response") %>%
  mutate(Percentage = Freq * 100) %>%
  ggplot(aes(x = factor(chr_arm, levels = chr_arm_levels),
             y = Percentage, fill = Response)) +
  geom_col() +
  facet_wrap(~segment_status, scales = "free_y") +
  theme_Publication() + coord_flip() +
  scale_fill_manual(values = c(Color_NR, Color_R)) +
  labs(x = "")

chr_arm_status_simplified_delta_df %>%
  
  mutate(Response_delta = 100 * Response_delta)  %>%
  ggplot(aes(x = factor(chr_arm_status,
                        levels = (chr_arm_status_levels)), y = Response_delta,
                    fill = Response_delta > 0)) +
  geom_col() + theme_Publication() + coord_flip() +
  facet_wrap(~segment_status, scales = "free_y") +
  labs(x = "", y = "Response Delta (Arm segments % in R - NR)") +
  scale_fill_manual(values = c(Color_NR, Color_R))

chr_arm_status_simplified_delta_df%>%
  filter(segment_status != "normal") %>%
  mutate(Response_delta = 100 * Response_delta)  %>%
  ggplot(aes(x = factor(chr_arm_status,
                        levels = (chr_arm_status_levels)), y = Response_delta,
                    fill = Response_delta > 0)) +
  geom_col() + theme_Publication() + coord_flip() +
  facet_wrap(~segment_status, scales = "free_y") +
  labs(x = "", y = "Response Delta (Arm segments % in R - NR)") +
  scale_fill_manual(values = c(Color_NR, Color_R))

```

# ----------

# check for specific gains and (deep) deletions

## FACETS

```{r}
facets_c_cytobands <- facets_c_cytobands %>%
  left_join(cytoband_pq_df %>% select(chr, chr_arm)) %>%
  mutate(chr_arm_cytoband = paste0(chr, "_", cytoband))
```

```{r, fig.width = 24, fig.height = 5}
facets_c_cytobands %>% filter(chr_arm == NA)
facets_c_cytobands %>% filter(type == "DEL") %>% view() 

facets_c_cytobands$type %>% table()
facets_c_cytobands %>% filter(type == "DEL") %>% pull(chr_arm) %>% table() %>% as.data.frame() %>% arrange(-Freq) %>% setNames(c("chr_arm", "deletions"))
facets_c_cytobands %>% filter(type == "DEL") %>% pull(chr_arm_cytoband) %>% table() %>% as.data.frame() %>% arrange(-Freq) %>% setNames(c("cytoband", "deletions"))

cytoband_pq_df %>% select(chr, chr_arm)

facets_c_cytobands_excl_dMMR_mean_cn <- facets_c_cytobands %>%
  filter(Trial != "NICHE-MSI") %>%
  group_by(chr_arm_cytoband, cytoband, chr, Response) %>%
  summarize(mean_cn = mean(cn)) %>%
  filter(chr_arm_cytoband != "NA_NA") %>%
  left_join(cytoband_df) 

facets_c_cytobands_excl_dMMR_mean_cn  %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_mean_cn.csv"))

cytoband_order <- cytoband_df %>%
  filter(!chr %in% c("chrX", "chrY")) %>%
  mutate(chr_arm_cytoband = paste0(chr, "_", cytoband))
cytoband_order$chr <- as.factor(
  cytoband_order$chr, levels = c("chr1", "chr2", "chr3",  "chr4", "chr5",
                                 "chr6", "chr7", "chr8", "chr9", "chr10",
                                 "chr11", "chr12", "chr13", "chr14", "chr15",
                                 "chr16", "chr17", "chr18", "chr19", "chr20",
                                 "chr21", "chr22"))

cytoband_order$chrom_num <- as.numeric(gsub("chr", "", cytoband_order$chr))

cytoband_order <- cytoband_order[order(cytoband_order$chrom_num), ]

cytoband_order_f <- cytoband_order %>%
  filter(chr_arm_cytoband %in% facets_c_cytobands_excl_dMMR_mean_cn$chr_arm_cytoband)

facets_c_cytobands_excl_dMMR_mean_cn$chr_arm_cytoband <- factor(
  facets_c_cytobands_excl_dMMR_mean_cn$chr_arm_cytoband,
  levels = cytoband_order_f$chr_arm_cytoband)

chr_label_positions <- facets_c_cytobands_excl_dMMR_mean_cn %>%
 group_by(chr) %>%
 slice(1) %>%
 ungroup()

chr_label_positions <- facets_c_cytobands_excl_dMMR_mean_cn[
  !duplicated(facets_c_cytobands_excl_dMMR_mean_cn$chr), ] %>%
  filter(!chr %in% c("chrX", "chrY"))

facets_c_cytobands_excl_dMMR_mean_cn %>%
  filter(!chr %in% c("chrX", "chrY")) %>%
  ggplot(aes(x = chr_arm_cytoband, y = mean_cn, col = Response)) +
  geom_point() + theme_Publication() +
  labs(x = "", y = "Average cytoband copy number") +
  scale_color_manual(values = c(Color_NR, Color_R)) +
  theme(axis.text.x = element_blank()) +
geom_text(data = chr_label_positions, aes(y = 4, label = chr),
          col = "black", vjust = -1.5, size = 3.5) 


```

# CDK2NA

```{r}
# CDKN2A  21,967,752 to 21,995,301
# TRAF2  136,881,912 - 136,926,607 (GRCh38/hg38) q34.3 
# CCND1 11q13.3
facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_p21.3") %>% view()
facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_p21.3") %>% select(cytoband, start, end, cn, type, cnlr.median) %>% factor_summary()
facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_q34.3") %>% view()
facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_q34.3") %>% select(cytoband, start, end, cn, type, cnlr.median) %>% factor_summary()
facets_c_cytobands %>% filter(chr_arm_cytoband == "chr11_q13.3") %>% select(cytoband, start, end, cn, type, cnlr.median) %>% factor_summary()

facets_c_cytobands %>% filter(chr == "chr9") %>% pull(chr_arm_cytoband)  %>% unique()

facets_chr9_p21.3_loss <- facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_p21.3") %>% filter(type %in% c("DEL", "HEMIZYG"))

facets_chr9_q34.3_loss <- facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_q34.3") %>% filter(type %in% c("DEL", "HEMIZYG"))

facets_chr9_q34.3_loss <- facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_q34.3") %>% filter(type %in% c("DEL", "HEMIZYG"))

facets_chr11_q13.3_gain <- facets_c_cytobands %>% filter(chr_arm_cytoband == "chr11_q13.3") %>% filter(type %in% c("DUP"))

facets_c_cytobands %>% filter(chr_arm_cytoband == "chr11_q13.3") %>% select(cytoband, start, end, cn, type, cnlr.median) %>% factor_summary()

facets_c_cytobands %>% filter(chr_arm_cytoband == "chr9_q34.3") %>% filter(type %in% c("DEL", "HEMIZYG")) %>% pull(CF_sample)

facets_amp_del_df <- facets_aneuploidy_score_df %>%
  mutate(chr9_q34_3_loss = CF_sample %in% facets_chr9_q34.3_loss$CF_sample,
         chr9_p21.3_loss = CF_sample %in% facets_chr9_p21.3_loss$CF_sample,
         chr11_q13.3_gain = CF_sample %in% facets_chr11_q13.3_gain$CF_sample)

table(facets_amp_del_df$chr9_q34_3_loss)
table(facets_amp_del_df$chr9_p21.3_loss)

table(facets_amp_del_df$chr9_q34_3_loss, facets_amp_del_df$chr9_p21.3_loss) %>%
    `rownames<-`( c("no chr9_q34_3_loss", "chr9_q34_3_loss")) %>%
    `colnames<-`( c("no chr9_p21.3_loss", "chr9_p21.3_loss"))
table(facets_amp_del_df$chr9_q34_3_loss, facets_amp_del_df$chr9_p21.3_loss) %>% prop.table() %>% round(2) %>%
    `rownames<-`( c("no chr9_q34_3_loss", "chr9_q34_3_loss")) %>%
    `colnames<-`( c("no chr9_p21.3_loss", "chr9_p21.3_loss"))

dim(facets_amp_del_df)
facets_amp_del_df %>% select(Patient, CF_sample, chr9p, chr9q, chr9_q34_3_loss, facets_chr9_p21.3_loss)
facets_amp_del_df %>% view()

table(facets_amp_del_df$Response, facets_amp_del_df$chr9_q34_3_loss)
table(facets_amp_del_df$Response, facets_amp_del_df$chr9_q34_3_loss) %>% prop.table(margin = 1) %>% round(2)
table(facets_amp_del_df$Response, facets_amp_del_df$chr9_q34_3_loss) %>% prop.table(margin = 2) %>% round(2)

fisher.test(table(facets_amp_del_df$Response, facets_amp_del_df$chr9_q34_3_loss))

facets_amp_del_df %>% write_csv2(here(output_dir, "facets_amp_del_df.csv"))
```

# Excluding dMMR
```{r}
facets_c_cytobands_excl_dMMR %>%
  ggplot(aes(x = log2(cn + 1))) +
  geom_density() + theme_Publication()
facets_c_cytobands_excl_dMMR %>%
  filter(cn > 2) %>% ggplot(aes(x = log2(cn))) +
  geom_density() + theme_Publication()
```

```{r}
facets_c_cytobands_excl_dMMR %>% pull(cn) %>% summary()
facets_c_cytobands_excl_dMMR %>% pull(cn) %>% summary()

facets_c_cytobands_excl_dMMR <- facets_c_cytobands %>%
  filter(!Trial %in% c("NICHE-MSI"))

facets_c_cytobands_excl_dMMR %>% select(Patient, Response) %>% unique() %>% pull(Response) %>% table()

facets_c_cytobands_excl_dMMR_sample_size <- facets_c_cytobands_excl_dMMR %>%
  select(Patient, Response) %>% unique() %>%
  pull(Response) %>% table() %>% as.data.frame() %>%
  setNames(c("Response", "Total"))

facets_c_cytobands_excl_dMMR_sample_size
facets_c_cytobands_excl_dMMR_status <- table(
  facets_c_cytobands_excl_dMMR$chr_arm_cytoband,
  facets_c_cytobands_excl_dMMR$type,
  facets_c_cytobands_excl_dMMR$Response) %>%
  as.data.frame() %>%
  setNames(c("chr_arm_cytoband", "type", "Response", "Freq"))

facets_c_cytobands_excl_dMMR_status$type %>% table()

facets_c_cytobands_excl_dMMR_del <- facets_c_cytobands_excl_dMMR_status %>%
  filter(type %in% c("DEL", "HEMIZYG")) %>%
  group_by(chr_arm_cytoband, Response) %>%
  summarize(Freq_del_hemizyg = sum(Freq)) %>%
  ungroup() %>%
  left_join(facets_c_cytobands_excl_dMMR_sample_size) %>%
  mutate(Prop_del_hemizyg = Freq_del_hemizyg / Total) %>%
  left_join(facets_c_cytobands %>%
              select(chr_arm_cytoband, chr) %>%
              unique())

facets_c_cytobands_excl_dMMR_del_freq <- facets_c_cytobands_excl_dMMR_del %>%
  select(-c(Freq_del_hemizyg, Total)) %>%
  pivot_wider(names_from = Response, values_from = Prop_del_hemizyg) %>%
  mutate(Delta_del_hemizyg = R - NR) %>%
  arrange(Delta_del_hemizyg)

facets_c_cytobands_excl_dMMR_del_freq <- facets_c_cytobands_excl_dMMR_del_freq %>%
  filter(!chr_arm_cytoband %in% c("NA_NA"))

facets_c_cytobands_excl_dMMR_del %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_del.csv"))

facets_c_cytobands_excl_dMMR_del_freq %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_del_freq.csv"))

facets_c_cytobands_excl_dMMR_gain <- facets_c_cytobands_excl_dMMR_status %>%
  filter(type %in% c("DUP", "DUP-LOH")) %>%
  group_by(chr_arm_cytoband, Response) %>%
  summarize(Freq_dup = sum(Freq)) %>%
  ungroup() %>%
  left_join(facets_c_cytobands_excl_dMMR_sample_size) %>%
  mutate(Prop_dup = Freq_dup / Total) %>%
  left_join(facets_c_cytobands %>%
              select(chr_arm_cytoband, chr) %>%
              unique())

facets_c_cytobands_excl_dMMR_gain_freq <- facets_c_cytobands_excl_dMMR_gain %>%
  select(-c(Freq_dup, Total)) %>%
  pivot_wider(names_from = Response, values_from = Prop_dup) %>%
  mutate(Delta_dup = R - NR) %>%
  arrange(Delta_dup)

facets_c_cytobands_excl_dMMR_gain_freq <- facets_c_cytobands_excl_dMMR_gain_freq %>%
  filter(!chr_arm_cytoband %in% c("NA_NA"))

facets_c_cytobands_excl_dMMR_gain %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_gain.csv"))
facets_c_cytobands_excl_dMMR_gain_freq %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_gain_freq.csv"))

cytoband_fisher_del_excl_dmmr <- data.frame(matrix(ncol = 0, nrow = 0))
for (current_cytoband in unique(facets_c_cytobands_excl_dMMR_del$chr_arm_cytoband)) {
  current_cytoband_data <- facets_c_cytobands_excl_dMMR_del %>% filter(chr_arm_cytoband == current_cytoband)
    if (any(current_cytoband_data$Freq_del_hemizyg == 0)) {
    next
  }
  current_cytoband_data_m <- current_cytoband_data %>%
    select(Response, Freq_del_hemizyg, Total) %>%
    column_to_rownames("Response") %>%
    as.matrix()
  current_cytoband_fisher <- fisher.test(current_cytoband_data_m) %>%
    tidy() %>% mutate(chr_arm_cytoband = current_cytoband)
  cytoband_fisher_del_excl_dmmr <- cytoband_fisher_del_excl_dmmr %>%
    rbind(current_cytoband_fisher)
}

cytoband_fisher_del_excl_dmmr
cytoband_fisher_del_excl_dmmr$fdrq <- p.adjust(
  cytoband_fisher_del_excl_dmmr$p.value, method = "BH")

cytoband_fisher_del_excl_dmmr %>%
  write_csv2(here(wes_dir, "cytoband_fisher_del_excl_dmmr.csv"))


facets_c_cytobands_excl_dMMR_del_freq[
  select_top_extremes(facets_c_cytobands_excl_dMMR_del_freq$Delta_del_hemizyg, top_n = 10) |
    facets_c_cytobands_excl_dMMR_del_freq$chr_arm_cytoband %in% c("chr9_p21.3", "chr9_q34.3"), ] %>%
  left_join(cytoband_fisher_del_excl_dmmr) %>%
  view()
```

# Cytoband losses freq

```{r, fig.height = 7, fig.width = 6}
facets_c_cytobands_excl_dMMR_del_freq[
  select_top_extremes(facets_c_cytobands_excl_dMMR_del_freq$Delta_del_hemizyg,
                      top_n = 10), ]

facets_c_cytobands_excl_dMMR_del_freq[
  select_top_extremes(facets_c_cytobands_excl_dMMR_del_freq$Delta_del_hemizyg, top_n = 10) |
    facets_c_cytobands_excl_dMMR_del_freq$chr_arm_cytoband %in% c("chr9_p21.3", "chr9_q34.3"), ] %>%
  ggplot(aes(x = reorder(chr_arm_cytoband, Delta_del_hemizyg),
             y = Delta_del_hemizyg)) +
  geom_col() + coord_flip() +
  theme_Publication() + labs(x = "", y = "Difference in proportions (R - NR)",
                             title = "Delta in cytoband losses")

facets_c_cytobands_excl_dMMR_del_freq %>% filter(chr == "chr9") %>%
  arrange(Delta_del_hemizyg)
facets_c_cytobands_excl_dMMR_del_freq %>% filter(chr == "chr9") %>%
  ggplot(aes(x = reorder(chr_arm_cytoband, Delta_del_hemizyg),
             y = Delta_del_hemizyg)) +
  geom_col() + coord_flip() + theme_Publication()

facets_c_cytobands_excl_dMMR_del_freq %>% filter(chr_arm_cytoband %in% c("chr9_p21.3", "chr9_q34.3"))

facets_chr9_p21.3_loss <- facets_c_cytobands %>%
  filter(chr_arm_cytoband == "chr9_p21.3") %>%
  filter(type %in% c("DEL", "HEMIZYG"))

facets_chr9_q34.3_loss <- facets_c_cytobands %>%
  filter(chr_arm_cytoband == "chr9_q34.3") %>%
  filter(type %in% c("DEL", "HEMIZYG"))

facets_chr9_q34.3_loss <- facets_c_cytobands %>%
  filter(chr_arm_cytoband == "chr9_q34.3") %>%
  filter(type %in% c("DEL", "HEMIZYG"))

facets_chr11_q13.3_gain <- facets_c_cytobands %>%
  filter(chr_arm_cytoband == "chr11_q13.3") %>%
  filter(type %in% c("DUP"))
facets_c_cytobands 


```

# Cytoband gains freq

```{r, fig.height = 7, fig.width = 6}
facets_c_cytobands_excl_dMMR_gain_freq[
  select_top_extremes(facets_c_cytobands_excl_dMMR_gain_freq$Delta_dup, top_n = 10), ]

facets_c_cytobands_excl_dMMR_gain_freq[
  select_top_extremes(facets_c_cytobands_excl_dMMR_gain_freq$Delta_dup, top_n = 10), ] %>%
  ggplot(aes(x = reorder(chr_arm_cytoband, Delta_dup), y = Delta_dup)) +
  geom_col() + coord_flip() + theme_Publication()

facets_c_cytobands_excl_dMMR_del_freq %>% filter(chr == "chr9") %>%
  ggplot(aes(x = reorder(chr_arm_cytoband, Delta_del_hemizyg),
             y = Delta_del_hemizyg)) +
  geom_col() + coord_flip() + theme_Publication()

```

# Amplifications (cn \> 5)

```{r}
facets_c_cytobands_excl_dMMR %>% filter(cn > 5) %>% ggplot(aes(x = log2(cn))) + geom_density() + theme_Publication()
facets_c_cytobands_excl_dMMR %>% filter(cn > 5) %>% factor_summary()
facets_c_cytobands_excl_dMMR %>% filter(cn > 5) %>% pull(chr_arm_cytoband) %>% table() %>% as.data.frame() %>% arrange(-Freq)
facets_c_cytobands_excl_dMMR_amp <- facets_c_cytobands_excl_dMMR %>% filter(cn > 5) 
facets_c_cytobands_excl_dMMR_amp$type %>% table() 
  
facets_c_cytobands_excl_dMMR_amp <- table(
  facets_c_cytobands_excl_dMMR_amp$chr_arm_cytoband,
  facets_c_cytobands_excl_dMMR_amp$Response) %>%
  as.data.frame() %>%
  setNames(c("chr_arm_cytoband", "Response", "Freq_amp")) %>%
  left_join(facets_c_cytobands_excl_dMMR_sample_size) %>%
  mutate(Prop_amp = Freq_amp / Total) %>%
  left_join(facets_c_cytobands %>%
              select(chr_arm_cytoband, chr) %>%
              unique())

facets_c_cytobands_excl_dMMR_amp_freq <- facets_c_cytobands_excl_dMMR_amp %>%
  select(-c(Freq_amp, Total)) %>%
  pivot_wider(names_from = Response, values_from = Prop_amp) %>%
  mutate(Delta_amp = R - NR) %>%
  arrange(Delta_amp)

facets_c_cytobands_excl_dMMR_amp_freq <- facets_c_cytobands_excl_dMMR_amp_freq %>%
  filter(!chr_arm_cytoband %in% c("NA_NA"))

facets_c_cytobands_excl_dMMR_amp %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_amp.csv"))

facets_c_cytobands_excl_dMMR_amp_freq %>%
  write_csv2(here(wes_dir, "facets_c_cytobands_excl_dMMR_amp_freq.csv"))

cytoband_fisher_amp_excl_dmmr <- data.frame(matrix(ncol = 0, nrow = 0))
for (current_cytoband in unique(facets_c_cytobands_excl_dMMR_amp$chr_arm_cytoband)) {
  
  current_cytoband_data <- facets_c_cytobands_excl_dMMR_amp %>%
    filter(chr_arm_cytoband == current_cytoband)
  if (any(current_cytoband_data$Freq_amp == 0)) {
    next
  }
  current_cytoband_data_m <- current_cytoband_data %>%
    select(Response, Freq_amp, Total) %>%
    column_to_rownames("Response") %>%
    as.matrix()
    
  current_cytoband_fisher <- fisher.test(current_cytoband_data_m) %>%
    tidy() %>%
    mutate(chr_arm_cytoband = current_cytoband)
  cytoband_fisher_amp_excl_dmmr <- cytoband_fisher_amp_excl_dmmr %>%
    rbind(current_cytoband_fisher)
}

cytoband_fisher_amp_excl_dmmr$fdrq <- p.adjust(
  cytoband_fisher_amp_excl_dmmr$p.value, method = "BH")

cytoband_fisher_amp_excl_dmmr %>%
  write_csv2(here(wes_dir, "cytoband_fisher_amp_excl_dmmr.csv"))
```

```{r}
facets_c_cytobands_excl_dMMR_amp_freq[select_top_extremes(facets_c_cytobands_excl_dMMR_amp_freq$Delta_amp, top_n = 10), ]
facets_c_cytobands_excl_dMMR_amp_freq[select_top_extremes(facets_c_cytobands_excl_dMMR_amp_freq$Delta_amp, top_n = 10), ] %>% ggplot(aes(x = reorder(chr_arm_cytoband, Delta_amp), y = Delta_amp)) + geom_col() + coord_flip() + theme_Publication()

facets_c_cytobands_excl_dMMR_del_freq %>% filter(chr == "chr9") %>% ggplot(aes(x = reorder(chr_arm_cytoband, Delta_del_hemizyg), y = Delta_del_hemizyg)) + geom_col() + coord_flip() + theme_Publication()

facets_c_cytobands_excl_dMMR_amp_freq %>% filter(chr == "chr9") %>% ggplot(aes(x = reorder(chr_arm_cytoband, Delta_amp), y = Delta_del_hemizyg)) + geom_col() + coord_flip() + theme_Publication()

facets_c_cytobands_excl_dMMR_amp_freq %>% ggplot(aes(x = reorder(chr_arm_cytoband, Delta_amp), y = Delta_del_hemizyg)) + geom_col() + coord_flip() + theme_Publication()

```

# ----------

# COSMIC Mutational signatures
## Filtered vcf files
```{r}
library('BSgenome.Hsapiens.UCSC.hg38') #v 1.4.3
package.version('BSgenome.Hsapiens.UCSC.hg38') # ow 1.4.5
ref_genome <- "BSgenome.Hsapiens.UCSC.hg38"

list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_BELLINI") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_IMCISION") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_MATISSE") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_NABUCCO") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_NICHE") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_OpACIN_neo") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_PRADO") %>% length()
list.files("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered") %>% length()

list.files(here("data", "raw"))

vcf_files_full_path <- list.files(path="/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered", pattern = "pass.vcf", full.names = TRUE)
vcf_files <- list.files(path="/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered", pattern = "pass.vcf", full.names = FALSE)
vcf_files <- vcf_files[seq(1,length(vcf_files),2)]
SampleName <- matrix(unlist(strsplit(vcf_files, "-vs-")), ncol=2, byrow=T)[,1]
```

```{r}
malformed_vcfs <- c(14, 43, 312, 348, 395)
SampleName[malformed_vcfs]
vcf_files_full_path[malformed_vcfs]

# Pseudocode below
# parse the vcf files, filter and export (slow method) 
# Did not fix the code to properly parse filenames and locate the cohorts, I just hard code adapted and ran it for each of the 5 samples
if (fix_vcf_filtering) {
    for (current_vcf_file in vcf_files_full_path[malformed_vcfs]) {
      print(current_vcf_file)
      # "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_NABUCCO/5519_21_CF33195_AACTCACC_S60-vs-5519_45_CF31087_GACAGTGC_S83-GATKFiltered.vcf"
      # "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_BELLINI/6793_1_CF54491_TAACTTGGTC-GTCGTGAATC_S2-vs-6793_34_CF64315_TTACGCGCAT-CTTAACGACA_S38-GATKFiltered.vcf"
      # "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_OpACIN_neo/S1224Nr33-vs-S1224Nr68-GATKFiltered.vcf"
      #  "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/vcf_IMCISION/S1678Nr1-vs-S1678Nr37-GATKFiltered.vcf"
      # "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_PRADO/S2697Nr23-vs-S2697Nr141-GATKFiltered.vcf"
      current_vcf <- VariantAnnotation::readVcf(
      "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files/vcf_PRADO/S2697Nr23-vs-S2697Nr141-GATKFiltered.vcf")
      
      # Filter for variants that passed filters
      variant_main_info <- data.frame(rowRanges(current_vcf))
      table(variant_main_info$FILTER)
      
      current_vcf_filtered <- current_vcf[
        variant_main_info$FILTER == "PASS", ]
      
      file.exists("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/5519_21_CF33195_AACTCACC_S60-vs-5519_45_CF31087_GACAGTGC_S83-GATKFiltered.pass.vcf")
      file.exists("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/6793_1_CF54491_TAACTTGGTC-GTCGTGAATC_S2-vs-6793_34_CF64315_TTACGCGCAT-CTTAACGACA_S38-GATKFiltered.pass.vcf")
      file.exists("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/S1224Nr33-vs-S1224Nr68-GATKFiltered.pass.vcf")
      file.exists("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/S1678Nr1-vs-S1678Nr37-GATKFiltered.pass.vcf")
      file.exists("/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/S2697Nr23-vs-S2697Nr141-GATKFiltered.pass.vcf")
    
      # Export filtered vcfs
      VariantAnnotation::writeVcf(obj = current_vcf_filtered,
               filename = "/shared/data/niche/Pancancer_DNA_preprocessed_Pedro/VCF_files_filtered/S2697Nr23-vs-S2697Nr141-GATKFiltered.pass.vcf")
      rm(current_vcf, variant_main_info, current_vcf_filtered)
  }

}

```

```{r}
vcfs <- read_vcfs_as_granges(vcf_files_full_path, SampleName, ref_genome)
# vcfs <- read_vcfs_as_granges(vcf_files_full_path, SampleName, ref_genome)
type_occurrences <- mut_type_occurrences(vcfs, ref_genome)
mut_mat <- mut_matrix(vcf_list = vcfs, ref_genome = ref_genome)

plot_spectrum(type_occurrences, CT = TRUE, legend = TRUE)

```

## Load cosmic signatures
```{r}
# Petros took the COSMIC signatures from this URL
# Previous signature collection matched GRCh37... rounded to 4 digits it's the same matrix
sp_url <- paste("https://cancer.sanger.ac.uk/cancergenome/assets/","signatures_probabilities.txt", sep = "")
#cancer_signatures = get_known_signatures(muttype = "snv")
cancer_signatures_raw = read.table(sp_url, sep = "\t", header = TRUE)
cancer_signatures = read.table(sp_url, sep = "\t", header = TRUE)
new_order = match(row.names(mut_mat), cancer_signatures$Somatic.Mutation.Type)
cancer_signatures = cancer_signatures[as.vector(new_order),]
row.names(cancer_signatures) = cancer_signatures$Somatic.Mutation.Type
cancer_signatures = as.matrix(cancer_signatures[,4:33])

# Reorder rows to match signatures and mut_mat profiles
COSMIC_v2_SBS_GRCh37_raw <- read.table(
  here("data", "external", "COSMIC", "COSMIC_v2",
       "COSMIC_v2_SBS_GRCh37.txt"), sep = "", header = TRUE) %>%
  column_to_rownames("Type")

table(row.names(mut_mat) == rownames(COSMIC_v2_SBS_GRCh37_raw))
new_order = match(row.names(mut_mat), rownames(COSMIC_v2_SBS_GRCh37_raw))
COSMIC_v2_SBS_GRCh37 = COSMIC_v2_SBS_GRCh37_raw[as.vector(new_order),]
table(row.names(mut_mat) == rownames(COSMIC_v2_SBS_GRCh37))

COSMIC_v2_SBS_GRCh38_raw <- read.table(
  here("data", "external", "COSMIC", "COSMIC_v2",
       "COSMIC_v2_SBS_GRCh38.txt"), sep = "", header = TRUE) %>%
  column_to_rownames("Type")

table(row.names(mut_mat) == rownames(COSMIC_v2_SBS_GRCh38_raw))
new_order = match(row.names(mut_mat), rownames(COSMIC_v2_SBS_GRCh38_raw))
COSMIC_v2_SBS_GRCh38 = COSMIC_v2_SBS_GRCh38_raw[as.vector(new_order),]
table(row.names(mut_mat) == rownames(COSMIC_v2_SBS_GRCh38))

COSMIC_v2_SBS_GRCh38_raw %>% as.matrix() %>% dist()

COSMIC_v3.4_SBS_GRCh38_raw <- read.table(
  here("data", "external", "COSMIC", "COSMIC_v3.4",
       "COSMIC_v3.4_SBS_GRCh38.txt"), sep = "", header = TRUE) %>%
  column_to_rownames("Type")
table(row.names(mut_mat) == rownames(COSMIC_v3.4_SBS_GRCh38_raw))
new_order = match(row.names(mut_mat), rownames(COSMIC_v3.4_SBS_GRCh38_raw))
COSMIC_v3.4_SBS_GRCh38 = COSMIC_v3.4_SBS_GRCh38_raw[as.vector(new_order),]
table(row.names(mut_mat) == rownames(COSMIC_v3.4_SBS_GRCh38))

```

## COSMIC v2 cosine distances
```{r}
# Compute cosine similarity
cosine_sim <- cosine(as.matrix(COSMIC_v2_SBS_GRCh38_raw))

# Convert similarity to distance
cosine_dist <- 1 - cosine_sim

# View the result
print(cosine_dist)

# Convert to long format for ggplot2
dist_melt <- melt(cosine_dist)

# Create heatmap
ggplot(dist_melt, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "white", high = "red") +
  labs(title = "Cosine Distance Heatmap of COSMIC Signatures",
       x = "Signature", y = "Signature", fill = "Cosine Distance") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text.y = element_text(size = 8))

# Compute cosine similarity
cosine_sim <- cosine(
  as.matrix(
    COSMIC_v2_SBS_GRCh38_raw[, c("Signature_1", "Signature_6", "Signature_14",
                                 "Signature_15", "Signature_20",
                                 "Signature_21", "Signature_26")]))

# Convert similarity to distance
cosine_dist <- 1 - cosine_sim

# View the result
print(cosine_dist)

# Convert to long format for ggplot2
dist_melt <- melt(cosine_dist)

# Create heatmap
ggplot(dist_melt, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile(color = "white") +
  geom_text(aes(label = sprintf("%.2f", value)), size = 3) +
  scale_fill_gradient(low = "white", high = "red") +
  labs(title = "Cosine Distance Heatmap of COSMIC Signatures",
       x = "Signature", y = "Signature", fill = "Cosine Distance") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text.y = element_text(size = 8)) 

# Create heatmap
ggplot(dist_melt, aes(x = Var1, y = Var2, fill = 1 - value)) +
  geom_tile(color = "white") +
  geom_text(aes(label = sprintf("%.2f", 1 - value)), size = 3)  +
  scale_fill_gradient(low = "white", high = "red") +
  labs(title = "Cosine Sim Heatmap of COSMIC Signatures",
       x = "Signature", y = "Signature", fill = "Cosine Sim") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.text.y = element_text(size = 8))
```

## Fit signatures
```{r}
fit_res <- fit_to_signatures(mut_mat, as.matrix(COSMIC_v2_SBS_GRCh38))
fit_res %>% saveRDS(here(output_dir, "fit_res_cosmic_v2_GRCh38.rds"))

fit_res_cosmic_v3.4 <- fit_to_signatures(mut_mat, as.matrix(COSMIC_v3.4_SBS_GRCh38))
fit_res_cosmic_v3.4 %>% saveRDS(here(output_dir, "fit_res_cosmic_v3.4_GRCh38.rds"))

plot_contribution(fit_res$contribution,
  coord_flip = FALSE,
  mode = "absolute"
)

fit_res$contribution %>% dim()
colnames(fit_res$contribution) %>% head()

Relative_contri_Perc <- matrix(data=0, ncol=ncol(fit_res$contribution), nrow=nrow(fit_res$contribution))

colnames(Relative_contri_Perc) <- colnames(fit_res$contribution)
rownames(Relative_contri_Perc) <- rownames(fit_res$contribution)

for (i in 1:length(vcf_files)){
  Relative_contri_Perc[,i]<- fit_res$contribution[,i]/sum(fit_res$contribution[,i])
}

Relative_contri_Perc %>% colSums()

Relative_contri_Perc_t = as.data.frame(t(Relative_contri_Perc))
Relative_contri_Perc_t$name_vcf = rownames(Relative_contri_Perc_t)
colnames(Relative_contri_Perc_t) <- gsub("_", ".", colnames(Relative_contri_Perc_t))
Relative_contri_Perc_t <- Relative_contri_Perc_t %>% rename(name_vcf = name.vcf)

fit_res = as.data.frame(fit_res$contribution)
fit_res_t = as.data.frame(t(fit_res))
fit_res_t$name_vcf = rownames(fit_res_t)
colnames(fit_res_t) <- gsub("_", ".", colnames(fit_res_t))
fit_res_t <- fit_res_t %>% rename(name_vcf = name.vcf)

write.table(Relative_contri_Perc_t, here(output_dir, "Muts_Contribution_Perc_cosmic.txt"), sep="\t",quote=F, row.names = T, col.names = T)
write.table(fit_res_t, file = here(output_dir, "Muts_Contribution_cosmic.txt"), sep="\t", quote=F)

```

## Parse cosmic
```{r}
combine_cosmic <- FALSE
if (combine_cosmic) {
  
  cosmic_combined_v2_GRCh38 <- fit_res_t
  cosmic_cf_samples <- cosmic_combined_v2_GRCh38$name_vcf %>%
    str_split("_") %>%
    unlist()
  cosmic_cf_samples <- cosmic_cf_samples[str_detect(
    cosmic_cf_samples, "^CF|Nr")]
  cosmic_combined_v2_GRCh38$CF_sample_tumor <- cosmic_cf_samples
  
  cosmic_cf_samples[!cosmic_cf_samples %in% dna_metadata_f$CF_sample_tumor]
  
  cosmic_combined_v2_GRCh38 <- cosmic_combined_v2_GRCh38 %>%
    left_join(dna_metadata_f)
  cosmic_combined_v2_GRCh38 %>%
    write_csv2(here(output_dir, "cosmic_combined_v2_GRCh38.csv"))
  
  cosmic_perc_combined_v2_GRCh38 <- Relative_contri_Perc_t
  cosmic_cf_samples <- cosmic_perc_combined_v2_GRCh38$name_vcf %>%
    str_split("_") %>%
    unlist()
  cosmic_cf_samples <- cosmic_cf_samples[
    str_detect(cosmic_cf_samples, "^CF|Nr")]
  cosmic_perc_combined_v2_GRCh38$CF_sample_tumor <- cosmic_cf_samples
  
  cosmic_perc_combined_v2_GRCh38 <- cosmic_perc_combined_v2_GRCh38 %>%
    left_join(dna_metadata_f)
  cosmic_perc_combined_v2_GRCh38 %>%
    write_csv2(here(output_dir, "cosmic_perc_combined_v2_GRCh38.csv"))
  
}

cosmic_combined_v2_GRCh38 <- read_csv2(
  here(output_dir, "cosmic_combined_v2_GRCh38.csv"))
cosmic_perc_combined_v2_GRCh38 <- read_csv2(
  here(output_dir, "cosmic_perc_combined_v2_GRCh38.csv"))

```

```{r}
# Updated aetiologies
cosmic_v3.4_aetiology <- read_excel(
  here("data", "external", "COSMIC", "COSMIC_SBS_Aetiology_20250728.xlsx"),
  sheet = "v3.4 v102")
cosmic_v2_aetiology <- read_excel(
  here("data","external", "COSMIC", "COSMIC_SBS_Aetiology_20250728.xlsx"),
  sheet = "v2.0")

cosmic_v2_aetiology_p <- cosmic_v2_aetiology %>%
  mutate(Aetiology_simplified = Aetiology_simplified_later,
         `Signature Number` = sprintf(
           "Signature %s", parse_number(COSMIC_SBS))) %>%
  select(COSMIC_SBS, `Signature Number`, Aetiology_simplified) %>%
  mutate(cosmic_signature = gsub(" ", ".", `Signature Number`))

# APOBEC Signature 2 + 13
# HRD Signature 3
# Smoking # SBS4
# dMMR SBS6 + SBS14 + SBS15 + SBS20 + SBS21 + SBS26
# UV Signature 7
# POLE_D1 Signature 10
cosmic_combined_v2_GRCh38$Trial %>% table()
cosmic_combined_v2_GRCh38 %>%
  select(all_of(c(paste0("Signature.", c(1:30))))) %>%
  as.matrix() %>%
  rowSums()

cosmic_combined_v2_GRCh38 %>%
  filter(!CF_sample_tumor %in% dna_metadata_f$CF_sample_tumor)

# Filter for data present in metadata, keep in mind NICHE2 MSI sensitivity analysis
cosmic_combined_v2_GRCh38 <- cosmic_combined_v2_GRCh38 %>%
  filter(CF_sample_tumor %in% dna_metadata_f$CF_sample_tumor)
cosmic_perc_combined_v2_GRCh38 <- cosmic_perc_combined_v2_GRCh38 %>%
  filter(CF_sample_tumor %in% dna_metadata_f$CF_sample_tumor)

cosmic_combined_v2_GRCh38 <- cosmic_combined_v2_GRCh38 %>%
  mutate(Signature_dMMR = Signature.6 + Signature.14 + Signature.15 + Signature.20 + Signature.21 + Signature.26,
         Signature_APOBEC = Signature.2 + Signature.13)

cosmic_perc_combined_v2_GRCh38 <- cosmic_perc_combined_v2_GRCh38 %>%
  mutate(Signature_dMMR = Signature.6 + Signature.14 + Signature.15 + Signature.20 + Signature.21 + Signature.26,
         Signature_APOBEC = Signature.2 + Signature.13)

cosmic_combined_l <- cosmic_combined_v2_GRCh38 %>%
  pivot_longer(cols = c(paste0("Signature.", c(1:30))),
               names_to = "cosmic_signature",
               values_to = "cosmic_scores") %>%
  left_join(cosmic_v2_aetiology_p)

cosmic_perc_combined_l <- cosmic_perc_combined_v2_GRCh38 %>%
  pivot_longer(cols = c(paste0("Signature.", c(1:30))),
               names_to = "cosmic_signature",
               values_to = "percentage") %>%
  left_join(cosmic_v2_aetiology_p)

cosmic_combined_l$cosmic_signature <- factor(
  cosmic_combined_l$cosmic_signature,
  levels = c(paste0("Signature.", c(1:30))))
cosmic_perc_combined_l$cosmic_signature <- factor(
  cosmic_perc_combined_l$cosmic_signature,
  levels = c(paste0("Signature.", c(1:30))))

cosmic_colors <- c(Aflatoxin = "#d0e0e3", Aging = "#3d85c6",
                   APOBEC = "#674ea7",
                   Aristolochic_acid = "#d9d2e9", dMMR = "#B73A3A",
                   HRD = "#e69138", Other = "#cccccc", POL1 = "#cc0000",
                   POLE_POLD1 = "#5b0f00", POLH = "#FEB5DA", ROS = "#6DB6FF",
                   SBS5 = "#ea9999", Smoking = "#38761d", TMZ = "#b5a7d6",
                   Tobacco_chewing = "#6aa84f", UV = "#f1c232")

cosmic_combined_p <- cosmic_combined_l %>%
  group_by(CF_sample_tumor, Aetiology_simplified) %>%
  summarize(cosmic_scores = sum(cosmic_scores)) %>%
  pivot_wider(names_from = Aetiology_simplified,
              values_from = cosmic_scores) %>%
  left_join(dna_metadata_f) 

cosmic_perc_combined_p <- cosmic_perc_combined_l %>%
  group_by(CF_sample_tumor, Aetiology_simplified) %>%
  summarize(percentage = sum(percentage)) %>%
  pivot_wider(names_from = Aetiology_simplified,
              values_from = percentage) %>%
  left_join(dna_metadata_f) 

mean_cosmic_abs <- cosmic_combined_l %>%
  group_by(Trial, cosmic_signature) %>% 
  summarize(mean_abs = mean(cosmic_scores))

mean_cosmic_percs <- cosmic_perc_combined_l %>%
  group_by(Trial, cosmic_signature) %>% 
  summarize(mean_percentage = mean(percentage))

mean_cosmic_abs %>%
  select(Trial, cosmic_signature, mean_abs) %>%
  pivot_wider(names_from = cosmic_signature, values_from = mean_abs) %>%
  summary()

mean_cosmic_percs %>%
  select(Trial, cosmic_signature, mean_percentage) %>%
  pivot_wider(names_from = cosmic_signature, values_from = mean_percentage) %>%
  summary()
```

```{r}
cosmic_combined_v2_GRCh38 %>%
  ggplot(aes(x = Trial, y = Signature_APOBEC, fill = Trial)) + geom_boxplot() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values = trial_colors)

cosmic_combined_v2_GRCh38 %>%
  ggplot(aes(x = Trial, y = Signature_dMMR, fill = Trial)) + geom_boxplot() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values = trial_colors)

cosmic_combined_v2_GRCh38 %>%
  ggplot(aes(x = Trial, y = Signature.7, fill = Trial)) + geom_boxplot() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values = trial_colors)

cosmic_combined_v2_GRCh38 %>%
  ggplot(aes(x = Trial, y = Signature.5, fill = Trial)) + geom_boxplot() +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values = trial_colors)

n_variants_dp20_vaf_tmb_status %>%
  left_join(cosmic_combined_p) %>%
  ggplot(aes(x = POLE, y = POLE_POLD1, col = POLE)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "POLE mutation") +
  expand_limits(y = 0)

n_variants_dp20_vaf_tmb_status  %>%
  left_join(cosmic_combined_p) %>%
  ggplot(aes(x = POLD1, y = POLE_POLD1, col = POLD1)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "POLD1 mutation") +
  expand_limits(y = 0)

n_variants_dp20_vaf_tmb_status %>%
  mutate(somatic_dMMR = dmmr_hits > 0) %>%
  left_join(cosmic_combined_p) %>%
  ggplot(aes(x = somatic_dMMR, y = dMMR, col = somatic_dMMR)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "dMMR mutation") +
  expand_limits(y = 0)

n_variants_dp20_vaf_tmb_status %>%
  mutate(somatic_dMMR_hit = dmmr_hits > 0) %>%
  left_join(cosmic_combined_p) %>%
  ggplot(aes(x = somatic_dMMR_hit, y = dMMR, col = somatic_dMMR_hit)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "dMMR mutation", y = "Cosmic dMMR signatures") +
  expand_limits(y = 0)

n_variants_dp20_vaf_tmb_status %>%
  mutate(NICHE_MSI = ifelse(Trial == "NICHE-MSI", "NICHE-MSI", "Other")) %>%
  mutate(somatic_dMMR_hit = dmmr_hits > 0) %>%
  left_join(cosmic_combined_p) %>%
  ggplot(aes(x = somatic_dMMR_hit, y = dMMR, col = somatic_dMMR_hit)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "dMMR mutation", y = "Cosmic dMMR signatures") +
  expand_limits(y = 0) +
  facet_wrap(~NICHE_MSI)

n_variants_dp20_vaf_tmb_status %>%
  mutate(NICHE_MSI = ifelse(Trial == "NICHE-MSI", "NICHE-MSI", "Other")) %>%
  mutate(somatic_dMMR_hit = dmmr_hits > 0) %>%
  left_join(cosmic_perc_combined_p) %>%
  ggplot(aes(x = somatic_dMMR_hit, y = dMMR, col = somatic_dMMR_hit)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "dMMR mutation", y = "Cosmic dMMR signatures") +
  expand_limits(y = 0) +
  facet_wrap(~NICHE_MSI)

n_variants_dp20_vaf_tmb_status  %>%
  mutate(somatic_dMMR = dmmr_hits > 0) %>%
  filter(somatic_dMMR) %>%
  left_join(cosmic_combined_p)  %>%
  ggplot(aes(x = Trial, y = log10(dMMR + 1), fill = Trial)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_point()  +
  theme_Publication() +
  labs(x = "dMMR mutation") +
  expand_limits(y = 0) +
  scale_fill_manual(values  = trial_colors)

```

```{r, fig.width = 12, fig.height = 12}
mean_cosmic_abs %>%
  left_join(cosmic_v2_aetiology_p) %>%
  select(Trial, Aetiology_simplified, mean_abs) %>%
  group_by(Trial, Aetiology_simplified) %>%
  summarize(mean_abs = sum(mean_abs)) %>%
  pivot_wider(names_from = Aetiology_simplified, values_from = mean_abs)

mean_cosmic_abs %>%
  left_join(cosmic_v2_aetiology_p) %>%
  select(Trial, Aetiology_simplified, mean_abs) %>%
  group_by(Trial, Aetiology_simplified) %>%
  summarize(mean_abs = sum(mean_abs)) %>%
  ggplot(aes(x = Trial, y = mean_abs, fill = Aetiology_simplified)) +
  geom_col() +
  facet_wrap(~Aetiology_simplified) +
  scale_fill_manual(values = cosmic_colors) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggsave(here(output_dir, "mean_cosmic_trial_abs.png"), height = 6, width = 7)

mean_cosmic_abs %>%
  left_join(cosmic_v2_aetiology_p) %>%
  select(Trial, Aetiology_simplified, mean_abs) %>%
  group_by(Trial, Aetiology_simplified) %>%
  summarize(mean_abs = sum(mean_abs)) %>%
  ggplot(aes(x = Trial, y = mean_abs, fill = Aetiology_simplified)) +
  geom_col() +
  facet_wrap(~Aetiology_simplified, scales = "free") +
  scale_fill_manual(values = cosmic_colors) +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
ggsave(here(output_dir, "mean_cosmic_trial_abs_facet.png"),
       height = 12, width = 12)

mean_cosmic_abs %>%
  left_join(cosmic_v2_aetiology_p)%>%
  mutate(Aetiology_simplified = ifelse(
    Aetiology_simplified %in% c("Aflatoxin", "Aristolochic_acid",
                                "POL1", "POLH"),
    "Other", Aetiology_simplified))  %>%
  ggplot(aes(x = Trial, y = mean_abs, fill = Aetiology_simplified)) +
  geom_col() + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = cosmic_colors) +
  labs(x = "", y = "Mean relative frequency", fill = "COSMIC aetiology")+
  theme(axis.title.x = element_blank())
ggsave(here(output_dir, "mean_cosmic_trial_abs.png"), height = 6, width = 7)

mean_cosmic_abs %>% write_csv2(here(output_dir, "mean_cosmic_abs.csv"))
```

```{r, fig.width = 8, fig.height = 8}
mean_cosmic_percs %>%
  left_join(cosmic_v2_aetiology_p) %>% 
  select(Trial, Aetiology_simplified, mean_percentage) %>%
  group_by(Trial, Aetiology_simplified) %>%
  summarize(mean_percentage = sum(mean_percentage)) %>%
  pivot_wider(names_from = Aetiology_simplified, values_from = mean_percentage)

mean_cosmic_percs %>%
  left_join(cosmic_v2_aetiology_p) %>%
  ggplot(aes(x = Trial, y = mean_percentage, fill = Aetiology_simplified)) +
  geom_col() + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = cosmic_colors) +
  labs(x = "", y = "Mean relative frequency", fill = "COSMIC aetiology")

mean_cosmic_percs %>%
  left_join(cosmic_v2_aetiology_p) %>%
  mutate(Aetiology_simplified = ifelse(
    Aetiology_simplified %in% c("Aflatoxin", "Aristolochic_acid",
                                "POL1", "POLH"),
    "Other", Aetiology_simplified)) %>%
  ggplot(aes(x = Trial, y = mean_percentage, fill = Aetiology_simplified)) +
  geom_col() + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_fill_manual(values = cosmic_colors) +
  labs(x = "", y = "Mean relative frequency", fill = "COSMIC aetiology") +
  guides(fill = guide_legend(nrow = 4)) +
  theme(axis.title.x = element_blank())

ggsave(here(output_dir, "mean_cosmic_trial_percs.png"), height = 6, width = 7)

mean_cosmic_percs %>% write_csv2(here(output_dir, "mean_cosmic_percs.csv"))

```

# Cosmic response

```{r, fig.width = 10, fig.height = 10}
cosmic_combined_l %>%
  ggplot(aes(x = cosmic_signature, y = cosmic_scores, fill = Response)) +
  geom_boxplot() +
  scale_fill_manual(values = c(Color_NR, Color_R)) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  guides(fill = FALSE) + labs(x = "COSMIC SBS", y = "Total contribution")

cosmic_combined_l %>%
  ggplot(aes(x = cosmic_signature, y = cosmic_scores, fill = Response)) +
  geom_boxplot() +
  scale_fill_manual(values = c(Color_NR, Color_R)) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  guides(fill = FALSE) + labs(x = "COSMIC SBS", y = "Total contribution") +
  facet_wrap(~Trial, scales = "free", ncol = 4) + coord_flip()

cosmic_perc_combined_l %>%
  ggplot(aes(x = cosmic_signature, y = percentage, fill = Response)) +
  geom_boxplot() +
  scale_fill_manual(values = c(Color_NR, Color_R)) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  guides(fill = FALSE) + labs(x = "COSMIC SBS", y = "Relative contribution")

cosmic_perc_combined_l %>%
  left_join(cosmic_v2_aetiology_p) %>%
  ggplot(aes(x = Aetiology_simplified, y = percentage, fill = Response)) +
  geom_boxplot() +
  scale_fill_manual(values = c(Color_NR, Color_R)) + theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(x = "") + facet_wrap(~Trial) + coord_flip()

```

# ----------

```{r}
table(specific_mutation_df$Trial, specific_mutation_df$SETDB1)
table(specific_mutation_df$Trial, specific_mutation_df$CD58)
table(specific_mutation_df$Trial, specific_mutation_df$POLE)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$POLD1)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$NRAS)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$EGFR)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$ATM)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$ATR)

table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$POLD1)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$NRAS)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$EGFR)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$ATM)
table(pathway_level_mutation_df$Trial, pathway_level_mutation_df$ATR)

```

# Export maf and combine WES sample df

```{r}
combined_maf %>% saveRDS(here(output_dir, "combined_maf_not_vaf_filtered.rds"))
combined_maf_dp20_vaf %>% saveRDS(here(output_dir, "combined_maf_dp20_vaf.rds"))
write.mafSummary(maf = combined_maf_dp20_vaf,
                 basename = here(output_dir, "combined_maf_dp20_vaf"))

combined_maf_dp20_vaf_df_specific_f %>%
  saveRDS(here(output_dir, "combined_maf_dp20_vaf_df_specific_f.rds"))

combined_maf_dp20_vaf_df_specific_f %>%
  write_csv2(here(output_dir, "combined_maf_dp20_vaf_df_specific_f.csv"))

combined_maf_specific %>% saveRDS(
  here(output_dir, "combined_maf_specific.rds"))
combined_maf_specific_f %>% saveRDS(
  here(output_dir, "combined_maf_specific_f.rds"))
combined_maf_specific_f <- readRDS(
  here(wes_dir, "combined_maf_specific_f.rds"))
   
wes_features_df <- n_variants_dp20_vaf_tmb %>%
  left_join(facets_data_metrics)%>%
  left_join(facets_aneuploidy_score_df)  %>%
  left_join(pathway_level_mutation_df) %>%
  left_join(cosmic_combined_p %>%
  select(c("CF_sample_tumor", "APOBEC", "Aflatoxin", "Aging",
           "Aristolochic_acid", "HRD", "Other", "POLE_POLD1", "POLH",
           "ROS", "SBS5", "Smoking", "TMZ", "Tobacco_chewing", "UV", "dMMR")) %>%
  left_join(cosmic_perc_combined_p %>%
              select(c("CF_sample_tumor", "APOBEC", "Aflatoxin", "Aging",
                       "Aristolochic_acid", "HRD", "Other", "POLE_POLD1",
                       "POLH", "ROS", "SBS5", "Smoking", "TMZ",
                       "Tobacco_chewing", "UV", "dMMR")),
            by = "CF_sample_tumor", suffix = c("_abs", "_perc"))
  ) %>%
  left_join(cnvkit_cnvs_aneuploidy_score_df)

wes_features_df %>% filter(is.na(Aneuploidy_score))
wes_features_df %>% filter(is.na(Purity))
wes_features_df %>% filter(is.na(Purity)) %>% factor_summary()
wes_features_df %>% write_csv2(here(wes_dir, "wes_features_df.csv"))

prado_opacin_neo_tmb <- n_variants_dp20_vaf_tmb %>%
  filter(Trial %in% c("PRADO", "OpACIN_neo")) 
prado_opacin_neo_tmb %>% write_csv2(here(output_dir, "prado_opacin_neo_tmb.csv"))
```

```{r}
wes_features_df <- read_csv2(here(output_dir, "wes_features_df.csv"))

wes_features_df %>%
  ggplot(aes(x = TMB, y = APOBEC_abs, col = Trial)) +
  geom_point() +
  theme_Publication() +
  scale_color_manual(values = trial_colors)
ggsave(here(output_dir, "TMB_APOBEC_trial.png"), height = 5, width = 6)

wes_features_df %>%
  ggplot(aes(x = TMB, y = UV_abs, col = Trial)) +
  geom_point() +
  theme_Publication() +
  scale_color_manual(values = trial_colors)
ggsave(here(output_dir, "TMB_UV_trial.png"), height = 5, width = 6)

wes_features_df %>%
  ggplot(aes(x = TMB, y = dMMR_abs, col = Trial)) +
  geom_point() +
  theme_Publication() +
  scale_color_manual(values = trial_colors)
ggsave(here(output_dir, "TMB_dMMR_trial.png"), height = 5, width = 6)

```

# ----------
# Lint script
```{r}
# lint("~/NKI_pancancer_neoadj_io/src/pancan_WES_analysis.Rmd")
lint(filename = "~/NKI_pancancer_neoadj_io/src/pancan_WES_analysis.Rmd",
  linters = unused_import_linter()
)
```
# ----------

# sessionInfo - Package Versions

```{r sessionInfo}
sessionInfo()
```
