---
title: "parse rnaseq data"
author: "Pedro Batista Tan"
date: "`r strftime(Sys.Date(), '%B %d %Y')`"
output: 
  html_document:
    toc: true
    number_sections: true
params:
  export: TRUE

---

Author comment:

File description comment:
The purpose of this script is to parse RNAseq data from neoadjuvant IO trial samples
It receives metadata and RNA counts as inputs, maps ensembl IDs to symbols,
selects IDs with multiple mapping, and aggregates IDs that map to the
same symbol if there is information on the location
It filters genes expressed in at least 20% of samples, calculates RPM, TPM
and exports the RNAseq data
It also exports QC plots on rRNA percentage and total counts

# Loading Packages
```{r, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, warning = FALSE, message = FALSE, dpi = 300, error = TRUE
  # fig.width = 10, , output: html_notebook
)

# Install if necessary install.packages("BiocManager", "sva")
# Install if necessary BiocManager::install("DESeq2")
library(tidyverse)
library(readxl)
library(ggrepel)
library(here)
library(ggpointdensity)
# library(sva)
source("helper_functions.R")

set.seed(42)
select <- dplyr::select
rename <- dplyr::rename
here <- here::here
```

# Setup
```{r}
export <- params$export

# generate output directory
output_dir <- here("data", "processed", "rnaseq_data")
print(output_dir)
metadata_dir <- here("data", "processed", "metadata")
data_dir <- here("data", "raw", "Pancancer_RNA_preprocessed_Pedro")

# if output_dir does not exist, create it
if (!dir.exists(output_dir)) {
    dir.create(output_dir)
} else {
  print("Output folder already exists")
}

```

# Function definitions
```{r}
process_htseq_mapping_stats <- function(count_data, mapping_data, output_file) {
  mapping_data <- mapping_data %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("CF_sample")
  
  mapping_data <-  mapping_data %>%
    setNames(c("CF_sample", "no_feature", "ambiguous", "too_low_aQual",
               "not_aligned", "alignment_not_unique"))
  
  mapping_data <- mapping_data %>% left_join(
colSums(count_data %>%
          select(-ensembl_gene_id)) %>%
  as.data.frame() %>%
  setNames("total_counts") %>%
  rownames_to_column("CF_sample"))
  
  mapping_data %>% write_csv2(output_file)
   
  plot <- mapping_data %>%
    ggplot(aes(x = reorder(CF_sample, total_counts), y = total_counts)) +
    geom_col() +
    coord_flip() +
    labs(x = "")
  print(plot)
  
  plot <- mapping_data %>%
    pivot_longer(cols = c("no_feature", "ambiguous", "too_low_aQual",
               "not_aligned", "alignment_not_unique", "total_counts"),
               names_to = "type", values_to = "value") %>%
    ggplot(aes(x = CF_sample, y = value, fill = type)) +
    geom_col() +
    coord_flip()
  print(plot)

  plot <- mapping_data %>%
    pivot_longer(cols = c("no_feature", "ambiguous", "too_low_aQual",
               "not_aligned", "alignment_not_unique", "total_counts"),
               names_to = "type", values_to = "value") %>%
    group_by(CF_sample) %>%
    mutate(value_perc = value / sum(value)) %>%
    ggplot(aes(x = CF_sample, y = value_perc, fill = type)) +
    geom_col() +
    coord_flip()
  print(plot)
}

map_ensembl_ids <- function(rnaseq_data, gene_map, exclusion_file,
                            filename, min_nonzero_fraction = 0.2) {
  # If multiple ensembl IDs map to the same gene symbol
  # check if genes can be confidently aggregated by location
  # use the ensembl ID as an identifier instead
  
  # Replace ensembl ID for gene symbol, when available
  df_counts_symbol <- rnaseq_data %>% left_join(gene_map)
  
  # Many IDs mapped to the same gene symbol
  duplicated_df <- table(df_counts_symbol$hgnc_symbol) %>%
    as.data.frame() %>%
    filter(Freq > 1)
  
  duplicated <- which(df_counts_symbol$hgnc_symbol %in% duplicated_df$Var1)
  
  # Inspect what led to duplicated entries
  # df_counts_symbol[duplicated, ] %>%
  #   select(ensembl_gene_id, hgnc_symbol) %>%
  #   left_join(gene_features) %>%
  #   view()
  
  # These genes have alternative transcripts under the same symbol
   aggregate <- duplicated_df$Var1[duplicated_df$Var1 != ""]
  
  df_counts_aggr <- df_counts_symbol %>%
    select(-ensembl_gene_id) %>%
    filter(hgnc_symbol %in% aggregate)
  
  # # Aggregate their counts
  df_counts_aggr <- df_counts_aggr %>%
    group_by(hgnc_symbol) %>%
    summarize_at(
      colnames(df_counts_aggr)[!colnames(df_counts_aggr) %in%
                               c("hgnc_symbol")], sum) %>%
    column_to_rownames("hgnc_symbol")
  
  # Get genes which are still duplicated (empty symbols)
  duplicated_df <- table(df_counts_symbol$hgnc_symbol) %>%
    as.data.frame() %>%
    filter(Freq > 1) %>%
    filter(!Var1 %in% rownames(df_counts_aggr))
  
  duplicated <- which(df_counts_symbol$hgnc_symbol %in% duplicated_df$Var1)
  
  df_counts <- df_counts_symbol %>% mutate(gene = hgnc_symbol)
  df_counts_map <- df_counts %>% select(ensembl_gene_id, hgnc_symbol, gene)
  
  # Replace duplicated ones with the ensembl gene IDs
  df_counts$gene[duplicated] <- df_counts_symbol$ensembl_gene_id[duplicated]
  rownames(df_counts) <- NULL
  
  # Replace NA symbols with the ensembl gene IDs
  na_symbols <- which(is.na(df_counts_symbol$hgnc_symbol))
  df_counts$gene[na_symbols] <- df_counts_symbol$ensembl_gene_id[na_symbols]

  df_counts <- df_counts %>%
    filter(!hgnc_symbol %in% rownames(df_counts_aggr)) %>%
    column_to_rownames("gene") %>%
    select(-c(ensembl_gene_id, hgnc_symbol)) %>%
    rbind(df_counts_aggr)

  # Keep genes that have non zero expression in at least 20% of samples
  n_samples <- ncol(df_counts)
  min_keep <- n_samples * min_nonzero_fraction
  table(rowSums(df_counts) != 0)
  table(rowSums(df_counts != 0) > min_keep)
  
  # Export gene exclusions
  excluded_genes <- rownames(
    df_counts[rowSums(df_counts != 0) <= min_keep, ])
  
  excluded_genes %>%  saveRDS(exclusion_file)

  df_counts <- df_counts[rowSums(df_counts != 0) > min_keep, ]
  rm(df_counts_symbol, duplicated_df, duplicated, df_counts_aggr)
  
  # Export
  df_counts %>% rownames_to_column("gene") %>% write_csv2(filename)
  
  df_counts
}

calculate_rpm <- function(rnaseq_data, filename) {
  rpm_counts <- sweep(x = rnaseq_data,
                    STATS = colSums(rnaseq_data) / 1e6, MARGIN = 2, FUN = "/")
  rpm_counts %>%
    rownames_to_column("gene") %>%
    write_csv2(filename)
  rpm_counts
}

get_rrna_fraction <- function(df_counts, rrna_data) {
  print(rownames(df_counts)[
    rownames(df_counts) %in% human_rrna_genes$Approved.symbol])
  
  qc_counts <- colSums(
    df_counts[rownames(df_counts) %in% human_rrna_genes$Approved.symbol, ]) %>%
    as.data.frame() %>%
    setNames("rRNA_counts") %>%
    rownames_to_column("CF_sample") %>%
    left_join(colSums(df_counts) %>%
                as.data.frame() %>%
                setNames("raw_total") %>%
                rownames_to_column("CF_sample")) %>%
    mutate(fraction_rRNA = rRNA_counts / raw_total)
  
  qc_counts
}
```

# Load metadata
```{r}
bellini_metadata_f <-
  read_csv2(here(metadata_dir, "bellini_rna_metadata.csv"))
imcision_metadata_f <-
  read_csv2(here(metadata_dir, "imcision_rna_metadata.csv"))
nabucco_metadata_f <-
  read_csv2(here(metadata_dir, "nabucco2_rna_metadata.csv"))
niche_metadata_f <-
  read_csv2(here(metadata_dir, "niche_rna_metadata.csv"))
opacin_neo_metadata_f <-
  read_csv2(here(metadata_dir, "opacin_neo_rna_metadata.csv"))
prado_metadata_f <-
  read_csv2(here(metadata_dir, "prado_rna_metadata.csv"))
matisse_metadata_f <-
  read_csv2(here(metadata_dir, "matisse_rna_metadata.csv"))

all_cohort_rna_metadata <-
  read_csv2(here(metadata_dir, "combined_rna_metadata.csv"))
```

# BELLINI
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts
bellini_rna_raw <- read.table(
  here(data_dir, "BELLINI", "RNAseq_output", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
bellini_rna <- bellini_rna_raw
# tail(bellini_rna)

bellini_rna_3B_raw <- read.table(
  here(data_dir, "BELLINI", "RNAseq_output", "Counts",
       "BELINNI_III_Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")

bellini_rna_3B_pre_raw <- read.table(
  here(data_dir, "BELLINI", "RNAseq_output", "Counts",
       "Count_matrix_HTseq_BELLINI_III_pre_samples.txt"), row.names = 1, sep = "\t")
bellini_rna_3B <- bellini_rna_3B_raw


sum(rownames(bellini_rna_3B_pre_raw) == rownames(bellini_rna_3B_raw)) == nrow(bellini_rna_3B_raw)
sum(rownames(bellini_rna_raw) == rownames(bellini_rna_3B_raw)) == nrow(bellini_rna_3B_raw)

bellini_rna <- bellini_rna_raw %>% cbind(bellini_rna_3B_raw) %>% cbind(bellini_rna_3B_pre_raw)


# bellini_gcf_7375 <- read.table(
#   here("/shared/gcf/p.batistatan/shared/m.d.maaker/7375/genecounts.txt"), row.names = 1, sep = "\t", header = TRUE)
# bellini_gcf_7695 <- read.table(
#   here("/shared/gcf/p.batistatan/shared/m.d.maaker/7695/genecounts.txt"), row.names = 1, sep = "\t", header = TRUE)
# 
# bellini_gcf_genes <- bellini_gcf_7375[30:34] %>% full_join(bellini_gcf_7695[29:33])
# bellini_gcf_genes %>% factor_summary()
# bellini_gcf_7375[30:34] %>% filter(!external_gene_id %in% bellini_gcf_7695$external_gene_id) %>% factor_summary()
# bellini_gcf_7695[29:33] %>% filter(!external_gene_id %in% bellini_gcf_7375$external_gene_id) %>% factor_summary()
# 
# bellini_gcf_7375[30:34] %>% filter(!external_gene_id %in% bellini_rna$external_gene_id) %>% factor_summary()
# bellini_gcf_7695[29:33] %>% filter(!external_gene_id %in% bellini_rna$external_gene_id) %>% factor_summary()

# bellini_gcf <- bellini_gcf_7375[, 1:29] %>% rownames_to_column("ensembl_gene_id") %>% full_join(bellini_gcf_7695[, 1:28] %>% rownames_to_column("ensembl_gene_id"))
# 
# bellini_gcf$ensembl_gene_id %>% table() %>% as.data.frame() %>% filter(Freq > 1)
# 
# bellini_gcf %>% filter(!ensembl_gene_id %in% bellini_rna$ensembl_gene_id) %>% nrow()
# bellini_rna <- bellini_rna_raw %>% cbind(bellini_rna_3B_raw)

# bellini_rna <- bellini_rna_raw  %>%
#   rownames_to_column("ensembl_gene_id") %>%
#   left_join(bellini_gcf) %>%
#   column_to_rownames("ensembl_gene_id")

# Parse column names
parsed_colnames <- colnames(bellini_rna)

parsed_colnames <- parsed_colnames %>%
  as.data.frame() %>%
  setNames("sample_raw")

cfmp_sample <- str_split(parsed_colnames$sample_raw, "_")

cfmp_sample <- unlist(cfmp_sample)
cfmp_sample <- cfmp_sample[cfmp_sample != ""]
cfmp_sample <- cfmp_sample[str_detect(cfmp_sample, "CF")]

parsed_colnames$cfmp_sample <- cfmp_sample
parsed_colnames$cfmp_sample %>% table() %>% as.data.frame() %>% filter(Freq > 1)

# 4 samples with apparent resequencing, duplicated names
# 7695_4_CF79327_CCTATGTGTA-TAGTGAGTCG_S50-mdup.cram
# 7695_3_CF79317_ACGGTCGCAT-GTACTCTCTA_S49-mdup.cram
# 7695_2_CF79278_TAGGAAGCGG-CCAATCCAGG_S48-mdup.cram
# 7695_1_CF79266_GCGCGGTTAA-AAGCTATAGC_S47-mdup.cram

parsed_colnames$sample_raw %>% table() 

sum(str_detect(parsed_colnames$sample_raw,
               parsed_colnames$cfmp_sample)) == length(
                 parsed_colnames$cfmp_sample)

parsed_colnames <- parsed_colnames %>%
  mutate(cfmp_sample = ifelse(sample_raw == "X7375_19_CF79317_TGCAACACTA.CGTTCTGGTA_S21", "X19_CF79317", cfmp_sample))
parsed_colnames <- parsed_colnames %>%
  mutate(cfmp_sample = ifelse(sample_raw == "X7375_29_CF79327_ACGCAACACA.CATTCAGGAG_S29", "X29_CF79327", cfmp_sample))
parsed_colnames <- parsed_colnames %>%
  mutate(cfmp_sample = ifelse(sample_raw == "X7375_3_CF79266_AATTGGCACA.CAGTCGGAGT_S4", "X3_CF79266", cfmp_sample))
parsed_colnames <- parsed_colnames %>%
  mutate(cfmp_sample = ifelse(sample_raw == "X7375_15_CF79278_GTACGCTCAA.ATTGATGGCG_S1", "X15_CF79278", cfmp_sample))


parsed_colnames$cfmp_sample %in% bellini_metadata_f$CF_sample
parsed_colnames$cfmp_sample[!parsed_colnames$cfmp_sample %in% bellini_metadata_f$CF_sample]

# Replace column names
colnames(bellini_rna)
colnames(bellini_rna) <- parsed_colnames$cfmp_sample
colnames(bellini_rna)

length(colnames(bellini_rna))

# Missing metadata for RNAseq files outside cohort
colnames(bellini_rna) %in% bellini_metadata_f$CF_sample

colnames(bellini_rna)[
  !colnames(bellini_rna) %in% bellini_metadata_f$CF_sample]

# All files with metadata are present
bellini_metadata_f %>% filter(!CF_sample %in% colnames(bellini_rna)) %>% factor_summary()

# Get mapping statistics
bellini_rna_mapping <- bellini_rna[61542 : 61546, ]

bellini_rna <- bellini_rna[1:61541, ] %>%
  rownames_to_column("ensembl_gene_id")

bellini_rna %>% ggplot(aes(x = CF79317, y = X19_CF79317)) + geom_point() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
bellini_rna %>% ggplot(aes(x = CF79327, y = X29_CF79327)) + geom_point() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
bellini_rna %>% ggplot(aes(x = CF79266, y = X3_CF79266)) + geom_point() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
bellini_rna %>% ggplot(aes(x = CF79278, y = X15_CF79278)) + geom_point() + theme_Publication() + geom_abline(slope = 1, intercept = 0)

# Impute zeros
# bellini_rna[is.na(bellini_rna)] <- 0

process_htseq_mapping_stats(bellini_rna, bellini_rna_mapping, here(output_dir, "bellini_rnaseq_mapping_stats.csv"))

bellini_rnaseq_mapping_stats <- read_csv2(here(output_dir, "bellini_rnaseq_mapping_stats.csv"))

# Select samples with metadata
bellini_rna <- bellini_rna %>% select(all_of(c("ensembl_gene_id", bellini_metadata_f$CF_sample)))


bellini_rna  %>%
  write_csv2(here(output_dir, "bellini_rnaseq_counts.csv"))
```

# Compare counts with GCF 3B
```{r}
# Parse column names
# parsed_colnames <- colnames(bellini_rna_3B)
# 
# parsed_colnames <- parsed_colnames %>%
#   as.data.frame() %>%
#   setNames("sample_raw")
# 
# cfmp_sample <- str_split(parsed_colnames$sample_raw, "_")
# 
# cfmp_sample <- unlist(cfmp_sample)
# cfmp_sample <- cfmp_sample[cfmp_sample != ""]
# cfmp_sample <- cfmp_sample[str_detect(cfmp_sample, "CF")]
# 
# parsed_colnames$cfmp_sample <- cfmp_sample
# parsed_colnames$cfmp_sample %>% table() %>% as.data.frame() %>% filter(Freq > 1)
# 
# sum(str_detect(parsed_colnames$sample_raw,
#                parsed_colnames$cfmp_sample)) == length(
#                  parsed_colnames$cfmp_sample)
# 
# parsed_colnames$cfmp_sample %in% bellini_metadata_f$CF_sample
# parsed_colnames$cfmp_sample[!parsed_colnames$cfmp_sample %in% bellini_metadata_f$CF_sample]
# 
# # Replace column names
# colnames(bellini_rna_3B)
# colnames(bellini_rna_3B) <- parsed_colnames$cfmp_sample
# colnames(bellini_rna_3B)
# 
# bellini_rna_3B <- bellini_rna_3B[, colnames(bellini_rna_3B) %in% colnames(bellini_rna)] %>% rownames_to_column("ensembl_gene_id")
# bellini_rna_3B <- bellini_rna_3B[1:61541,]
# bellini_rna_3B <- bellini_rna_3B %>%
#   rownames_to_column("ensembl_gene_id")
# 
# bellini_rna_3B %>% filter(!ensembl_gene_id %in% bellini_gcf$ensembl_gene_id) %>%
#   column_to_rownames("ensembl_gene_id") %>% rowSums() %>% summary()
# 
# test <- bellini_rna_3B %>% filter(ensembl_gene_id %in% bellini_gcf$ensembl_gene_id) %>% left_join(bellini_rna, by = "ensembl_gene_id", suffix = c("_htseq", "_gcf"))
# 
# test[is.na(test)] <- 0
# 
# colnames(test)
# 
# test %>% ggplot(aes(x = CF85285_htseq, y = CF85285_gcf)) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
# test %>% ggplot(aes(x = CF85286_htseq, y = CF85286_gcf)) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
# test %>% ggplot(aes(x = CF85288_htseq, y = CF85288_gcf)) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
# test %>% ggplot(aes(x = CF85292_htseq, y = CF85292_gcf)) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
# 
# 
# test %>% ggplot(aes(x = log1p(CF85285_htseq), y = log1p(CF85285_gcf))) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0) 
# test %>% ggplot(aes(x = log1p(CF85286_htseq), y = log1p(CF85286_gcf))) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
# test %>% ggplot(aes(x = log1p(CF85288_htseq), y = log1p(CF85288_gcf))) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)
# test %>% ggplot(aes(x = log1p(CF85292_htseq), y = log1p(CF85292_gcf))) + geom_pointdensity() + theme_Publication() + geom_abline(slope = 1, intercept = 0)

```

# IMCISION
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts
imcision_rna_raw <- read.table(
  here(data_dir, "IMCISION", "RNAseq_output", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
imcision_rna <- imcision_rna_raw
# tail(imcision_rna)

# No missing metadata for RNAseq files
colnames(imcision_rna) %in% imcision_metadata_f$CF_sample
colnames(imcision_rna)[
  !colnames(imcision_rna) %in% imcision_metadata_f$CF_sample]
imcision_metadata_f %>% filter(!CF_sample %in% colnames(imcision_rna))

# Get mapping statistics
imcision_rna_mapping <- imcision_rna[61542 : 61546, ]
imcision_rna <- imcision_rna[1:61541, ]  %>%
  rownames_to_column("ensembl_gene_id")

process_htseq_mapping_stats(imcision_rna, imcision_rna_mapping,
                            here(output_dir, "imcision_rnaseq_mapping_stats.csv"))

imcision_rna %>%
  write_csv2(here(output_dir, "imcision_rnaseq_counts.csv"))
```
# NABUCCO
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts
nabucco_rna_raw <- read.table(
  here(data_dir, "NABUCCO", "RNAseq_output", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
nabucco_rna <- nabucco_rna_raw

nabucco_rna_v2_raw <- read.table(
  here(data_dir, "NABUCCO", "RNAseq_output_v2", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
nabucco_rna_v2 <- nabucco_rna_v2_raw
# tail(nabucco_rna)

# Parse column names
parsed_colnames <- colnames(nabucco_rna)

parsed_colnames <- parsed_colnames %>%
  as.data.frame() %>%
  setNames("sample_raw")

cfmp_sample <- str_split(parsed_colnames$sample_raw, "_")

cfmp_sample <- unlist(cfmp_sample)
cfmp_sample <- cfmp_sample[cfmp_sample != ""]
cfmp_sample <- cfmp_sample[str_detect(cfmp_sample, "CF|unknown")]

parsed_colnames$cfmp_sample <- cfmp_sample

sum(str_detect(parsed_colnames$sample_raw,
               parsed_colnames$cfmp_sample)) == length(
                 parsed_colnames$cfmp_sample)

# Replace column names
colnames(nabucco_rna)
colnames(nabucco_rna) <- parsed_colnames$cfmp_sample
colnames(nabucco_rna)

# Do the same for NABUCCO 2
# Parse column names
parsed_colnames <- colnames(nabucco_rna_v2)

parsed_colnames <- parsed_colnames %>%
  as.data.frame() %>%
  setNames("sample_raw")

cfmp_sample <- str_split(parsed_colnames$sample_raw, "_")

cfmp_sample <- unlist(cfmp_sample)
cfmp_sample <- cfmp_sample[cfmp_sample != ""]
cfmp_sample <- cfmp_sample[str_detect(cfmp_sample, "CF|unknown")]

parsed_colnames$cfmp_sample <- cfmp_sample

sum(str_detect(parsed_colnames$sample_raw,
               parsed_colnames$cfmp_sample)) == length(
                 parsed_colnames$cfmp_sample)

# Replace column names
colnames(nabucco_rna_v2)
colnames(nabucco_rna_v2) <- parsed_colnames$cfmp_sample
colnames(nabucco_rna_v2)

# Missing metadata for some 1 "unknown" RNAseq file
colnames(nabucco_rna)[
  !colnames(nabucco_rna) %in% nabucco_metadata_f$CF_sample]
colnames(nabucco_rna_v2)[
  !colnames(nabucco_rna_v2) %in% nabucco_metadata_f$CF_sample]

nabucco_metadata_f$CF_sample %in% colnames(nabucco_rna)
# 1 sample CF33081 present in both datasets, choose 1
sum(colnames(nabucco_rna) == "CF33081")
sum(colnames(nabucco_rna_v2) == "CF33081")

# Join the RNA data
nabucco_rna_c <- nabucco_rna %>% select(-CF33081) %>% cbind(nabucco_rna_v2)

# Some samples with metadata are missing from RNA
nabucco_metadata_f %>% filter(!CF_sample %in% colnames(nabucco_rna_c))
colnames(nabucco_rna_c)[!colnames(nabucco_rna_c) %in%
                          nabucco_metadata_f$CF_sample]
 
n_distinct(nabucco_metadata_f$CF_sample[
                                 nabucco_metadata_f$CF_sample %in% colnames(nabucco_rna_c)])
# Filter for samples with metadata
nabucco_rna_c <- nabucco_rna_c[,
                               nabucco_metadata_f$CF_sample[
                                 nabucco_metadata_f$CF_sample %in% colnames(nabucco_rna_c)]]

# Get mapping statistics
nabucco_rna_mapping <- nabucco_rna_c[61542 : 61546, ]
nabucco_rna_c <- nabucco_rna_c[1:61541, ] %>%
  rownames_to_column("ensembl_gene_id")

process_htseq_mapping_stats(
  nabucco_rna_c, nabucco_rna_mapping,
  here(output_dir, "nabucco_rnaseq_mapping_stats.csv"))

nabucco_rnaseq_mapping_stats <- read_csv2(
  here(output_dir, "nabucco_rnaseq_mapping_stats.csv"))

# Based on QC I'll remove these samples
# CF66475   Flag Alberto
# CF61419   Flag Alberto
# CF61263   Flag Alberto
# CF64112   Flag Alberto
# CF61102   Flag Alberto
# CF33174   Flag Alberto
# CF33085   Not Flagged but I'll remove
# CF33079   Not Flagged but I'll remove
# CF33077   Not Flagged but I'll remove
# Alberto also flagged "CF43834", "CF64574" but these seem fine to me
# They are distant on the PCA plot though

nabucco_rnaseq_mapping_stats %>%
  filter(CF_sample %in% c("CF66475", "CF61419", "CF61263", "CF64112",
                          "CF61102", "CF33174", "CF33085", "CF33079", "CF33077")) %>%
    pivot_longer(cols = c("no_feature", "ambiguous", "too_low_aQual",
               "not_aligned", "alignment_not_unique", "total_counts"),
               names_to = "type", values_to = "value") %>%
    ggplot(aes(x = CF_sample, y = value, fill = type)) +
    geom_col() +
    coord_flip() + theme_Publication()

nabucco_rna_c_f <- nabucco_rna_c
c("CF66475", "CF61419", "CF61263", "CF64112",
  "CF61102", "CF33174", "CF33085", "CF33079", "CF33077") %in% colnames(nabucco_rna_c)
# nabucco_rna_c_f <- nabucco_rna_c %>%
#   select(-c("CF66475", "CF61419", "CF61263", "CF64112",
#             "CF61102", "CF33174", "CF33085", "CF33079", "CF33077"))

nabucco_rna_c_f %>%
  write_csv2(here(output_dir, "nabucco_rnaseq_counts.csv"))
```


# NICHE
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts v2 did not have the parsed HTseq count matrix
# niche_rna_raw <- read.table(
#   here(data_dir, "NICHE", "RNAseq_output_v2", "Counts",
#        "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
# niche_rna <- niche_rna_raw
# tail(niche_rna)

niche_rna_files <- list.files(here(data_dir, "NICHE", "RNAseq_output_v2",
                                   "Counts"))

for (current_sample in niche_rna_files) {
  print(current_sample)
  
  current_sample_counts <- read.delim(
    here(data_dir, "NICHE", "RNAseq_output_v2", "Counts", current_sample),
    header = FALSE)
  
  if (current_sample == niche_rna_files[1]) {
    niche_rna_raw <- current_sample_counts[, 1, drop = FALSE]
  }
  
  if (all(current_sample_counts$V1 == niche_rna_raw$V1)) {
    niche_rna_raw <- niche_rna_raw %>%
      cbind(current_sample_counts[, 2, drop = FALSE] %>%
            setNames(current_sample))
  } else {
    print(sprintf("Error for sample: %s", current_sample))
    print(sprintf("Ensembl IDs do not match, stopping loop"))
    break
  }
}

niche_rna <- niche_rna_raw
niche_rna <- niche_rna %>% column_to_rownames("V1")

# Parse column names
parsed_colnames <- colnames(niche_rna)

parsed_colnames <- parsed_colnames %>%
  as.data.frame() %>%
  setNames("sample_raw")

cfmp_sample <- str_split(parsed_colnames$sample_raw, "_")
cfmp_sample <- unlist(cfmp_sample)
cfmp_sample <- str_split(cfmp_sample, "\\.")
cfmp_sample <- unlist(cfmp_sample)
cfmp_sample <- str_split(cfmp_sample, "-")
cfmp_sample <- unlist(cfmp_sample)

cfmp_sample <- cfmp_sample[cfmp_sample != ""]
cfmp_sample <- cfmp_sample[str_detect(cfmp_sample, "CF|unknown")]

parsed_colnames$cfmp_sample <- cfmp_sample

sum(str_detect(parsed_colnames$sample_raw,
               parsed_colnames$cfmp_sample)) == length(
                 parsed_colnames$cfmp_sample)

# one sample is present twice CF35017
table(parsed_colnames$cfmp_sample)

parsed_colnames <- parsed_colnames %>%
  mutate(cfmp_sample = ifelse(
    sample_raw == "6894_1_CF35017_GCGCGGTTAA-AAGCTATAGC_S2_HTseqCount.txt",
    "CF35017_6894", cfmp_sample))

# Replace column names
colnames(niche_rna)
colnames(niche_rna) <- parsed_colnames$cfmp_sample
colnames(niche_rna)

# No missing metadata for RNAseq files, only CF63953
colnames(niche_rna)[!colnames(niche_rna) %in% niche_metadata_f$CF_sample]

niche_metadata_f %>% filter(!CF_sample %in% colnames(niche_rna))

niche_metadata_f$CF_sample[!niche_metadata_f$CF_sample %in% colnames(niche_rna)]

# Filter for samples with metadata
niche_rna <- niche_rna[, niche_metadata_f$CF_sample[
  niche_metadata_f$CF_sample %in% colnames(niche_rna)]]
# niche_rna <- niche_rna %>% select(-CF35017_6894)

# Get mapping statistics
niche_rna_mapping <- niche_rna[61542 : 61546, ]
niche_rna <- niche_rna[1:61541, ] %>%
  rownames_to_column("ensembl_gene_id")

process_htseq_mapping_stats(niche_rna, niche_rna_mapping,
                            here(output_dir, "niche_rnaseq_mapping_stats.csv"))

# excluding the resequenced sample CF35017_6894, using the older one
niche_rna %>% 
  write_csv2(here(output_dir, "niche_rnaseq_counts.csv"))
```
# Opacin-NEO
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts
opacin_neo_rna_raw <- read.table(
  here(data_dir, "OpACIN_neo", "RNAseq_output", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
opacin_neo_rna <- opacin_neo_rna_raw
# tail(opacin_neo_rna)

# Parse column names
parsed_colnames <- colnames(opacin_neo_rna)

parsed_colnames <- parsed_colnames %>%
  as.data.frame() %>%
  setNames("sample_raw")

cfmp_sample <- str_split(parsed_colnames$sample_raw, "_")

cfmp_sample <- unlist(cfmp_sample)
cfmp_sample <- cfmp_sample[cfmp_sample != ""]
cfmp_sample <- cfmp_sample[str_detect(cfmp_sample, "CF|unknown")]

parsed_colnames$cfmp_sample <- cfmp_sample

sum(str_detect(parsed_colnames$sample_raw,
               parsed_colnames$cfmp_sample)) == length(
                 parsed_colnames$cfmp_sample)

# Replace column names
colnames(opacin_neo_rna)
colnames(opacin_neo_rna) <- parsed_colnames$cfmp_sample
colnames(opacin_neo_rna)

# Missing metadata for some RNAseq files
colnames(opacin_neo_rna) %in% opacin_neo_metadata_f$CF_sample

colnames(opacin_neo_rna)[
  !colnames(opacin_neo_rna) %in% opacin_neo_metadata_f$CF_sample]

opacin_neo_metadata_f$CF_sample %in% colnames(opacin_neo_rna) 

opacin_neo_metadata_f %>% filter(!CF_sample %in% colnames(opacin_neo_rna))


# Filter for samples with metadata
opacin_neo_rna <- opacin_neo_rna[, opacin_neo_metadata_f$CF_sample]

# Get mapping statistics
opacin_neo_rna_mapping <- opacin_neo_rna[61542 : 61546, ]
opacin_neo_rna <- opacin_neo_rna[1:61541, ] %>%
  rownames_to_column("ensembl_gene_id")

process_htseq_mapping_stats(opacin_neo_rna, opacin_neo_rna_mapping, here(output_dir, "opacin_neo_rnaseq_mapping_stats.csv"))

opacin_neo_rna  %>%
  write_csv2(here(output_dir, "opacin_neo_rnaseq_counts.csv"))

```
# PRADO
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts
prado_rna_raw <- read.table(
  here(data_dir, "PRADO", "RNAseq_output", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
prado_rna <- prado_rna_raw
# tail(prado_rna)

# Missing metadata for many RNAseq files
colnames(prado_rna) %in% prado_metadata_f$CF_sample

colnames(prado_rna)[
  !colnames(prado_rna) %in% prado_metadata_f$CF_sample]

prado_metadata_f$CF_sample %in% colnames(prado_rna) 

prado_metadata_f %>% filter(!CF_sample %in% colnames(prado_rna))

# Filter for samples with metadata
prado_rna <- prado_rna[, prado_metadata_f$CF_sample]

# Get mapping statistics
prado_rna_mapping <- prado_rna[61542 : 61546, ]
prado_rna <- prado_rna[1:61541, ] %>%
  rownames_to_column("ensembl_gene_id")

process_htseq_mapping_stats(prado_rna, prado_rna_mapping,
                            here(output_dir, "prado_rnaseq_mapping_stats.csv"))

prado_rna %>%
  write_csv2(here(output_dir, "prado_rnaseq_counts.csv"))

```

# MATISSE
```{r, fig.width = 6, fig.height = 15}
# Load RNAseq counts
matisse_rna_raw <- read.table(
  here(data_dir, "MATISSE", "RNAseq_output", "Counts",
       "Count_matrix_HTseq.txt"), row.names = 1, sep = "\t")
matisse_rna <- matisse_rna_raw
# tail(matisse_rna)

# No missing metadata for RNAseq files
colnames(matisse_rna) %in% matisse_metadata_f$CF_sample

colnames(matisse_rna)[
  !colnames(matisse_rna) %in% matisse_metadata_f$CF_sample]

matisse_metadata_f$CF_sample %in% colnames(matisse_rna) 

matisse_metadata_f %>% filter(!CF_sample %in% colnames(matisse_rna))

# Filter for samples with metadata
matisse_rna <- matisse_rna[, matisse_metadata_f$CF_sample]

# Get mapping statistics
matisse_rna_mapping <- matisse_rna[61542 : 61546, ]
matisse_rna <- matisse_rna[1:61541, ] %>%
  rownames_to_column("ensembl_gene_id")

process_htseq_mapping_stats(matisse_rna, matisse_rna_mapping,
                            here(output_dir,
                                 "matisse_rnaseq_mapping_stats.csv"))

matisse_rna %>%
  write_csv2(here(output_dir, "matisse_rnaseq_counts.csv"))

```

# Compare mapping data across cohorts
```{r}
bellini_rna_mapping <- read_csv2(
  here(output_dir, "bellini_rnaseq_mapping_stats.csv"))
imcision_rna_mapping <- read_csv2(
  here(output_dir, "imcision_rnaseq_mapping_stats.csv"))
nabucco_rna_mapping <- read_csv2(
  here(output_dir, "nabucco_rnaseq_mapping_stats.csv"))
niche_rna_mapping <- read_csv2(
  here(output_dir, "niche_rnaseq_mapping_stats.csv"))
opacin_neo_rna_mapping <- read_csv2(
  here(output_dir, "opacin_neo_rnaseq_mapping_stats.csv"))
prado_rna_mapping <- read_csv2(
  here(output_dir, "prado_rnaseq_mapping_stats.csv"))
matisse_rna_mapping <- read_csv2(
  here(output_dir, "matisse_rnaseq_mapping_stats.csv"))
  
all_cohort_rna_mapping <- bellini_rna_mapping %>% mutate(Trial = "BELLINI") %>%
  full_join(imcision_rna_mapping %>% mutate(Trial = "IMCISION")) %>%
  full_join(nabucco_rna_mapping %>% mutate(Trial = "NABUCCO")) %>%
  full_join(niche_rna_mapping %>% mutate(Trial = "NICHE")) %>%
  full_join(opacin_neo_rna_mapping %>% mutate(Trial = "OpACIN_neo")) %>%
  full_join(prado_rna_mapping %>% mutate(Trial = "PRADO"))%>%
  full_join(matisse_rna_mapping %>% mutate(Trial = "MATISSE"))

all_cohort_rna_mapping <- all_cohort_rna_mapping %>% arrange(Trial)

all_cohort_rna_mapping_l <- all_cohort_rna_mapping %>%
    pivot_longer(cols = c("no_feature", "ambiguous", "too_low_aQual",
               "not_aligned", "alignment_not_unique", "total_counts"),
               names_to = "type", values_to = "value") %>%
    group_by(CF_sample) %>%
    mutate(value_perc = value / sum(value))

 all_cohort_rna_mapping_l %>%
    ggplot(aes(x = Trial, y = log10(value), fill = type)) +
    geom_boxplot() +
  theme_Publication()

all_cohort_rna_mapping_l %>%
    ggplot(aes(x = Trial, y = value_perc, fill = type)) +
    geom_boxplot() +
  theme_Publication()

all_cohort_rna_mapping %>%
  ggplot(aes(x = Trial, y = log10(total_counts), fill = Trial)) +
  geom_boxplot() +
  theme_Publication()

all_cohort_rna_mapping_l %>%
  filter(type == "total_counts", Trial == "IMCISION") %>%
  summary()
```
# Load rnaseq data
```{r}
bellini_rna <- read_csv2(here(output_dir, "bellini_rnaseq_counts.csv"))
imcision_rna <- read_csv2(here(output_dir, "imcision_rnaseq_counts.csv"))
nabucco_rna <- read_csv2(here(output_dir, "nabucco_rnaseq_counts.csv"))
niche_rna <- read_csv2(here(output_dir, "niche_rnaseq_counts.csv"))
opacin_neo_rna <- read_csv2(here(output_dir, "opacin_neo_rnaseq_counts.csv"))
prado_rna <- read_csv2(here(output_dir, "prado_rnaseq_counts.csv"))
matisse_rna <- read_csv2(here(output_dir, "matisse_rnaseq_counts.csv"))
```

# ensembl id map
```{r}
nrow(bellini_rna)
all.equal(nrow(bellini_rna), nrow(imcision_rna), nrow(nabucco_rna),
          nrow(niche_rna), nrow(opacin_neo_rna), nrow(prado_rna), 
          nrow(matisse_rna))
all.equal(bellini_rna$ensembl_gene_id, imcision_rna$ensembl_gene_id,
          nabucco_rna$ensembl_gene_id, niche_rna$ensembl_gene_id,
          opacin_neo_rna$ensembl_gene_id, prado_rna$ensembl_gene_id,
          matisse_rna$ensembl_gene_id)

combined_rna <- bellini_rna %>%
  left_join(imcision_rna) %>%
  left_join(nabucco_rna) %>%
  left_join(niche_rna) %>%
  left_join(opacin_neo_rna) %>%
  left_join(prado_rna) %>% 
  left_join(matisse_rna)

ensembl_id_set <- unique(combined_rna$ensembl_gene_id)

gene_features <- read.csv2(
  file = here("data", "external", "biomaRt_gene_features.csv"))

gene_metadata <- gene_features %>%
  filter(ensembl_gene_id %in% ensembl_id_set) %>%
  select(ensembl_gene_id, gene_biotype, chromosome_name, start_position,
         end_position, description) %>% unique()

gene_map <- bellini_rna %>%
  select(ensembl_gene_id)  %>%
  left_join(gene_features %>%
              select(ensembl_gene_id, hgnc_symbol) %>%
              unique()) %>%
  unique()

# Two ensembl IDs have 2 different gene symbols
table(gene_map$ensembl_gene_id) %>% as.data.frame() %>% filter(Freq > 1)
gene_map %>% filter(ensembl_gene_id == "ENSG00000230417")
gene_map %>% filter(ensembl_gene_id == "ENSG00000276085")

# Choosing one of the 2 symbols for each (CCL3L1 and STRA6LP, LINC00595)
# discarding the others
gene_map <- gene_map %>%
  filter(!hgnc_symbol %in% c("CCL3L3", "LINC00856")) %>%
  unique()

# There are ensembl IDs without gene symbols (mostly novel transcripts)
gene_map %>% filter(hgnc_symbol == "") %>% nrow()

table(gene_map$hgnc_symbol) %>% as.data.frame() %>% filter(Freq > 1)


```

# Map ensembl IDs to gene symbols
```{r}
combined_rna_p <- map_ensembl_ids(
  combined_rna, gene_map, here(output_dir, "combined_excluded_genes.RDS"),
  here(output_dir, "combined_rnaseq_counts_p.csv"))

bellini_rna_p <- map_ensembl_ids(
  bellini_rna, gene_map, here(output_dir, "bellini_excluded_genes.RDS"),
  here(output_dir, "bellini_rnaseq_counts_p.csv"))

imcision_rna_p <- map_ensembl_ids(
  imcision_rna, gene_map, here(output_dir, "imcision_excluded_genes.RDS"),
  here(output_dir, "imcision_rnaseq_counts_p.csv"))
  
nabucco_rna_p <- map_ensembl_ids(
  nabucco_rna, gene_map, here(output_dir, "nabucco_excluded_genes.RDS"),
  here(output_dir, "nabucco_rnaseq_counts_p.csv"))

niche_rna_p <- map_ensembl_ids(
  niche_rna, gene_map, here(output_dir, "niche_excluded_genes.RDS"),
  here(output_dir,  "niche_rnaseq_counts_p.csv"))

opacin_neo_rna_p <- map_ensembl_ids(
  opacin_neo_rna, gene_map, here(output_dir, "opacin_neo_excluded_genes.RDS"),
  here(output_dir, "opacin_neo_rnaseq_counts_p.csv"))

prado_rna_p <- map_ensembl_ids(
  prado_rna, gene_map, here(output_dir, "prado_excluded_genes.RDS"),
  here(output_dir, "prado_rnaseq_counts_p.csv"))

matisse_rna_p <- map_ensembl_ids(
  matisse_rna, gene_map, here(output_dir, "matisse_excluded_genes.RDS"),
  here(output_dir, "matisse_rnaseq_counts_p.csv"))

combined_rna_p <- read_csv2(
  here(output_dir, "combined_rnaseq_counts_p.csv")) %>%
  column_to_rownames("gene")

sum(colnames(combined_rna_p) %in% all_cohort_rna_metadata$CF_sample)
colnames(combined_rna_p) [!colnames(combined_rna_p) %in% all_cohort_rna_metadata$CF_sample]

all_cohort_rna_metadata$CF_sample[
  !all_cohort_rna_metadata$CF_sample %in% colnames(combined_rna_p)]
dim(combined_rna_p)
```

# Inspect QC and rRNA counts
```{r}
human_rrna_genes <- read.csv2(
  file = here("data", "external", "human_rRNA_genes.csv"))

bellini_rrna <- get_rrna_fraction(bellini_rna_p, human_rrna_genes) %>%
  mutate(Trial = "BELLINI")

imcision_rrna <- get_rrna_fraction(imcision_rna_p, human_rrna_genes) %>%
  mutate(Trial = "IMCISION")

nabucco_rrna <- get_rrna_fraction(nabucco_rna_p, human_rrna_genes) %>%
  mutate(Trial = "NABUCCO")

niche_rrna <- get_rrna_fraction(niche_rna_p, human_rrna_genes) %>%
  mutate(Trial = "NICHE")

opacin_neo_rrna <- get_rrna_fraction(opacin_neo_rna_p, human_rrna_genes) %>%
  mutate(Trial = "OpACIN_neo")

prado_rrna <- get_rrna_fraction(opacin_neo_rna_p, human_rrna_genes) %>%
  mutate(Trial = "PRADO")

matisse_rrna <- get_rrna_fraction(matisse_rna_p, human_rrna_genes) %>%
  mutate(Trial = "MATISSE")

rrna_c <- bellini_rrna %>%
  rbind(imcision_rrna) %>%
  rbind(nabucco_rrna) %>%
  rbind(niche_rrna) %>%
  rbind(opacin_neo_rrna) %>%
  rbind(prado_rrna) %>%
  rbind(matisse_rrna)

 rrna_c %>%
    ggplot(aes(x = Trial, y = fraction_rRNA, fill = Trial)) +
    geom_boxplot() +
  theme_Publication()
```

# Calculate RPM
```{r}
bellini_rpm_counts <- calculate_rpm(
  bellini_rna_p, here(output_dir, "bellini_rnaseq_rpm.csv"))
imcision_rpm_counts <- calculate_rpm(
  imcision_rna_p, here(output_dir, "imcision_rnaseq_rpm.csv"))
nabucco_rpm_counts <- calculate_rpm(
  nabucco_rna_p, here(output_dir, "nabucco_rnaseq_rpm.csv"))
niche_rpm_counts <- calculate_rpm(
  niche_rna_p, here(output_dir, "niche_rnaseq_rpm.csv"))
opacin_neo_rpm_counts <- calculate_rpm(
  opacin_neo_rna_p, here(output_dir, "opacin_neo_rnaseq_rpm.csv"))
prado_rpm_counts <- calculate_rpm(
  prado_rna_p, here(output_dir, "prado_rnaseq_rpm.csv"))
matisse_rpm_counts <- calculate_rpm(
  matisse_rna_p, here(output_dir, "matisse_rnaseq_rpm.csv"))

combined_rpm_counts <- calculate_rpm(
  combined_rna_p, here(output_dir, "combined_rnaseq_rpm.csv"))

colSums(combined_rpm_counts)
```

# Estimate TPM
```{r}
# # Get information of gene lengths
# gene_features_f <- gene_features %>%
#   filter(chromosome_name %in%
#            c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
#              15, 16, 17, 18, 19, 20, 21, 22, "X", "Y", "MT")) %>%
#   select(-entrezgene_id) %>%
#   filter(hgnc_symbol %in% rownames(df_counts) |
#            ensembl_gene_id %in% rownames(df_counts)) %>%
#   mutate(gene_length = end_position - start_position) %>%
#   unique()
# 
# 
# # Check unmapped genes, likely outside the main chromosomes
# unmapped <- rownames(df_counts)[
#   !rownames(df_counts) %in%
#     c(gene_features_f$hgnc_symbol, gene_features_f$ensembl_gene_id)]
# 
# unmapped %in% c(gene_metadata$ensembl_gene_id, gene_metadata$external_gene_id)
# 
# # Calculate the lengths for these missing genes and add them
# missing_lengths_rnaseq <- gene_metadata %>%
#   filter(external_gene_id %in% unmapped | ensembl_gene_id %in% unmapped) %>%
#   mutate(gene_length = end_position - start_position) %>%
#   select(c(ensembl_gene_id, gene_biotype, chromosome_name, start_position,
#            end_position, external_gene_id, description, gene_length)) %>%
#   rename(hgnc_symbol = external_gene_id) %>%
#   mutate(ensembl_gene_id_version = ensembl_gene_id)
# 
# missing_lengths_rnaseq <- missing_lengths_rnaseq[, colnames(gene_features_f)]
# 
# gene_features_f <- gene_features_f %>%
#   rbind(missing_lengths_rnaseq) %>%
#   mutate(
#     gene = ifelse(
#       hgnc_symbol %in% c("") |
#         !hgnc_symbol %in% df_counts_map$gene, ensembl_gene_id, hgnc_symbol)) %>%
#   unique()
# 
# # A few genes may be duplicated, summarize and get the max gene length
# duplicated <- table(gene_features_f$gene) %>%
#   as.data.frame() %>%
#   filter(Freq > 1)
# 
# gene_lengths <- gene_features_f %>%
#   dplyr::group_by(gene) %>%
#   dplyr::summarize(gene_length = max(gene_length)) %>%
#   mutate(gene_length_kb = gene_length / 1000)
# 
# 
# gene_lengths %>% summary()
# 
# gene_lengths %>%
#   ggplot(aes(x = log10(gene_length_kb))) +
#   geom_histogram()
# 
# df_counts_map_f <- df_counts_map[
#   df_counts_map$ensembl_gene_id %in% rownames(df_counts) |
#     df_counts_map$external_gene_id %in% rownames(df_counts), ]
# 
# # Remove genes that are still unmapped, may be due to different genome builds
# df_counts_f <- df_counts[rownames(df_counts) %in% c(gene_lengths$gene), ]
# 
# # Reorder gene_lengths to ensure it has the same order
# gene_lengths <- rownames(df_counts) %>%
#   as.data.frame() %>%
#   setNames("gene") %>%
#   left_join(gene_lengths) %>%
#   filter(!is.na(gene_length))
# 
# nrow(df_counts_f) == sum(rownames(df_counts_f) == gene_lengths$gene)
# 
# # Calculate TPM
# rpk_counts <- sweep(x = df_counts_f,
#                     STATS = gene_lengths$gene_length_kb, MARGIN = 1, FUN = "/")
# 
# tpm_counts <- sweep(x = rpk_counts,
#                     STATS = colSums(rpk_counts) / 1e6, MARGIN = 2, FUN = "/")
# colSums(tpm_counts)
```

```{r}
# write.table(rpm_counts %>%
#   rownames_to_column("gene"), file = here(output_dir, "rnaseq_rpm.tsv"),
#             sep = "\t", row.names = FALSE)
# 
# tpm_counts %>%
#   rownames_to_column("gene") %>%
#   write_csv2(here(output_dir, "rnaseq_tpm.csv"))
# 
# write.table(tpm_counts %>%
#   rownames_to_column("gene"),
#             file = here(output_dir, "rnaseq_tpm.tsv"),
#             sep = "\t", row.names = FALSE)
# 
# rpm_counts <- read_csv2(here(output_dir, "rnaseq_rpm.csv"))
# 
# rpm_logpcounts <- rpm_counts %>%
#   column_to_rownames("gene")
# 
# rpm_logpcounts <- log2(rpm_logpcounts + 1)
# 
# rpm_logpcounts %>%
#   rownames_to_column("GENE") %>%
#   write_delim(here(output_dir, "EXPR.txt"), delim = "\t")
# 
# df_counts <- read.csv2(here(output_dir, "rnaseq_counts_p.csv"))
# df_counts <- df_counts %>%
#   column_to_rownames("gene")
```

# sessionInfo - Package Versions
```{r sessionInfo}
sessionInfo()
```
